# HG changeset patch
# Parent  404f34ae6db7a10b8b119ae997941337c693c480

diff --git a/src/fileio.c b/src/fileio.c
--- a/src/fileio.c
+++ b/src/fileio.c
@@ -5234,6 +5234,9 @@ msg_add_lines(
 #ifdef LONG_LONG_OFF_T
 	sprintf((char *)p,
 		"%ldL, %lldC", lnum, (long long)nchars);
+#elif defined(WIN3264)
+	sprintf((char *)p,
+		"%ldL, %I64dC", lnum, (__int64)nchars);
 #else
 	sprintf((char *)p,
 		/* Explicit typecast avoids warning on Mac OS X 10.6 */
@@ -5252,6 +5255,9 @@ msg_add_lines(
 #ifdef LONG_LONG_OFF_T
 	    sprintf((char *)p,
 		    _("%lld characters"), (long long)nchars);
+#elif defined(WIN3264)
+	    sprintf((char *)p,
+		    _("%I64d characters"), (__int64)nchars);
 #else
 	    sprintf((char *)p,
 		    /* Explicit typecast avoids warning on Mac OS X 10.6 */
