# HG changeset patch
# Parent 12e327f895c2ed1ecc0072b3c82dd01875d3bde1

diff --git a/src/Make_cyg.mak b/src/Make_cyg.mak
--- a/src/Make_cyg.mak
+++ b/src/Make_cyg.mak
@@ -193,31 +193,37 @@
 ##############################
 ifdef RUBY
 
+ifndef DYNAMIC_RUBY
+DYNAMIC_RUBY=yes
+endif
+#  Set default value
 ifndef RUBY_VER
-RUBY_VER=16
+RUBY_VER = 16
+endif
+ifndef RUBY_VER_LONG
+RUBY_VER_LONG = 1.6
+endif
+ifndef RUBY_API_VER
+RUBY_API_VER = $(subst .,,$(RUBY_VER_LONG))
 endif
 
-ifndef RUBY_VER_LONG
-RUBY_VER_LONG=1.6
+ifndef RUBY_PLATFORM
+ifeq ($(RUBY_VER), 16)
+RUBY_PLATFORM = i586-mswin32
+else
+ifneq ($(wildcard $(RUBY)/lib/ruby/$(RUBY_VER_LONG)/i386-mingw32),)
+RUBY_PLATFORM = i386-mingw32
+else
+RUBY_PLATFORM = i386-mswin32
+endif
+endif
 endif
 
-ifndef DYNAMIC_RUBY
-DYNAMIC_RUBY = yes
-endif
-
+ifndef RUBY_INSTALL_NAME
 ifeq ($(RUBY_VER), 16)
-ifndef RUBY_PLATFORM
-RUBY_PLATFORM = i586-mswin32
-endif
-ifndef RUBY_INSTALL_NAME
-RUBY_INSTALL_NAME = mswin32-ruby$(RUBY_VER)
-endif
+RUBY_INSTALL_NAME = mswin32-ruby$(RUBY_API_VER)
 else
-ifndef RUBY_PLATFORM
-RUBY_PLATFORM = i386-mswin32
-endif
-ifndef RUBY_INSTALL_NAME
-RUBY_INSTALL_NAME = msvcrt-ruby$(RUBY_VER)
+RUBY_INSTALL_NAME = msvcrt-ruby$(RUBY_API_VER)
 endif
 endif
 
