From f53d4f475b1d984ad736f6378a57a35517f6e663 Mon Sep 17 00:00:00 2001
From: "K.Takata" <kentkt@csc.jp>
Date: Tue, 2 Apr 2019 19:40:19 +0900
Subject: [PATCH] Remove unnecessary check for ANSI API

---
 src/misc1.c | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/src/misc1.c b/src/misc1.c
index c19a6528c..3e41eb661 100644
--- a/src/misc1.c
+++ b/src/misc1.c
@@ -5784,12 +5784,7 @@ dos_expandpath(
     STRCPY(s, "*.*");
     wn = enc_to_utf16(buf, NULL);
     if (wn != NULL)
-    {
 	hFind = FindFirstFileW(wn, &wfb);
-	if (hFind == INVALID_HANDLE_VALUE
-			  && GetLastError() == ERROR_CALL_NOT_IMPLEMENTED)
-	    VIM_CLEAR(wn);
-    }
     ok = (hFind != INVALID_HANDLE_VALUE);
 
     while (ok)
-- 
2.17.0

