[PATCH 25/25] kaoriya-feat-kaoriya

diff --git a/src/eval.c b/src/eval.c
--- a/src/eval.c
+++ b/src/eval.c
@@ -12220,6 +12220,7 @@
 #ifdef FEAT_JUMPLIST
 	"jumplist",
 #endif
+	"kaoriya",
 #ifdef FEAT_KEYMAP
 	"keymap",
 #endif
diff --git a/src/version.c b/src/version.c
--- a/src/version.c
+++ b/src/version.c
@@ -285,6 +285,7 @@
 #else
 	"-keymap",
 #endif
+	"+kaoriya",
 #ifdef FEAT_LANGMAP
 	"+langmap",
 #else
