# HG changeset patch
# Parent  7122a1b0a7aa78d56f2547986edfb79eceaddc36

diff --git a/src/ops.c b/src/ops.c
--- a/src/ops.c
+++ b/src/ops.c
@@ -5475,8 +5475,8 @@ do_addsub(op_type, pos, length, Prenum1)
     char_u	buf2[NUMBUFLEN];
     int		pre;		/* 'X'/'x': hex; '0': octal; 'B'/'b': bin */
     static int	hexupper = FALSE;	/* 0xABC */
-    unsigned long n;
-    long_u	oldn;
+    uvarnumber_T	n;
+    uvarnumber_T	oldn;
     char_u	*ptr;
     int		c;
     int		todel;
@@ -5682,9 +5682,9 @@ do_addsub(op_type, pos, length, Prenum1)
 
 	oldn = n;
 	if (subtract)
-	    n -= (unsigned long)Prenum1;
+	    n -= (uvarnumber_T)Prenum1;
 	else
-	    n += (unsigned long)Prenum1;
+	    n += (uvarnumber_T)Prenum1;
 	/* handle wraparound for decimal numbers */
 	if (!pre)
 	{
@@ -5692,7 +5692,7 @@ do_addsub(op_type, pos, length, Prenum1)
 	    {
 		if (n > oldn)
 		{
-		    n = 1 + (n ^ (unsigned long)-1);
+		    n = 1 + (n ^ (uvarnumber_T)-1);
 		    negative ^= TRUE;
 		}
 	    }
@@ -5701,7 +5701,7 @@ do_addsub(op_type, pos, length, Prenum1)
 		/* add */
 		if (n < oldn)
 		{
-		    n = (n ^ (unsigned long)-1);
+		    n = (n ^ (uvarnumber_T)-1);
 		    negative ^= TRUE;
 		}
 	    }
@@ -5777,7 +5777,7 @@ do_addsub(op_type, pos, length, Prenum1)
 	{
 	    int i;
 	    int bit = 0;
-	    int bits = sizeof(unsigned long) * 8;
+	    int bits = sizeof(uvarnumber_T) * 8;
 
 	    /* leading zeros */
 	    for (bit = bits; bit > 0; bit--)
@@ -5789,13 +5789,13 @@ do_addsub(op_type, pos, length, Prenum1)
 	    buf2[i] = '\0';
 	}
 	else if (pre == 0)
-	    sprintf((char *)buf2, "%lu", n);
+	    sprintf((char *)buf2, "%"VARNUMBER_FMT"u", n);
 	else if (pre == '0')
-	    sprintf((char *)buf2, "%lo", n);
+	    sprintf((char *)buf2, "%"VARNUMBER_FMT"o", n);
 	else if (pre && hexupper)
-	    sprintf((char *)buf2, "%lX", n);
+	    sprintf((char *)buf2, "%"VARNUMBER_FMT"X", n);
 	else
-	    sprintf((char *)buf2, "%lx", n);
+	    sprintf((char *)buf2, "%"VARNUMBER_FMT"x", n);
 	length -= (int)STRLEN(buf2);
 
 	/*
