# HG changeset patch
# Parent  22c94b1ecd6978a2bbc710bda82d47d2c02dca24

diff --git a/src/option.c b/src/option.c
--- a/src/option.c
+++ b/src/option.c
@@ -8651,6 +8651,11 @@ set_bool_option(
     /* 'termguicolors' */
     else if ((int *)varp == &p_tgc)
     {
+# if !defined(FEAT_GUI) && defined(WIN3264)
+	/* Do not turn on 'tgc' when 24-bit colors are not supported. */
+	if (!vtp_get_stat())
+	    p_tgc = 0;
+# endif
 # ifdef FEAT_GUI
 	if (!gui.in_use && !gui.starting)
 # endif
