# HG changeset patch
# Parent  663e22df8ae9f6f97792dd93a8d6a198390612f3

diff --git a/src/misc2.c b/src/misc2.c
--- a/src/misc2.c
+++ b/src/misc2.c
@@ -2057,6 +2057,8 @@ ga_grow(garray_T *gap, int n)
     {
 	if (n < gap->ga_growsize)
 	    n = gap->ga_growsize;
+	if (n < gap->ga_len / 2)
+	    n = gap->ga_len / 2;
 	new_len = gap->ga_itemsize * (gap->ga_len + n);
 	pp = vim_realloc(gap->ga_data, new_len);
 	if (pp == NULL)
