# HG changeset patch
# Parent  07f8a2d169a0bdf696679c7f6dc599a4da809447

diff --git a/src/arglist.c b/src/arglist.c
--- a/src/arglist.c
+++ b/src/arglist.c
@@ -969,9 +969,9 @@ do_arg_all(
     old_curwin = curwin;
     old_curtab = curtab;
 
-# ifdef FEAT_GUI
+#ifdef FEAT_GUI
     need_mouse_correct = TRUE;
-# endif
+#endif
 
     // Try closing all windows that are not in the argument list.
     // Also close windows that are not full width;
diff --git a/src/change.c b/src/change.c
--- a/src/change.c
+++ b/src/change.c
@@ -1638,9 +1638,9 @@ open_line(
 #ifdef FEAT_CINDENT
     // May do indenting after opening a new line.
     do_cindent = !p_paste && (curbuf->b_p_cin
-#  ifdef FEAT_EVAL
+# ifdef FEAT_EVAL
 		    || *curbuf->b_p_inde != NUL
-#  endif
+# endif
 		)
 	    && in_cinkeys(dir == FORWARD
 		? KEY_OPEN_FORW
diff --git a/src/ex_cmds.c b/src/ex_cmds.c
--- a/src/ex_cmds.c
+++ b/src/ex_cmds.c
@@ -99,9 +99,9 @@ do_ascii(exarg_T *eap UNUSED)
 	    IObuff[len++] = ' ';
 	IObuff[len++] = '<';
 	if (enc_utf8 && utf_iscomposing(c)
-# ifdef USE_GUI
+#ifdef USE_GUI
 		&& !gui.in_use
-# endif
+#endif
 		)
 	    IObuff[len++] = ' '; // draw composing char on top of a space
 	len += (*mb_char2bytes)(c, IObuff + len);
diff --git a/src/gui.c b/src/gui.c
--- a/src/gui.c
+++ b/src/gui.c
@@ -292,11 +292,11 @@ gui_do_fork(void)
     }
     // Child
 
-#ifdef FEAT_GUI_GTK
+# ifdef FEAT_GUI_GTK
     // Call gtk_init_check() here after fork(). See gui_init_check().
     if (gui_mch_init_check() != OK)
 	getout_preserve_modified(1);
-#endif
+# endif
 
 # if defined(HAVE_SETSID) || defined(HAVE_SETPGID)
     /*
@@ -348,11 +348,11 @@ gui_do_fork(void)
 gui_read_child_pipe(int fd)
 {
     long	bytes_read;
-#define READ_BUFFER_SIZE 10
+# define READ_BUFFER_SIZE 10
     char	buffer[READ_BUFFER_SIZE];
 
     bytes_read = read_eintr(fd, buffer, READ_BUFFER_SIZE - 1);
-#undef READ_BUFFER_SIZE
+# undef READ_BUFFER_SIZE
     close(fd);
     if (bytes_read < 0)
 	return GUI_CHILD_IO_ERROR;
@@ -459,7 +459,7 @@ gui_init_check(void)
     gui.scrollbar_width = gui.scrollbar_height = SB_DEFAULT_WIDTH;
     gui.prev_wrap = -1;
 
-# ifdef FEAT_GUI_GTK
+#ifdef FEAT_GUI_GTK
     CLEAR_FIELD(gui.ligatures_map);
 #endif
 
@@ -1435,7 +1435,7 @@ gui_position_components(int total_width 
 	text_area_y += gui.menu_height;
 #endif
 
-# if defined(FEAT_GUI_TABLINE) && (defined(FEAT_GUI_MSWIN) \
+#if defined(FEAT_GUI_TABLINE) && (defined(FEAT_GUI_MSWIN) \
 	|| defined(FEAT_GUI_MOTIF))
     if (gui_has_tabline())
 	text_area_y += gui.tabline_height;
@@ -1453,7 +1453,7 @@ gui_position_components(int total_width 
     }
 #endif
 
-# if defined(FEAT_GUI_TABLINE) && defined(FEAT_GUI_HAIKU)
+#if defined(FEAT_GUI_TABLINE) && defined(FEAT_GUI_HAIKU)
     gui_mch_set_tabline_pos(0, text_area_y,
     gui.menu_width, gui.tabline_height);
     if (gui_has_tabline())
@@ -5235,10 +5235,10 @@ gui_update_screen(void)
     {
 	if (has_cursormoved())
 	    apply_autocmds(EVENT_CURSORMOVED, NULL, NULL, FALSE, curbuf);
-#ifdef FEAT_PROP_POPUP
+# ifdef FEAT_PROP_POPUP
 	if (popup_visible)
 	    popup_check_cursor_pos();
-#endif
+# endif
 # ifdef FEAT_CONCEAL
 	if (curwin->w_p_cole > 0)
 	{
diff --git a/src/hashtab.c b/src/hashtab.c
--- a/src/hashtab.c
+++ b/src/hashtab.c
@@ -189,13 +189,13 @@ hash_lookup(hashtab_T *ht, char_u *key, 
     void
 hash_debug_results(void)
 {
-#ifdef HT_DEBUG
+# ifdef HT_DEBUG
     fprintf(stderr, "\r\n\r\n\r\n\r\n");
     fprintf(stderr, "Number of hashtable lookups: %ld\r\n", hash_count_lookup);
     fprintf(stderr, "Number of perturb loops: %ld\r\n", hash_count_perturb);
     fprintf(stderr, "Percentage of perturb loops: %ld%%\r\n",
 				hash_count_perturb * 100 / hash_count_lookup);
-#endif
+# endif
 }
 #endif
 
diff --git a/src/indent.c b/src/indent.c
--- a/src/indent.c
+++ b/src/indent.c
@@ -2173,7 +2173,7 @@ f_indent(typval_T *argvars, typval_T *re
     void
 f_lispindent(typval_T *argvars UNUSED, typval_T *rettv)
 {
-#ifdef FEAT_LISP
+# ifdef FEAT_LISP
     pos_T	pos;
     linenr_T	lnum;
 
@@ -2191,7 +2191,7 @@ f_lispindent(typval_T *argvars UNUSED, t
     else if (in_vim9script())
 	semsg(_(e_invalid_line_number_nr), lnum);
     else
-#endif
+# endif
 	rettv->vval.v_number = -1;
 }
 #endif
diff --git a/src/ops.c b/src/ops.c
--- a/src/ops.c
+++ b/src/ops.c
@@ -3667,9 +3667,9 @@ do_pending_operator(cmdarg_T *cap, int o
 		curbuf->b_visual.vi_mode = VIsual_mode;
 		restore_visual_mode();
 		curbuf->b_visual.vi_curswant = curwin->w_curswant;
-# ifdef FEAT_EVAL
+#ifdef FEAT_EVAL
 		curbuf->b_visual_mode_eval = VIsual_mode;
-# endif
+#endif
 	    }
 
 	    // In Select mode, a linewise selection is operated upon like a
diff --git a/src/os_win32.c b/src/os_win32.c
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -241,11 +241,11 @@ static WCHAR *exe_pathw = NULL;
 
 static BOOL win8_or_later = FALSE;
 
-# if defined(__GNUC__) && !defined(__MINGW32__)  && !defined(__CYGWIN__)
-#  define UChar UnicodeChar
-# else
-#  define UChar uChar.UnicodeChar
-# endif
+#if defined(__GNUC__) && !defined(__MINGW32__)  && !defined(__CYGWIN__)
+# define UChar UnicodeChar
+#else
+# define UChar uChar.UnicodeChar
+#endif
 
 /*
  * Get version number including build number
@@ -2078,13 +2078,13 @@ theend:
 	buf[len++] = typeahead[0];
 	mch_memmove(typeahead, typeahead + 1, --typeaheadlen);
     }
-#  ifdef FEAT_JOB_CHANNEL
+# ifdef FEAT_JOB_CHANNEL
     if (len > 0)
     {
 	buf[len] = NUL;
 	ch_log(NULL, "raw key input: \"%s\"", buf);
     }
-#  endif
+# endif
     return len;
 
 #else // FEAT_GUI_MSWIN
@@ -7858,12 +7858,12 @@ vtp_sgr_bulk(
     vtp_sgr_bulks(1, args);
 }
 
-#define FAST256(x) \
+# define FAST256(x) \
     if ((*p-- = "0123456789"[(n = x % 10)]) \
 	    && x >= 10 && (*p-- = "0123456789"[((m = x % 100) - n) / 10]) \
 	    && x >= 100 && (*p-- = "012"[((x & 0xff) - m) / 100]));
 
-#define FAST256CASE(x) \
+# define FAST256CASE(x) \
     case x: \
 	FAST256(newargs[x - 1]);
 
@@ -7872,8 +7872,8 @@ vtp_sgr_bulks(
     int argc,
     int *args)
 {
-#define MAXSGR 16
-#define SGRBUFSIZE 2 + 4 * MAXSGR + 1 // '\033[' + SGR + 'm'
+# define MAXSGR 16
+# define SGRBUFSIZE 2 + 4 * MAXSGR + 1 // '\033[' + SGR + 'm'
     char_u  buf[SGRBUFSIZE];
     char_u  *p;
     int	    in, out;
