# HG changeset patch
# Parent  87da22716dd7104149f968ba9a3225a1e82c1840

diff --git a/src/po/Make_all.mak b/src/po/Make_all.mak
--- a/src/po/Make_all.mak
+++ b/src/po/Make_all.mak
@@ -76,7 +76,6 @@ MOFILES = \
 		zh_CN.UTF-8.mo \
 		zh_CN.mo \
 		zh_TW.UTF-8.mo \
-		zh_TW.mo \
 
 
 MOCONVERTED = \
@@ -89,6 +88,7 @@ MOCONVERTED = \
 		sk.cp1250.mo \
 		uk.cp1251.mo \
 		zh_CN.cp936.mo \
+		zh_TW.mo \
 
 
 CHECKFILES = \
diff --git a/src/po/Makefile b/src/po/Makefile
--- a/src/po/Makefile
+++ b/src/po/Makefile
@@ -113,6 +113,17 @@ sk.cp1250.po: sk.po
 	iconv -f iso-8859-2 -t cp1250 sk.po | \
 		sed -e 's/charset=ISO-8859-2/charset=cp1250/' -e 's/# Original translations/# Generated from sk.po, DO NOT EDIT/' > sk.cp1250.po
 
+# Convert zh_TW.UTF-8.po to create zh_TW.po.  Requires doubling backslashes
+# in the second byte.  Don't depend on big5corr, it should only be compiled
+# when zh_TW.po is outdated.
+zh_TW.po: zh_TW.UTF-8.po
+	@$(MAKE) big5corr
+	rm -f zh_TW.po
+	iconv -f utf-8 -t big5 zh_TW.UTF-8.po | ./big5corr > zh_TW.po
+
+big5corr: big5corr.c
+	$(CC) -o big5corr big5corr.c
+
 # Convert zh_CN.UTF-8.po to create zh_CN.cp936.po.
 # Set 'charset' to gbk to avoid that msfmt generates a warning.
 # This used to convert from zh_CN.po, but that results in a conversion error.
diff --git a/src/po/big5corr.c b/src/po/big5corr.c
new file mode 100644
--- /dev/null
+++ b/src/po/big5corr.c
@@ -0,0 +1,43 @@
+/*
+ * Simplistic program to correct Big5 inside strings.  When a trail byte is a
+ * backslash it needs to be doubled.
+ * Public domain.
+ */
+#include <stdio.h>
+#include <string.h>
+
+	int
+main(int argc, char **argv)
+{
+	char buffer[BUFSIZ];
+	char *p;
+
+	while (fgets(buffer, BUFSIZ, stdin) != NULL)
+	{
+		for (p = buffer; *p != 0; p++)
+		{
+			if (strncmp(p, "charset=utf-8", 13) == 0
+				|| strncmp(p, "charset=UTF-8", 13) == 0)
+			{
+				fputs("charset=big5", stdout);
+				p += 12;
+			}
+			else if (strncmp(p, "# Original translations", 23) == 0)
+			{
+				fputs("# generated from zh_TW.UTF-8.po, DO NOT EDIT", stdout);
+				while (p[1] != '\n')
+					++p;
+			}
+			else
+			{
+				if (*(unsigned char *)p >= 0xA1)
+				{
+					putchar(*p++);
+					if (*p == '\\')
+						putchar(*p);
+				}
+				putchar(*p);
+			}
+		}
+	}
+}
