# HG changeset patch
# Parent  d5c084733f01155dcd906da35581a23e8a3651ad

diff --git a/src/gui.c b/src/gui.c
--- a/src/gui.c
+++ b/src/gui.c
@@ -1034,6 +1034,7 @@ gui_set_cursor(int row, int col)
     static void
 gui_check_pos(void)
 {
+    Trace("gui_check_pos: col:%d, row:%d, screen_Rows:%d, screen_Columns:%d", gui.col, gui.row, screen_Rows, screen_Columns);
     if (gui.row >= screen_Rows)
 	gui.row = screen_Rows - 1;
     if (gui.col >= screen_Columns)
@@ -1066,6 +1067,7 @@ gui_update_cursor(
     int		attr;
     attrentry_T *aep = NULL;
 
+    Trace("gui_update_cursor");
     /* Don't update the cursor when halfway busy scrolling or the screen size
      * doesn't match 'columns' and 'lines.  ScreenLines[] isn't valid then. */
     if (!can_update_cursor || screen_Columns != gui.num_cols
@@ -1076,7 +1078,9 @@ gui_update_cursor(
     if (!gui.cursor_is_valid || force
 		    || gui.row != gui.cursor_row || gui.col != gui.cursor_col)
     {
+	Trace("in gui_update_cursor > gui_undraw_cursor: col:%d, row:%d", gui.col, gui.row);
 	gui_undraw_cursor();
+	Trace("in gui_update_cursor < gui_undraw_cursor: col:%d, row:%d", gui.col, gui.row);
 	if (gui.row < 0)
 	    return;
 #ifdef FEAT_MBYTE
@@ -1977,6 +1981,7 @@ gui_write(
     gui.dragged_sb = SBAR_NONE;
 #endif
 
+    Trace("in gui_write > gui_may_flush");
     gui_may_flush();		    /* In case vim decides to take a nap */
 }
 
@@ -1991,8 +1996,10 @@ gui_dont_update_cursor(int undraw)
     if (gui.in_use)
     {
 	/* Undraw the cursor now, we probably can't do it after the change. */
+	Trace("in gui_dont_update_cursor > gui_undraw_cursor");
 	if (undraw)
 	    gui_undraw_cursor();
+	Trace("in gui_dont_update_cursor < gui_undraw_cursor");
 	can_update_cursor = FALSE;
     }
 }
@@ -2625,6 +2632,7 @@ gui_outstr_nowrap(
     void
 gui_undraw_cursor(void)
 {
+    Trace("gui_undraw_cursor");
     if (gui.cursor_is_valid)
     {
 #ifdef FEAT_HANGULIN
@@ -2987,6 +2995,7 @@ gui_wait_for_chars(long wtime)
 	return FAIL;
 
     /* Before waiting, flush any output to the screen. */
+    Trace("in gui_wait_for_chars");
     gui_mch_flush();
 
     if (wtime > 0)
@@ -3699,6 +3708,7 @@ gui_update_tabline(void)
     {
 	/* Updating the tabline uses direct GUI commands, flush
 	 * outstanding instructions first. (esp. clear screen) */
+	Trace("in gui_update_tabline");
 	out_flush();
 
 	if (!showit != !shown)
@@ -4107,6 +4117,7 @@ gui_drag_scrollbar(scrollbar_T *sb, long
 	}
 	if (old_leftcol != curwin->w_leftcol)
 	{
+	    Trace("in gui_drag_scrollbar 1");
 	    updateWindow(curwin);   /* update window, status and cmdline */
 	    setcursor();
 	}
@@ -4131,6 +4142,7 @@ gui_drag_scrollbar(scrollbar_T *sb, long
 #  endif
 			))))
     {
+	Trace("in gui_drag_scrollbar 2");
 	do_check_scrollbind(TRUE);
 	/* need to update the window right here */
 	FOR_ALL_WINDOWS(wp)
@@ -4139,6 +4151,7 @@ gui_drag_scrollbar(scrollbar_T *sb, long
 	setcursor();
     }
 # endif
+    Trace("in gui_drag_scrollbar");
     out_flush_cursor(FALSE, TRUE);
 #else
     add_to_input_buf(bytes, byte_count);
@@ -4502,9 +4515,11 @@ gui_do_scroll(void)
 	 * disappear when losing focus after a scrollbar drag. */
 	if (wp->w_redr_type < type)
 	    wp->w_redr_type = type;
+	Trace("> in gui_do_scroll");
 	gui_disable_flush();
 	updateWindow(wp);   /* update window, status line, and cmdline */
 	gui_enable_flush();
+	Trace("< in gui_do_scroll");
     }
 
 #ifdef FEAT_INS_EXPAND
@@ -4815,6 +4830,7 @@ gui_focus_change(int in_focus)
  */
 #if 1
     gui.in_focus = in_focus;
+    Trace("in gui_focus_change");
     out_flush_cursor(TRUE, FALSE);
 
 # ifdef FEAT_XIM
@@ -5160,6 +5176,7 @@ gui_update_screen(void)
     }
 #endif
 
+    Trace("in gui_update_screen > update_screen");
     update_screen(0);	/* may need to update the screen */
     setcursor();
 # if defined(FEAT_CONCEAL)
@@ -5174,6 +5191,7 @@ gui_update_screen(void)
 	curwin->w_valid &= ~VALID_CROW;
     }
 # endif
+    Trace("in gui_update_screen > out_flush_cursor");
     out_flush_cursor(TRUE, FALSE);
 }
 #endif
diff --git a/src/gui_w32.c b/src/gui_w32.c
--- a/src/gui_w32.c
+++ b/src/gui_w32.c
@@ -6420,6 +6420,7 @@ gui_mch_draw_string(
 	{
 	    pcliprect = &rc;
 	    foptions = ETO_CLIPPED;
+	    Trace("Drawing cursor: col:%d, row:%d", col, row);
 	}
     }
     SetTextColor(s_hdc, gui.currFgColor);
@@ -6655,6 +6656,7 @@ gui_mch_flush(void)
     BOOL  __stdcall GdiFlush(void);
 #   endif
 
+    Trace("** gui_mch_flush");
 #if defined(FEAT_DIRECTX)
     if (IS_ENABLE_DIRECTX())
 	DWriteContext_Flush(s_dwc);
diff --git a/src/main.c b/src/main.c
--- a/src/main.c
+++ b/src/main.c
@@ -1243,6 +1243,7 @@ main_loop(
 		update_curbuf(INVERTED);/* update inverted part */
 	    else if (must_redraw)
 	    {
+		Trace("in main_loop > update_screen");
 #ifdef FEAT_GUI
 		/* Stop issuing gui_mch_flush(). */
 		gui_disable_flush();
diff --git a/src/message.c b/src/message.c
--- a/src/message.c
+++ b/src/message.c
@@ -2319,6 +2319,7 @@ msg_scroll_up(void)
 #ifdef FEAT_GUI
     /* Remove the cursor before scrolling, ScreenLines[] is going
      * to become invalid. */
+    Trace("in msg_scroll_up > gui_undraw_cursor");
     if (gui.in_use)
 	gui_undraw_cursor();
 #endif
diff --git a/src/screen.c b/src/screen.c
--- a/src/screen.c
+++ b/src/screen.c
@@ -374,6 +374,7 @@ redraw_asap(int type)
 #endif
 	}
 
+	Trace("in redraw_asap");
 	update_screen(0);
 	ret = 3;
 
@@ -443,6 +444,7 @@ redraw_asap(int type)
     void
 redraw_after_callback(int call_update_screen)
 {
+    Trace("redraw_after_callback");
     ++redrawing_for_callback;
 
     if (State == HITRETURN || State == ASKMORE)
@@ -522,6 +524,7 @@ redrawWinline(
 update_curbuf(int type)
 {
     redraw_curbuf_later(type);
+    Trace("in update_curbuf");
     update_screen(type);
 }
 
@@ -550,6 +553,7 @@ update_screen(int type_arg)
     if (!screen_valid(TRUE))
 	return FAIL;
 
+    Trace("update_screen: %d", type_arg);
     if (type == VALID_NO_UPDATE)
     {
 	no_update = TRUE;
@@ -745,13 +749,17 @@ update_screen(int type_arg)
 		{
 		    gui_cursor_col = gui.cursor_col;
 		    gui_cursor_row = gui.cursor_row;
+		    Trace("in update_screen > gui_undraw_cursor");
 		    gui_undraw_cursor();
+		    Trace("in update_screen < gui_undraw_cursor");
 		    did_undraw = TRUE;
 		}
 #endif
 	    }
 #endif
+	    Trace("in update_screen > win_update");
 	    win_update(wp);
+	    Trace("in update_screen < win_update");
 	}
 
 	/* redraw status line after the window to minimize cursor movement */
@@ -812,7 +820,9 @@ update_screen(int type_arg)
 	    gui.col = mb_fix_col(gui.col, gui.row);
 # endif
 	    gui_update_cursor(FALSE, FALSE);
+	    Trace("in update_screen > gui_may_flush: col:%d, row:%d", gui.col, gui.row);
 	    gui_may_flush();
+	    Trace("in update_screen < gui_may_flush: col:%d, row:%d", gui.col, gui.row);
 	    screen_cur_col = gui.col;
 	    screen_cur_row = gui.row;
 	}
@@ -837,6 +847,7 @@ update_prepare(void)
 #ifdef FEAT_GUI
     /* Remove the cursor before starting to do anything, because scrolling may
      * make it difficult to redraw the text under it. */
+    Trace("in update_prepare > gui_undraw_cursor");
     if (gui.in_use)
 	gui_undraw_cursor();
 #endif
@@ -867,6 +878,7 @@ update_finish(void)
      * done. */
     if (gui.in_use)
     {
+	Trace("in update_finish > out_flush_cursor");
 	out_flush_cursor(FALSE, FALSE);
 	gui_update_scrollbars(FALSE);
     }
@@ -957,6 +969,7 @@ update_single_line(win_T *wp, linenr_T l
 	    row += wp->w_lines[j].wl_size;
 	}
 
+	Trace("in update_single_line");
 	update_finish();
 
 #ifdef SYN_TIME_LIMIT
@@ -1065,6 +1078,7 @@ updateWindow(win_T *wp)
 	    )
 	win_redr_status(wp);
 
+    Trace("in updateWindow");
     update_finish();
 }
 #endif
@@ -9893,6 +9907,7 @@ screen_ins_lines(
 #ifdef FEAT_GUI
     /* Don't update the GUI cursor here, ScreenLines[] is invalid until the
      * scrolling is actually carried out. */
+    Trace("in screen_ins_lines > gui_dont_update_cursor");
     gui_dont_update_cursor(row + off <= gui.cursor_row);
 #endif
 
@@ -10109,8 +10124,10 @@ screen_del_lines(
 #ifdef FEAT_GUI
     /* Don't update the GUI cursor here, ScreenLines[] is invalid until the
      * scrolling is actually carried out. */
+    Trace("in screen_del_lines > gui_dont_update_cursor");
     gui_dont_update_cursor(gui.cursor_row >= row + off
 						&& gui.cursor_row < end + off);
+    Trace("in screen_del_lines < gui_dont_update_cursor");
 #endif
 
     if (*T_CCS != NUL)	    /* cursor relative to region */
@@ -10230,8 +10247,10 @@ screen_del_lines(
 
 #ifdef FEAT_GUI
     gui_can_update_cursor();
+    Trace("in screen_del_lines > out_flush");
     if (gui.in_use)
 	out_flush();	/* always flush after a scroll */
+    Trace("in screen_del_lines < out_flush");
 #endif
 
     return OK;
diff --git a/src/term.c b/src/term.c
--- a/src/term.c
+++ b/src/term.c
@@ -2521,8 +2521,10 @@ out_flush_cursor(
     gui_enable_flush();
     if (gui.in_use)
     {
+	Trace("in out_flush_cursor >");
 	gui_update_cursor(force, clear_selection);
 	gui_may_flush();
+	Trace("in out_flush_cursor <");
     }
 #endif
 }
