# HG changeset patch
# Parent  3cc0ddf2abd0c4d56eba36b8e70f10b5bdcde45e

diff --git a/src/os_mswin.c b/src/os_mswin.c
--- a/src/os_mswin.c
+++ b/src/os_mswin.c
@@ -936,7 +936,7 @@ DumpPutS(const char *psz UNUSED)
 # endif
 }
 
-#ifdef _DEBUG
+#if 1 //#ifdef _DEBUG
 
 void __cdecl
 Trace(
diff --git a/src/os_win32.h b/src/os_win32.h
--- a/src/os_win32.h
+++ b/src/os_win32.h
@@ -191,6 +191,9 @@ Trace(char *pszFormat, ...);
 # define TRACE3(sz, p1, p2, p3)
 # define TRACE4(sz, p1, p2, p3, p4)
 
+void __cdecl
+Trace(char *pszFormat, ...);
+
 #endif // !_DEBUG
 
 
