# HG changeset patch
# Parent 87f41f5dd56b39e1f23e7cdb4e8cf0a9be72ce55
diff --git a/runtime/doc/tags b/runtime/doc/tags
--- a/runtime/doc/tags
+++ b/runtime/doc/tags
@@ -5394,6 +5394,8 @@
 digraphs-use	digraph.txt	/*digraphs-use*
 dip	motion.txt	/*dip*
 dircolors.vim	syntax.txt	/*dircolors.vim*
+directwrite	options.txt	/*directwrite*
+directx	options.txt	/*directx*
 dis	motion.txt	/*dis*
 disable-menus	gui.txt	/*disable-menus*
 discard	editing.txt	/*discard*
