# HG changeset patch
# Parent  761dae0b59a4e68d74deeda78963d0f7da509e89

diff --git a/src/testdir/Make_all.mak b/src/testdir/Make_all.mak
--- a/src/testdir/Make_all.mak
+++ b/src/testdir/Make_all.mak
@@ -158,7 +158,6 @@ NEW_TESTS = test_arglist.res \
 	    test_hardcopy.res \
 	    test_hide.res \
 	    test_history.res \
-	    test_hlsearch.res \
 	    test_increment.res \
 	    test_increment_dbcs.res \
 	    test_job_fails.res \
