# HG changeset patch
# Parent  5c93ebed3be1b92262bce13debf076883ebf9e2c

diff --git a/src/gui_gtk_f.c b/src/gui_gtk_f.c
--- a/src/gui_gtk_f.c
+++ b/src/gui_gtk_f.c
@@ -97,9 +97,6 @@ static void gtk_form_send_configure(GtkF
 static void gtk_form_child_map(GtkWidget *widget, gpointer user_data);
 static void gtk_form_child_unmap(GtkWidget *widget, gpointer user_data);
 
-#if !GTK_CHECK_VERSION(3,0,0)
-static GtkWidgetClass *parent_class = NULL;
-#endif
 
 /* Public interface
  */
@@ -109,11 +106,7 @@ gtk_form_new(void)
 {
     GtkForm *form;
 
-#if GTK_CHECK_VERSION(3,0,0)
     form = g_object_new(GTK_TYPE_FORM, NULL);
-#else
-    form = gtk_type_new(gtk_form_get_type());
-#endif
 
     return GTK_WIDGET(form);
 }
@@ -212,30 +205,7 @@ gtk_form_thaw(GtkForm *form)
 
 /* Basic Object handling procedures
  */
-#if GTK_CHECK_VERSION(3,0,0)
 G_DEFINE_TYPE(GtkForm, gtk_form, GTK_TYPE_CONTAINER)
-#else
-    GtkType
-gtk_form_get_type(void)
-{
-    static GtkType form_type = 0;
-
-    if (!form_type)
-    {
-	GtkTypeInfo form_info;
-
-	vim_memset(&form_info, 0, sizeof(form_info));
-	form_info.type_name = "GtkForm";
-	form_info.object_size = sizeof(GtkForm);
-	form_info.class_size = sizeof(GtkFormClass);
-	form_info.class_init_func = (GtkClassInitFunc)gtk_form_class_init;
-	form_info.object_init_func = (GtkObjectInitFunc)gtk_form_init;
-
-	form_type = gtk_type_unique(GTK_TYPE_CONTAINER, &form_info);
-    }
-    return form_type;
-}
-#endif /* !GTK_CHECK_VERSION(3,0,0) */
 
     static void
 gtk_form_class_init(GtkFormClass *klass)
@@ -246,10 +216,6 @@ gtk_form_class_init(GtkFormClass *klass)
     widget_class = (GtkWidgetClass *) klass;
     container_class = (GtkContainerClass *) klass;
 
-#if !GTK_CHECK_VERSION(3,0,0)
-    parent_class = gtk_type_class(gtk_container_get_type());
-#endif
-
     widget_class->realize = gtk_form_realize;
     widget_class->unrealize = gtk_form_unrealize;
     widget_class->map = gtk_form_map;
@@ -433,13 +399,8 @@ gtk_form_unrealize(GtkWidget *widget)
 	tmp_list = tmp_list->next;
     }
 
-#if GTK_CHECK_VERSION(3,0,0)
     if (GTK_WIDGET_CLASS (gtk_form_parent_class)->unrealize)
 	 (* GTK_WIDGET_CLASS (gtk_form_parent_class)->unrealize) (widget);
-#else
-    if (GTK_WIDGET_CLASS (parent_class)->unrealize)
-	 (* GTK_WIDGET_CLASS (parent_class)->unrealize) (widget);
-#endif
 }
 
     static void
diff --git a/src/gui_gtk_f.h b/src/gui_gtk_f.h
--- a/src/gui_gtk_f.h
+++ b/src/gui_gtk_f.h
@@ -22,17 +22,10 @@ extern "C" {
 #endif
 
 #define GTK_TYPE_FORM		       (gtk_form_get_type ())
-#ifdef USE_GTK3
 #define GTK_FORM(obj)		       (G_TYPE_CHECK_INSTANCE_CAST((obj), GTK_TYPE_FORM, GtkForm))
 #define GTK_FORM_CLASS(klass)	       (G_TYPE_CHECK_CLASS_CAST((klass), GTK_TYPE_FORM, GtkFormClass))
 #define GTK_IS_FORM(obj)	       (G_TYPE_CHECK_INSTANCE_TYPE((obj), GTK_TYPE_FORM))
 #define GTK_IS_FORM_CLASS(klass)       (G_TYPE_CHECK_CLASS_TYPE((klass), GTK_TYPE_FORM))
-#else
-#define GTK_FORM(obj)		       (GTK_CHECK_CAST ((obj), GTK_TYPE_FORM, GtkForm))
-#define GTK_FORM_CLASS(klass)	       (GTK_CHECK_CLASS_CAST ((klass), GTK_TYPE_FORM, GtkFormClass))
-#define GTK_IS_FORM(obj)	       (GTK_CHECK_TYPE ((obj), GTK_TYPE_FORM))
-#define GTK_IS_FORM_CLASS(klass)       (GTK_CHECK_CLASS_TYPE ((klass), GTK_TYPE_FORM))
-#endif
 
 
 typedef struct _GtkForm GtkForm;
@@ -52,11 +45,7 @@ struct _GtkFormClass
     GtkContainerClass parent_class;
 };
 
-#ifdef USE_GTK3
 GType gtk_form_get_type(void);
-#else
-GtkType gtk_form_get_type(void);
-#endif
 
 GtkWidget *gtk_form_new(void);
 
