# HG changeset patch
# Parent  6227f1f5bd72ed76e6f52bd4719a46a3b2162069

diff --git a/src/gui_dwrite.cpp b/src/gui_dwrite.cpp
--- a/src/gui_dwrite.cpp
+++ b/src/gui_dwrite.cpp
@@ -1034,7 +1034,7 @@ DWriteContext::DrawText(const WCHAR *tex
 
 	TextRenderer renderer(this);
 	TextRendererContext context = { color, FLOAT(cellWidth), 0.0f };
-	textLayout->Draw(&context, &renderer, FLOAT(x), FLOAT(y) - 0.5f);
+	textLayout->Draw(&context, &renderer, FLOAT(x), FLOAT(y) - 1.0f);
     }
 
     SafeRelease(&textLayout);
