# HG changeset patch
# Parent  0e1687fa0e9eea9880c09d614c07f86c87215e7a

diff --git a/src/gui_dwrite.cpp b/src/gui_dwrite.cpp
--- a/src/gui_dwrite.cpp
+++ b/src/gui_dwrite.cpp
@@ -930,6 +930,8 @@ DWriteContext::BindDC(HDC hdc, const REC
     }
 }
 
+extern "C" void redraw_later_clear(void);
+
     HRESULT
 DWriteContext::SetDrawingMode(DrawingMode mode)
 {
@@ -952,6 +954,7 @@ DWriteContext::SetDrawingMode(DrawingMod
 		    hr = S_OK;
 		    DiscardDeviceResources();
 		    CreateDeviceResources();
+		    redraw_later_clear();
 		}
 		mDrawing = false;
 	    }
