# HG changeset patch
# Parent  8d67cafbe1416d14203dc248f99b1b489f9a397a

diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -67,7 +67,7 @@ CSCOPE=yes
 # Set to yes to enable Netbeans support (requires CHANNEL).
 NETBEANS=$(GUI)
 # Set to yes to enable inter process communication.
-CHANNEL=$(GUI)
+CHANNEL=yes
 
 
 # Link against the shared version of libstdc++ by default.  Set
diff --git a/src/Make_mvc.mak b/src/Make_mvc.mak
--- a/src/Make_mvc.mak
+++ b/src/Make_mvc.mak
@@ -101,7 +101,7 @@
 #	Netbeans Debugging Support: NBDEBUG=[yes or no] (should be no, yes
 #	doesn't work)
 #
-#	Inter process communication: CHANNEL=[yes or no] (default is yes if GUI
+#	Inter process communication: CHANNEL=[yes or no] (default is yes)
 #	is yes)
 #
 #	XPM Image Support: XPM=[path to XPM directory]
@@ -295,7 +295,7 @@ NETBEANS = $(GUI)
 !endif
 
 !ifndef CHANNEL
-CHANNEL = $(GUI)
+CHANNEL = yes
 !endif
 
 # Only allow NETBEANS and XPM for a GUI build and CHANNEL.
