# HG changeset patch
# Parent  cc6f57f8bee1f08a2cfb92871fe6a17d6efb1572

diff --git a/src/userfunc.c b/src/userfunc.c
--- a/src/userfunc.c
+++ b/src/userfunc.c
@@ -2326,7 +2326,8 @@ ex_function(exarg_T *eap)
     fp->uf_lines = newlines;
     if ((flags & FC_CLOSURE) != 0)
     {
-	++fp->uf_refcount;
+	if (fudi.fd_dict == NULL)
+	    ++fp->uf_refcount;
 	if (register_closure(fp) == FAIL)
 	    goto erret;
     }
