# HG changeset patch
# Parent ada5696068bebb4548cf4048433390149bf740a5
diff --git a/src/os_win32.c b/src/os_win32.c
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -2855,14 +2855,16 @@
 	    reparseTag = findDataW.dwReserved0;
 	}
     }
-#else
-    hFind = FindFirstFile(fname, &findDataA);
-    if (hFind != INVALID_HANDLE_VALUE)
+    else
+#endif
     {
-	fileFlags = findDataA.dwFileAttributes;
-	reparseTag = findDataA.dwReserved0;
+	hFind = FindFirstFile(fname, &findDataA);
+	if (hFind != INVALID_HANDLE_VALUE)
+	{
+	    fileFlags = findDataA.dwFileAttributes;
+	    reparseTag = findDataA.dwReserved0;
+	}
     }
-#endif
 
     if (hFind != INVALID_HANDLE_VALUE)
 	FindClose(hFind);
