# HG changeset patch
# Parent  7c5a4372118c971fedf6e6482a9fb0a7c87d9192

diff --git a/src/option.h b/src/option.h
--- a/src/option.h
+++ b/src/option.h
@@ -429,7 +429,7 @@ EXTERN char_u	*p_csprg;	/* 'cscopeprg' *
 EXTERN int	p_csre;		/* 'cscoperelative' */
 # ifdef FEAT_QUICKFIX
 EXTERN char_u	*p_csqf;	/* 'cscopequickfix' */
-#  define	CSQF_CMDS   "sgdctefi"
+#  define	CSQF_CMDS   "sgdctefia"
 #  define	CSQF_FLAGS  "+-0"
 # endif
 EXTERN int	p_cst;		/* 'cscopetag' */
