# HG changeset patch
# Parent 0a6a6e985a0a913f594d2ff5d1180630a47864e1

diff --git a/src/testdir/test71.in b/src/testdir/test71.in
--- a/src/testdir/test71.in
+++ b/src/testdir/test71.in
@@ -2,6 +2,9 @@
 
 STARTTEST
 :so small.vim
+:set enc=utf-8
+:bwipe!
+:r ++enc=latin1 test71a.in
 :/^start of text/+1
 :let text_lines = getline('.', line('.') + 2)
 :/^start of cm=zip bytes/+1
@@ -60,16 +63,3 @@
 :qa!
 ENDTEST
 
-start of text
-01234567890123456789012345678901234567
-line 2  foo bar blah
-line 3 xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
-end of text
-
-start of cm=zip bytes
-VimCrypt~01!lV'Þ}Mg ê£V©çE#3Ž2Ué—
-end of cm=zip bytes
-
-start of cm=blowfish bytes
-VimCrypt~02!k)¾—#ÝSœõ=ºàÈ#¥M´†JÃAÍ¥M´†!€›õáÒ‚˜÷Ú
-end of cm=blowfish bytes
diff --git a/src/testdir/test71a.in b/src/testdir/test71a.in
new file mode 100644
--- /dev/null
+++ b/src/testdir/test71a.in
@@ -0,0 +1,13 @@
+start of text
+01234567890123456789012345678901234567
+line 2  foo bar blah
+line 3 xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+end of text
+
+start of cm=zip bytes
+VimCrypt~01!lV'Þ}Mg ê£V©çE#3Ž2Ué—
+end of cm=zip bytes
+
+start of cm=blowfish bytes
+VimCrypt~02!k)¾—#ÝSœõ=ºàÈ#¥M´†JÃAÍ¥M´†!€›õáÒ‚˜÷Ú
+end of cm=blowfish bytes
