# HG changeset patch
# Parent 561e223412ae790fc893722a3ae4c1989567fb73
diff --git a/src/edit.c b/src/edit.c
--- a/src/edit.c
+++ b/src/edit.c
@@ -5184,8 +5184,13 @@
 	}
 	else if (ctrl_x_mode == CTRL_X_FILES)
 	{
-	    while (--startcol >= 0 && vim_isfilec(line[startcol]))
-		;
+	    char_u	*p;
+
+	    p = line + startcol;
+	    mb_ptr_back(line, p);
+	    while (vim_isfilec(PTR2CHAR(p)) && p >= line)
+		mb_ptr_back(line, p);
+	    startcol = p - line;
 	    compl_col += ++startcol;
 	    compl_length = (int)curs_col - startcol;
 	    compl_pattern = addstar(line + compl_col, compl_length,
