# HG changeset patch
# Parent 85f41c5dbc77a06e44418240b8efa73dca58645c
diff --git a/runtime/syntax/help.vim b/runtime/syntax/help.vim
--- a/runtime/syntax/help.vim
+++ b/runtime/syntax/help.vim
@@ -44,7 +44,7 @@
 syn match helpVim		"VIM REFERENCE.*"
 syn match helpOption		"'[a-z]\{2,\}'"
 syn match helpOption		"'t_..'"
-syn match helpCommand		"`[^` ]\+`"hs=s+1,he=e-1 contains=helpBacktick
+syn match helpCommand		"`[^` \t]\+`"hs=s+1,he=e-1 contains=helpBacktick
 syn match helpHeader		"\s*\zs.\{-}\ze\s\=\~$" nextgroup=helpIgnore
 syn match helpGraphic		".* \ze`$" nextgroup=helpIgnore
 if has("conceal")
