# HG changeset patch
# Parent  d833b4e08bed520ee810f11d8147a77bd7f53d6e

diff --git a/src/spell.c b/src/spell.c
--- a/src/spell.c
+++ b/src/spell.c
@@ -1759,6 +1759,11 @@ spell_move_to(
 	}
 	else
 	{
+	    /* If we are back at the starting line and there is no match then
+	     * give up. */
+	    if (lnum == wp->w_cursor.lnum && wrapped)
+		break;
+
 	    if (lnum < wp->w_buffer->b_ml.ml_line_count)
 		++lnum;
 	    else if (!p_ws)
@@ -1775,7 +1780,7 @@ spell_move_to(
 
 	    /* If we are back at the starting line and there is no match then
 	     * give up. */
-	    if (lnum == wp->w_cursor.lnum && (!found_one || wrapped))
+	    if (lnum == wp->w_cursor.lnum && !found_one)
 		break;
 
 	    /* Skip the characters at the start of the next line that were
