# HG changeset patch
# Parent  653dd0fc53a21312c3d53fd2e6f3f702d5cc47f3

diff --git a/runtime/doc/Makefile b/runtime/doc/Makefile
--- a/runtime/doc/Makefile
+++ b/runtime/doc/Makefile
@@ -343,7 +343,7 @@ xxd.man: xxd.1
 	nroff -man xxd.1 | sed -e s/.//g > xxd.man
 
 uganda.nsis.txt: uganda.txt
-	sed -e 's/[ 	]*\*[-a-zA-Z0-9.]*\*//g' -e 's/vim:tw=78://' \
+	sed -e 's/[ 	]*\*[-a-zA-Z0-9.]*\*//g' -e 's/vim:tw=78:.*//' \
 		uganda.txt | uniq >uganda.nsis.txt
 
 # Awk version of .txt to .html conversion.
