# HG changeset patch
# Parent c8f754c4bd7340fff6944060978822d3f8fa548b

diff --git a/src/Makefile b/src/Makefile
--- a/src/Makefile
+++ b/src/Makefile
@@ -1668,20 +1668,24 @@
 	    ! grep -x -F 'ac_cv_env_CFLAGS_value=$(CFLAGS)' auto/config.cache > /dev/null; then \
 		rm auto/config.cache; \
 	fi
-	GUI_INC_LOC="$(GUI_INC_LOC)" GUI_LIB_LOC="$(GUI_LIB_LOC)" \
-		CC="$(CC)" CPPFLAGS="$(CPPFLAGS)" CFLAGS="$(CFLAGS)" \
-		LDFLAGS="$(LDFLAGS)" $(CONF_SHELL) srcdir="$(srcdir)" \
-		./configure $(CONF_OPT_GUI) $(CONF_OPT_X) $(CONF_OPT_XSMP) \
-		$(CONF_OPT_DARWIN) $(CONF_OPT_FAIL) \
-		$(CONF_OPT_PERL) $(CONF_OPT_PYTHON) $(CONF_OPT_PYTHON3) \
-		$(CONF_OPT_TCL) $(CONF_OPT_RUBY) $(CONF_OPT_NLS) \
-		$(CONF_OPT_CSCOPE) $(CONF_OPT_MULTIBYTE) $(CONF_OPT_INPUT) \
-		$(CONF_OPT_OUTPUT) $(CONF_OPT_GPM) $(CONF_OPT_WORKSHOP) \
-		$(CONF_OPT_SNIFF) $(CONF_OPT_FEAT) $(CONF_TERM_LIB) \
-		$(CONF_OPT_COMPBY) $(CONF_OPT_ACL)  $(CONF_OPT_NETBEANS) \
-		$(CONF_ARGS) $(CONF_OPT_MZSCHEME) $(CONF_OPT_PLTHOME) \
-		$(CONF_OPT_LUA) $(CONF_OPT_LUA_PREFIX) \
-		$(CONF_OPT_SYSMOUSE)
+	if test "X$(MAKECMDGOALS)" != "Xclean" -a "X$(MAKECMDGOALS)" != "Xdistclean"; then \
+		GUI_INC_LOC="$(GUI_INC_LOC)" GUI_LIB_LOC="$(GUI_LIB_LOC)" \
+			CC="$(CC)" CPPFLAGS="$(CPPFLAGS)" CFLAGS="$(CFLAGS)" \
+			LDFLAGS="$(LDFLAGS)" $(CONF_SHELL) srcdir="$(srcdir)" \
+			./configure $(CONF_OPT_GUI) $(CONF_OPT_X) \
+			$(CONF_OPT_XSMP) $(CONF_OPT_DARWIN) $(CONF_OPT_FAIL) \
+			$(CONF_OPT_PERL) $(CONF_OPT_PYTHON) \
+			$(CONF_OPT_PYTHON3) $(CONF_OPT_TCL) $(CONF_OPT_RUBY) \
+			$(CONF_OPT_NLS) $(CONF_OPT_CSCOPE) \
+			$(CONF_OPT_MULTIBYTE) $(CONF_OPT_INPUT) \
+			$(CONF_OPT_OUTPUT) $(CONF_OPT_GPM) \
+			$(CONF_OPT_WORKSHOP) $(CONF_OPT_SNIFF) \
+			$(CONF_OPT_FEAT) $(CONF_TERM_LIB) $(CONF_OPT_COMPBY) \
+			$(CONF_OPT_ACL) $(CONF_OPT_NETBEANS) \
+			$(CONF_ARGS) $(CONF_OPT_MZSCHEME) $(CONF_OPT_PLTHOME) \
+			$(CONF_OPT_LUA) $(CONF_OPT_LUA_PREFIX) \
+			$(CONF_OPT_SYSMOUSE); \
+	fi
 
 # Use "make reconfig" to rerun configure without cached values.
 # When config.h changes, most things will be recompiled automatically.
