# HG changeset patch
# Parent  74df6fba07cda32890348c0c8ffe95deb7a34f62

diff --git a/src/testdir/test_highlight.vim b/src/testdir/test_highlight.vim
--- a/src/testdir/test_highlight.vim
+++ b/src/testdir/test_highlight.vim
@@ -519,7 +519,7 @@ func Test_termguicolors()
   if !exists('+termguicolors')
     return
   endif
-  if has('vtp') && !has('vcon')
+  if has('vtp') && !has('vcon') && !has('gui_running')
     " Win32: 'guicolors' doesn't work without virtual console.
     call assert_fails('set termguicolors', 'E954:')
     return
