# HG changeset patch
# Parent  4c4904e57e3ff889ebb0972de138cc6df254c657

diff --git a/src/testdir/Make_dos.mak b/src/testdir/Make_dos.mak
--- a/src/testdir/Make_dos.mak
+++ b/src/testdir/Make_dos.mak
@@ -27,9 +27,6 @@ DOSTMP_INFILES = $(DOSTMP_OUTFILES:.out=
 
 .SUFFIXES: .in .out .res .vim
 
-# Must run test1 first to create small.vim.
-$(SCRIPTS) $(SCRIPTS_GUI) $(SCRIPTS_WIN32) $(NEW_TESTS): $(SCRIPTS_FIRST)
-
 nongui:	nolog $(SCRIPTS_FIRST) $(SCRIPTS) newtests report
 
 small:	nolog report
@@ -67,6 +64,9 @@ win32:	nolog $(SCRIPTS_FIRST) $(SCRIPTS)
 		 & echo $* FAILED >> test.log ) \
 		else ( move /y test.out $*.out )
 
+# Must run test1 first to create small.vim.
+$(SCRIPTS) $(SCRIPTS_GUI) $(SCRIPTS_WIN32) $(NEW_TESTS): $(SCRIPTS_FIRST)
+
 report:
 	@echo ""
 	@echo Test results:
