# HG changeset patch
# Parent d445f5c032c6f7e8a33eaef8b278f896531de347
diff --git a/src/if_python3.c b/src/if_python3.c
--- a/src/if_python3.c
+++ b/src/if_python3.c
@@ -734,6 +734,8 @@
 	 * as the system trace hook), will be lost between invocations of
 	 * Python code. */
 	PyEval_InitThreads();
+	/* The first python thread is vim's, release the lock. */
+	PyGILState_Release(pygilstate);
 	pygilstate = PyGILState_Ensure();
 
 #ifdef DYNAMIC_PYTHON3
