# HG changeset patch
# Parent  7e96225f48f463466c0a1a7683cc5be29afe8b74

diff --git a/src/drawline.c b/src/drawline.c
--- a/src/drawline.c
+++ b/src/drawline.c
@@ -1945,11 +1945,11 @@ win_line(
 		    c_final = NUL;
 		    if (VIM_ISWHITE(c))
 		    {
-#ifdef FEAT_CONCEAL
+# ifdef FEAT_CONCEAL
 			if (c == TAB)
 			    // See "Tab alignment" below.
 			    FIX_FOR_BOGUSCOLS;
-#endif
+# endif
 			if (!wp->w_p_list)
 			    c = ' ';
 		    }
