# HG changeset patch
# Parent  12b1d55b42f1956fb88273d0cb7f6b2d79c96644

diff --git a/src/if_py_both.h b/src/if_py_both.h
--- a/src/if_py_both.h
+++ b/src/if_py_both.h
@@ -6231,6 +6231,11 @@ ConvertFromPyObject(PyObject *obj, typva
 
 	Py_DECREF(num);
     }
+    else if (obj == Py_None)
+    {
+	tv->v_type = VAR_SPECIAL;
+	tv->vval.v_number = VVAL_NULL;
+    }
     else
     {
 	PyErr_FORMAT(PyExc_TypeError,
