# HG changeset patch
# Parent  ed9d2f4103ec083d72c42773f1589a6a964914e5

diff --git a/src/if_ruby.c b/src/if_ruby.c
--- a/src/if_ruby.c
+++ b/src/if_ruby.c
@@ -199,6 +199,12 @@ static void ruby_vim_init(void);
 # endif
 #endif
 
+#if defined(RUBY19_OR_LATER) && !defined(DYNAMIC_RUBY)
+# ifdef WIN3264
+#  define NtInitialize			ruby_sysinit
+# endif
+#endif
+
 #if defined(DYNAMIC_RUBY) || defined(PROTO)
 # if defined(PROTO) && !defined(HINSTANCE)
 #  define HINSTANCE int		/* for generating prototypes */
@@ -865,11 +871,7 @@ static int ensure_ruby_initialized(void)
 	    int argc = 1;
 	    char *argv[] = {"gvim.exe"};
 	    char **argvp = argv;
-# ifdef RUBY19_OR_LATER
-	    ruby_sysinit(&argc, &argvp);
-# else
 	    NtInitialize(&argc, &argvp);
-# endif
 #endif
 	    {
 #if defined(RUBY19_OR_LATER) || defined(RUBY_INIT_STACK)
