# HG changeset patch
# Parent  a5da25ca6c52cf326e6dd7fa13892029497c00e0

diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -896,7 +896,7 @@ if_perl.c: if_perl.xs typemap
 	     $(PERLTYPEMAP) if_perl.xs -output $@
 
 $(OUTDIR)/iscygpty.o:	iscygpty.c $(CUI_INCL)
-	$(CC) -c $(CFLAGS) iscygpty.c -o $(OUTDIR)/iscygpty.o -D_WIN32_WINNT=0x0600 -DUSE_DYNFILEID -DENABLE_STUB_IMPL
+	$(CC) -c $(CFLAGS) iscygpty.c -o $(OUTDIR)/iscygpty.o -U_WIN32_WINNT -D_WIN32_WINNT=0x0600 -DUSE_DYNFILEID -DENABLE_STUB_IMPL
 
 $(OUTDIR)/main.o:		main.c $(INCL) $(CUI_INCL)
 	$(CC) -c $(CFLAGS) main.c -o $(OUTDIR)/main.o
