# HG changeset patch
# Parent 4bda13dff6a7642be1bf88c6032277211ee2b727

diff --git a/src/VisVim/OleAut.cpp b/src/VisVim/OleAut.cpp
--- a/src/VisVim/OleAut.cpp
+++ b/src/VisVim/OleAut.cpp
@@ -654,7 +654,7 @@
 	}
 
 	char Buf[256];
-	sprintf (Buf, "An OLE error occured:\r\nCode = %s\r\nResult = %lx.",
+	sprintf (Buf, "An OLE error occurred:\r\nCode = %s\r\nResult = %lx.",
 		 (char*) ErrName, m_hResult);
 	MessageBox (NULL, Buf, "OLE Error", MB_OK);
 }
diff --git a/src/configure.in b/src/configure.in
--- a/src/configure.in
+++ b/src/configure.in
@@ -2993,7 +2993,7 @@
 if test "x$vim_cv_tty_group" != "xworld" ; then
   AC_DEFINE_UNQUOTED(PTYGROUP,$vim_cv_tty_group)
   if test "x$vim_cv_tty_mode" = "x" ; then
-    AC_MSG_ERROR([It seems you're cross compiling and have 'vim_cv_tty_group' set, please also set the environment variable 'vim_cv_tty_mode' to the correct mode (propably 0620)])
+    AC_MSG_ERROR([It seems you're cross compiling and have 'vim_cv_tty_group' set, please also set the environment variable 'vim_cv_tty_mode' to the correct mode (probably 0620)])
   else
     AC_DEFINE(PTYMODE, 0620)
   fi
diff --git a/vimtutor.com b/vimtutor.com
--- a/vimtutor.com
+++ b/vimtutor.com
@@ -30,7 +30,7 @@
 $ !
 $	if f$search ("vim:vim.exe") .eqs. ""
 $	then
-$	    write sys$error "Error - Can't run tutoral. VIM not found."
+$	    write sys$error "Error - Can't run tutorial. VIM not found."
 $	    exit
 $	endif
 $ !
