# HG changeset patch
# Parent  d6495e9ff13f5c3fd7bb124015758bad027cb59a

diff --git a/src/option.c b/src/option.c
--- a/src/option.c
+++ b/src/option.c
@@ -8314,9 +8314,9 @@ set_bool_option(opt_idx, varp, value, op
     if (!starting)
     {
 	char_u buf_old[2], buf_new[2], buf_type[7];
-	snprintf((char *)buf_old, 2, "%d", old_value ? TRUE: FALSE);
-	snprintf((char *)buf_new, 2, "%d", value ? TRUE: FALSE);
-	sprintf((char *)buf_type, "%s", (opt_flags & OPT_LOCAL) ? "local" : "global");
+	vim_snprintf((char *)buf_old, 2, "%d", old_value ? TRUE: FALSE);
+	vim_snprintf((char *)buf_new, 2, "%d", value ? TRUE: FALSE);
+	vim_snprintf((char *)buf_type, 7, "%s", (opt_flags & OPT_LOCAL) ? "local" : "global");
 	set_vim_var_string(VV_OPTION_NEW, buf_new, -1);
 	set_vim_var_string(VV_OPTION_OLD, buf_old, -1);
 	set_vim_var_string(VV_OPTION_TYPE, buf_type, -1);
@@ -8864,9 +8864,9 @@ set_num_option(opt_idx, varp, value, err
     if (!starting && errmsg == NULL)
     {
 	char_u buf_old[11], buf_new[11], buf_type[7];
-	snprintf((char *)buf_old, 10, "%ld", old_value);
-	snprintf((char *)buf_new, 10, "%ld", value);
-	snprintf((char *)buf_type, 7, "%s", (opt_flags & OPT_LOCAL) ? "local" : "global");
+	vim_snprintf((char *)buf_old, 10, "%ld", old_value);
+	vim_snprintf((char *)buf_new, 10, "%ld", value);
+	vim_snprintf((char *)buf_type, 7, "%s", (opt_flags & OPT_LOCAL) ? "local" : "global");
 	set_vim_var_string(VV_OPTION_NEW, buf_new, -1);
 	set_vim_var_string(VV_OPTION_OLD, buf_old, -1);
 	set_vim_var_string(VV_OPTION_TYPE, buf_type, -1);
