# HG changeset patch
# Parent  5024da352a108a928f1f27bbe5a1a6ff04f03495

diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -1060,7 +1060,7 @@ auto/if_perl.c:		if_perl.xs typemap
 	     $(PERLTYPEMAP) if_perl.xs -output $@
 
 $(OUTDIR)/if_perl.o:	auto/if_perl.c $(INCL)
-	$(CC) -c $(CFLAGS) auto/if_perl.c -o $(OUTDIR)/if_perl.o
+	$(CC) -c $(CFLAGS) -I. auto/if_perl.c -o $(OUTDIR)/if_perl.o
 
 
 $(OUTDIR)/if_ruby.o:	if_ruby.c $(INCL)
