# HG changeset patch
# Parent  98a1914ed3e518259a84449287b352d86cddcb2d

diff --git a/src/os_win32.c b/src/os_win32.c
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -4418,7 +4418,10 @@ mch_call_shell_terminal(
     argvar[1].v_type = VAR_UNKNOWN;
     buf = term_start(argvar, NULL, &opt, TERM_START_SYSTEM);
     if (buf == NULL)
+    {
+	vim_free(newcmd);
 	return 255;
+    }
 
     job = term_getjob(buf->b_term);
     ++job->jv_refcount;
