# HG changeset patch
# Parent  464dbaf76e96768cf80422af5caf87e18b9efe29

diff --git a/src/GvimExt/Makefile b/src/GvimExt/Makefile
--- a/src/GvimExt/Makefile
+++ b/src/GvimExt/Makefile
@@ -16,6 +16,9 @@ NODEBUG = 1
 # On Windows NT
 ! ifndef CPU
 CPU = i386
+!  if !defined(PLATFORM) && defined(TARGET_CPU)
+PLATFORM = $(TARGET_CPU)
+!  endif
 !  ifdef PLATFORM
 !   if ("$(PLATFORM)" == "x64") || ("$(PLATFORM)" == "X64")
 CPU = AMD64
diff --git a/src/Make_mvc.mak b/src/Make_mvc.mak
--- a/src/Make_mvc.mak
+++ b/src/Make_mvc.mak
@@ -216,6 +216,9 @@ CPU = i386
 !  endif
 ! else  # !CPU
 CPU = i386
+!  if !defined(PLATFORM) && defined(TARGET_CPU)
+PLATFORM = $(TARGET_CPU)
+!  endif
 !  ifdef PLATFORM
 !   if ("$(PLATFORM)" == "x64") || ("$(PLATFORM)" == "X64")
 CPU = AMD64
