# HG changeset patch
# Parent 1f5a633fc3a946204795875eb3ab40bcc43c3908
diff --git a/src/gui_w32.c b/src/gui_w32.c
--- a/src/gui_w32.c
+++ b/src/gui_w32.c
@@ -1688,7 +1688,7 @@ gui_mch_init(void)
 	HANDLE	hIcon = NULL;
 
 	if (mch_icon_load(&hIcon) == OK && hIcon != NULL)
-	    SendMessage(s_hwnd, WM_SETICON, ICON_SMALL, hIcon);
+	    SendMessage(s_hwnd, WM_SETICON, ICON_SMALL, (LPARAM)hIcon);
     }
 
 #ifdef FEAT_MENU
