# HG changeset patch
# Parent b16ee1950b8e8ffccd8e1f7a87f5fbba6965d599
diff --git a/src/if_perl.xs b/src/if_perl.xs
--- a/src/if_perl.xs
+++ b/src/if_perl.xs
@@ -27,7 +27,9 @@
 /* Work around for perl-5.18.
  * Don't include "perl\lib\CORE\inline.h" for now,
  * include it after Perl_sv_free2 is defined. */
-#define PERL_NO_INLINE_FUNCTIONS
+#ifdef DYNAMIC_PERL
+# define PERL_NO_INLINE_FUNCTIONS
+#endif
 
 /*
  * Prevent including winsock.h.  perl.h tries to detect whether winsock.h is
