# HG changeset patch
# Parent  3763198b8993790c5c3a39f414040f37b6872270

diff --git a/src/testdir/test_vartabs.vim b/src/testdir/test_vartabs.vim
--- a/src/testdir/test_vartabs.vim
+++ b/src/testdir/test_vartabs.vim
@@ -445,8 +445,8 @@ endfunc
 func Test_vartabstop_latin1()
   let save_encoding = &encoding
   new
-  set encoding=iso8859
-  silent norm :se 
+  set encoding=latin1
+  silent norm :se 
   set vartabstop=400
   norm i00	
   bwipe!
