# HG changeset patch
# Parent 856415bcfae44c46a00c4bd919a560a6fd59a26a
diff --git a/src/if_perl.xs b/src/if_perl.xs
--- a/src/if_perl.xs
+++ b/src/if_perl.xs
@@ -54,6 +54,12 @@
 #include <XSUB.h>
 
 
+/* Work around for clang on Cygwin */
+#if defined(__CYGWIN__) && defined(__clang__)
+# undef XS
+# define XS(name) XSPROTO(name)
+#endif
+
 /*
  * Work around clashes between Perl and Vim namespace.	proto.h doesn't
  * include if_perl.pro and perlsfio.pro when IN_PERL_FILE is defined, because
