# HG changeset patch
# Parent  0c03afe40d9f3a3a82317419cafb665146aa5e33

diff --git a/src/INSTALL b/src/INSTALL
--- a/src/INSTALL
+++ b/src/INSTALL
@@ -11,10 +11,10 @@ 4. Atari MiNT
 
 See INSTALLami.txt              for Amiga
 See INSTALLmac.txt              for Macintosh
-See INSTALLpc.txt               for PC (MS-DOS, Windows 95/98/NT/XP)
+See INSTALLpc.txt               for PC (Windows 95/98/NT/XP/Vista/7/8/10)
 See INSTALLvms.txt              for VMS
 See INSTALLx.txt		for cross-compiling on Unix
-See ../README_390.txt           for OS/390 Unix
+See ../READMEdir/README_390.txt for OS/390 Unix
 See ../runtime/doc/os_beos.txt  for BeBox
 
 
