# HG changeset patch
# Parent adda2c7d4ac29f9a3e89420340c74c0407f028b4
diff --git a/src/if_perl.xs b/src/if_perl.xs
--- a/src/if_perl.xs
+++ b/src/if_perl.xs
@@ -24,13 +24,6 @@
 # define _USE_32BIT_TIME_T
 #endif
 
-/* Work around for perl-5.18.
- * Don't include "perl\lib\CORE\inline.h" for now,
- * include it after Perl_sv_free2 is defined. */
-#ifdef DYNAMIC_PERL
-# define PERL_NO_INLINE_FUNCTIONS
-#endif
-
 /*
  * Prevent including winsock.h.  perl.h tries to detect whether winsock.h is
  * already included before including winsock2.h, because winsock2.h isn't
@@ -49,6 +42,13 @@
 
 #include "vim.h"
 
+/* Work around for perl-5.18.
+ * Don't include "perl\lib\CORE\inline.h" for now,
+ * include it after Perl_sv_free2 is defined. */
+#ifdef DYNAMIC_PERL
+# define PERL_NO_INLINE_FUNCTIONS
+#endif
+
 #include <EXTERN.h>
 #include <perl.h>
 #include <XSUB.h>
