# HG changeset patch
# Parent 139058cadda1f59cf11c44a5824d69d192b7e154
diff --git a/src/VisVim/OleAut.cpp b/src/VisVim/OleAut.cpp
--- a/src/VisVim/OleAut.cpp
+++ b/src/VisVim/OleAut.cpp
@@ -654,7 +654,7 @@
 	}
 
 	char Buf[256];
-	sprintf (Buf, "An OLE error occured:\r\nCode = %s\r\nResult = %lx.",
+	sprintf (Buf, "An OLE error occurred:\r\nCode = %s\r\nResult = %lx.",
 		 (char*) ErrName, m_hResult);
 	MessageBox (NULL, Buf, "OLE Error", MB_OK);
 }
