# HG changeset patch
# Parent d5b125cc1185e1ac174d4657261532c56461de47
diff --git a/src/testdir/test86.in b/src/testdir/test86.in
--- a/src/testdir/test86.in
+++ b/src/testdir/test86.in
@@ -1384,6 +1384,7 @@ EOF
 :debug silent! py test_keyboard_interrupt()
 :redir END
 :0 debuggreedy
+:call inputrestore()
 :silent $put =output
 :unlet output
 :py del test_keyboard_interrupt
diff --git a/src/testdir/test87.in b/src/testdir/test87.in
--- a/src/testdir/test87.in
+++ b/src/testdir/test87.in
@@ -1361,6 +1361,7 @@ EOF
 :debug silent! py3 test_keyboard_interrupt()
 :redir END
 :0 debuggreedy
+:call inputrestore()
 :silent $put =output
 :unlet output
 :py3 del test_keyboard_interrupt
