# HG changeset patch
# Parent  f742f20758a69dcc94ce532e2e3ba2d231507d72

diff --git a/runtime/autoload/netrw.vim b/runtime/autoload/netrw.vim
--- a/runtime/autoload/netrw.vim
+++ b/runtime/autoload/netrw.vim
@@ -11089,12 +11089,14 @@ fun! s:LocalListing()
 
    if w:netrw_liststyle == s:LONGLIST
     let sz   = getfsize(filename)
-    let fsz  = strpart("               ",1,15-strlen(sz)).sz
     if g:netrw_sizestyle =~# "[hH]"
      let sz= s:NetrwHumanReadable(sz)
     endif
-    let longfile= printf("%-".(g:netrw_maxfilenamelen+1)."s",pfile)
-    let pfile   = longfile.sz." ".strftime(g:netrw_timefmt,getftime(filename))
+    let longfile= printf("%-".g:netrw_maxfilenamelen."S",pfile)
+    let szlen = 15 - (strdisplaywidth(longfile) - g:netrw_maxfilenamelen)
+    let szlen = (szlen > 0) ? szlen : 0
+    let fsz  = printf("%".szlen."S",sz)
+    let pfile   = longfile."  ".fsz." ".strftime(g:netrw_timefmt,getftime(filename))
 "    call Decho("longlist support: sz=".sz." fsz=".fsz,'~'.expand("<slnum>"))
    endif
 
