# HG changeset patch
# Parent  373ee3ae03c152c5a41c0142cd27072e8c8a3adb

diff --git a/src/testdir/test_pyx2.vim b/src/testdir/test_pyx2.vim
--- a/src/testdir/test_pyx2.vim
+++ b/src/testdir/test_pyx2.vim
@@ -28,7 +28,7 @@ endfunc
 func Test_pyxdo()
   pyx import sys
   enew
-  pyxdo return sys.version
+  pyxdo return sys.version.split("\n")[0]
   call assert_match(s:py2pattern, split(getline('.'))[0])
 endfunc
 
diff --git a/src/testdir/test_pyx3.vim b/src/testdir/test_pyx3.vim
--- a/src/testdir/test_pyx3.vim
+++ b/src/testdir/test_pyx3.vim
@@ -28,7 +28,7 @@ endfunc
 func Test_pyxdo()
   pyx import sys
   enew
-  pyxdo return sys.version
+  pyxdo return sys.version.split("\n")[0]
   call assert_match(s:py3pattern, split(getline('.'))[0])
 endfunc
 
