diff --git a/src/testdir/test95.in b/src/testdir/test95.in
--- a/src/testdir/test95.in
+++ b/src/testdir/test95.in
@@ -7,7 +7,7 @@
 STARTTEST
 :so small.vim
 :so mbyte.vim
-:set nocp encoding=utf-8 viminfo+=nviminfo
+:set nocp encoding=utf-8 viminfo+=nviminfo nomore
 :" tl is a List of Lists with:
 :"    regexp pattern
 :"    text to test the pattern on
@@ -30,10 +30,10 @@
 :call add(tl, ['\p\+', 'ìa', 'ìa'])
 
 :"""" Test recognition of some character classes
-:call add(tl, ['\i\+', '&*§xx ', 'xx'])
-:call add(tl, ['\%#=1\i\+', '&*§xx ', 'xx'])
+:call add(tl, ['\i\+', '&*¨xx ', 'xx'])
+:call add(tl, ['\%#=1\i\+', '&*¨xx ', 'xx'])
 :call add(tl, ['\f\+', '&*fname ', 'fname'])
-:call add(tl, ['\%#=1\i\+', '&*fname ', 'fname'])
+:call add(tl, ['\%#=1\f\+', '&*fname ', 'fname'])
 
 :"""" Combining different tests and features
 :call add(tl, ['[^[=a=]]\+', 'ddaãâbcd', 'dd'])
diff --git a/src/testdir/test95.ok b/src/testdir/test95.ok
--- a/src/testdir/test95.ok
+++ b/src/testdir/test95.ok
@@ -8,5 +8,5 @@
 OK - \i\+
 OK - \%#=1\i\+
 OK - \f\+
-OK - \%#=1\i\+
+OK - \%#=1\f\+
 OK - [^[=a=]]\+
