# HG changeset patch
# Parent 9f9587558f5be0397eec9319c490597398af9fbc

diff --git a/src/screen.c b/src/screen.c
--- a/src/screen.c
+++ b/src/screen.c
@@ -5839,7 +5839,7 @@
 		hl = ScreenAttrs[off_to];
 		if (hl > HL_ALL)
 		    hl = syn_attr2attr(hl);
-		if (hl & HL_BOLD)
+		if (hl & (HL_BOLD | HL_ITALIC))
 		    redraw_next = TRUE;
 	    }
 #endif
@@ -5930,7 +5930,7 @@
 	    if (gui.in_use && (col > startCol || !redraw_this))
 	    {
 		hl = ScreenAttrs[off_to];
-		if (hl > HL_ALL || (hl & HL_BOLD))
+		if (hl > HL_ALL || (hl & (HL_BOLD | HL_ITALIC)))
 		{
 		    int prev_cells = 1;
 # ifdef FEAT_MBYTE
@@ -7160,7 +7160,8 @@
 	    /* The bold trick makes a single row of pixels appear in the next
 	     * character.  When a bold character is removed, the next
 	     * character should be redrawn too.  This happens for our own GUI
-	     * and for some xterms. */
+	     * and for some xterms.  The next character to an italic character
+	     * should be redrawn too. */
 	    if (need_redraw && ScreenLines[off] != ' ' && (
 # ifdef FEAT_GUI
 		    gui.in_use
@@ -7177,7 +7178,7 @@
 
 		if (n > HL_ALL)
 		    n = syn_attr2attr(n);
-		if (n & HL_BOLD)
+		if (n & (HL_BOLD | HL_ITALIC))
 		    force_redraw_next = TRUE;
 	    }
 #endif
@@ -8083,7 +8084,7 @@
 		{
 		    if (ScreenLines[off] != ' '
 			    && (ScreenAttrs[off] > HL_ALL
-				|| ScreenAttrs[off] & HL_BOLD))
+				|| ScreenAttrs[off] & (HL_BOLD | HL_ITALIC)))
 			force_next = TRUE;
 		    else
 			force_next = FALSE;
