# HG changeset patch
# Parent  83dd41cc6039fab4168a7aadf972315b23bc220e

diff --git a/src/os_win32.c b/src/os_win32.c
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -4837,20 +4837,17 @@ mch_call_shell(
 			    (msg_silent != 0 || (options & SHELL_DOOUT))
 								 ? "-s " : "",
 			    p_sh, p_shcf, cmd);
-		else
+		else if (
 # ifdef VIMDLL
-		if (gui.in_use || gui.starting)
+			(gui.in_use || gui.starting) &&
 # endif
+			STRCMP(p_shcf, "/c") == 0)
 		    vim_snprintf((char *)newcmd, cmdlen, "%s %s %s %s %s",
 					   p_sh, p_shcf, p_sh, p_shcf, cmd);
-# ifdef VIMDLL
 		else
-# endif
 #endif
-#if !defined(FEAT_GUI_MSWIN) || defined(VIMDLL)
 		    vim_snprintf((char *)newcmd, cmdlen, "%s %s %s",
 							   p_sh, p_shcf, cmd);
-#endif
 		x = mch_system((char *)newcmd, options);
 		vim_free(newcmd);
 	    }
