# HG changeset patch
# Parent  352df9a140db374d767e95b0737c82f7a20d087f

diff --git a/src/testdir/test_recover.vim b/src/testdir/test_recover.vim
--- a/src/testdir/test_recover.vim
+++ b/src/testdir/test_recover.vim
@@ -7,7 +7,7 @@ func Test_recover_root_dir()
   call assert_fails('recover', 'E305:')
   close!
 
-  if has('win32')
+  if has('win32') || filewritable('/') == 2
     " can write in / directory on MS-Windows
     set dir=/notexist/
   endif
