# HG changeset patch
# Parent  fc8d6d66ae6a97393eea7eda681bd985749f6256

diff --git a/src/Makefile b/src/Makefile
--- a/src/Makefile
+++ b/src/Makefile
@@ -2129,7 +2129,8 @@ installrtbase: $(HELPSOURCE)/vim.1 $(DES
 # different from the repository.
 	cd $(HELPSOURCE); if test -f tags; then mv -f tags tags.dist; fi
 	@echo generating help tags
-	-@cd $(HELPSOURCE); $(MAKE) VIMEXE=$(DEST_BIN)/$(VIMTARGET) vimtags
+	-@cd $(HELPSOURCE); $(MAKE) VIMEXE=$(DEST_BIN)/$(VIMTARGET) vimtags || \
+		if test ! -f tags; then cp tags.dist tags; fi
 	cd $(HELPSOURCE); \
 		files=`ls *.txt tags`; \
 		files="$$files `ls *.??x tags-?? 2>/dev/null || true`"; \
