# HG changeset patch
# Parent  55947fb7b189fd4d61c26ca7a86cbf3305bf5852

diff --git a/src/terminal.c b/src/terminal.c
--- a/src/terminal.c
+++ b/src/terminal.c
@@ -339,6 +339,7 @@ term_start(char_u *cmd, jobopt_T *opt)
     {
 	/* store the size we ended up with */
 	vterm_get_size(term->tl_vterm, &term->tl_rows, &term->tl_cols);
+	term_report_winsize(term, term->tl_rows, term->tl_cols);
     }
     else
     {
