# HG changeset patch
# Parent  2ce9557a80f7855a5eef55790f1dda98d50a7e39

diff --git a/src/os_mswin.c b/src/os_mswin.c
--- a/src/os_mswin.c
+++ b/src/os_mswin.c
@@ -2920,7 +2920,7 @@ get_logfont(
      */
     for (p = name; *p && *p != ':'; p++)
     {
-	if (p - name + 1 > LF_FACESIZE)
+	if (p - name + 1 >= LF_FACESIZE)
 	    goto theend;			/* Name too long */
 	lf->lfFaceName[p - name] = *p;
     }
