# HG changeset patch
# Parent  237fbabaa9ffac7c4b71dfb5bc9a69e633de3438

diff --git a/src/testdir/Make_dos.mak b/src/testdir/Make_dos.mak
--- a/src/testdir/Make_dos.mak
+++ b/src/testdir/Make_dos.mak
@@ -20,16 +20,13 @@ default: nongui
 
 SCRIPTS = $(SCRIPTS_ALL) $(SCRIPTS_MORE1) $(SCRIPTS_MORE3) $(SCRIPTS_MORE4)
 
-TEST_OUTFILES = $(SCRIPTS_FIRST) $(SCRIPTS) $(SCRIPTS_WIN32) $(SCRIPTS_GUI)
+TEST_OUTFILES = $(SCRIPTS) $(SCRIPTS_WIN32) $(SCRIPTS_GUI)
 DOSTMP = dostmp
 DOSTMP_OUTFILES = $(TEST_OUTFILES:test=dostmp\test)
 DOSTMP_INFILES = $(DOSTMP_OUTFILES:.out=.in)
 
 .SUFFIXES: .in .out
 
-# Must run test1 first to create small.vim.
-$(SCRIPTS) $(SCRIPTS_GUI) $(SCRIPTS_WIN32) $(NEW_TESTS): $(SCRIPTS_FIRST)
-
 nongui:	nolog $(SCRIPTS_FIRST) $(SCRIPTS) report
 
 small:	nolog report
@@ -39,14 +36,19 @@ gui:	nolog $(SCRIPTS_FIRST) $(SCRIPTS) $
 win32:	nolog $(SCRIPTS_FIRST) $(SCRIPTS) $(SCRIPTS_WIN32) report
 
 # Copy the input files to dostmp, changing the fileformat to dos.
-$(DOSTMP_INFILES): $(*B).in
+{}.in{$(DOSTMP)}.in:
 	if not exist $(DOSTMP)\NUL md $(DOSTMP)
 	if exist $@ del $@
 	$(VIMPROG) -u dos.vim --noplugin "+set ff=dos|f $@|wq" $(*B).in
 
+# Must run test1 first to create small.vim.
+$(DOSTMP_INFILES): $(*B).in $(SCRIPTS_FIRST)
+
+$(DOSTMP)\$(SCRIPTS_FIRST:.out=.in): $(SCRIPTS_FIRST:.out=.in)
+
 # For each input file dostmp/test99.in run the tests.
 # This moves test99.in to test99.in.bak temporarily.
-$(TEST_OUTFILES): $(DOSTMP)\$(*B).in
+{$(DOSTMP)}.in{}.out:
 	-@if exist test.out DEL test.out
 	move $(*B).in $(*B).in.bak
 	copy $(DOSTMP)\$(*B).in $(*B).in
@@ -67,6 +69,10 @@ win32:	nolog $(SCRIPTS_FIRST) $(SCRIPTS)
 		 & echo $* FAILED >> test.log ) \
 		else ( move /y test.out $*.out )
 
+$(TEST_OUTFILES): $(DOSTMP)\$(*B).in
+
+$(SCRIPTS_FIRST): $(DOSTMP)\$(SCRIPTS_FIRST:.out=.in)
+
 report:
 	@echo ""
 	@echo Test results:
