# HG changeset patch
# Parent  caee9641248e38101f1ecbe64c1e1726d6cffd8c

diff --git a/runtime/doc/eval.txt b/runtime/doc/eval.txt
--- a/runtime/doc/eval.txt
+++ b/runtime/doc/eval.txt
@@ -8274,7 +8274,7 @@ See |:verbose-cmd| for more information.
 				:    let x += 1
 				:    return x
 				:  endfunction
-				:  return function('Bar')
+				:  return Bar
 				:endfunction
 
 				:let F = Foo()
@@ -8284,6 +8284,14 @@ See |:verbose-cmd| for more information.
 <				2 >
 				:echo F()
 <				3
+								*Exxx*
+			A closure is actually an unnamed function similar to
+			|lambda| with the internal name starting with
+			"<closure>" instead of "<lambda>".  In the above
+			example, the Funcref to the closure is set to a local
+			variable "Bar".  A closure is a local function because
+			it uses local variables of the outer function, thus
+			it cannot have a prefix "s:" nor "g:".
 
 						*function-search-undo*
 			The last used search pattern and the redo command "."
diff --git a/src/testdir/test_lambda.vim b/src/testdir/test_lambda.vim
--- a/src/testdir/test_lambda.vim
+++ b/src/testdir/test_lambda.vim
@@ -219,17 +219,19 @@ endfunction
 function! Test_closure_counter()
   function! s:foo()
     let x = 0
-    function! s:bar() closure
+    function! Bar() closure
       let x += 1
       return x
     endfunction
-    return function('s:bar')
+    return Bar
   endfunction
 
   let l:F = s:foo()
   call test_garbagecollect_now()
   call assert_equal(1, l:F())
   call assert_equal(2, l:F())
+  let l:F2 = s:foo()
+  call assert_equal(1, l:F2())
   call assert_equal(3, l:F())
   call assert_equal(4, l:F())
 endfunction
@@ -237,13 +239,47 @@ endfunction
 function! Test_closure_unlet()
   function! s:foo()
     let x = 1
-    function! s:bar() closure
+    function! Bar() closure
       unlet x
     endfunction
-    call s:bar()
+    call Bar()
     return l:
   endfunction
 
   call assert_false(has_key(s:foo(), 'x'))
   call test_garbagecollect_now()
 endfunction
+
+function! Test_closure_redefine()
+  function! s:foo()
+    function Bar() closure
+    endfunction
+    function Bar() closure
+    endfunction
+  endfunction
+
+  call assert_fails('call s:foo()', 'E707:')
+
+  function! s:foo()
+    function Bar() closure
+    endfunction
+    function! Bar() closure
+    endfunction
+  endfunction
+
+  call s:foo()	" no error
+endfunction
+
+function! Test_closure_nonlocal()
+  function! s:foo1()
+    function! s:bar() closure
+    endfunction
+  endfunction
+  function! s:foo2()
+    function! g:Bar() closure
+    endfunction
+  endfunction
+
+  call assert_fails('call s:foo1()', 'Exxx:')
+  call assert_fails('call s:foo2()', 'Exxx:')
+endfunction
diff --git a/src/userfunc.c b/src/userfunc.c
--- a/src/userfunc.c
+++ b/src/userfunc.c
@@ -1195,6 +1195,24 @@ builtin_function(char_u *name, int len)
     return p == NULL || (len > 0 && p > name + len);
 }
 
+/*
+ * Check if the function "name" is an unnamed function.
+ * Return TRUE if it is an unnamed function, FALSE otherwise.
+ */
+    static int
+is_unnamed_func(char_u *name)
+{
+    if (name == NULL)
+	return FALSE;
+    if (isdigit(*name))
+	return TRUE;
+    else if (STRNCMP(name, "<lambda>", 8) == 0)
+	return TRUE;
+    else if (STRNCMP(name, "<closure>", 9) == 0)
+	return TRUE;
+    return FALSE;
+}
+
     int
 func_call(
     char_u	*name,
@@ -1378,8 +1396,7 @@ call_func(
 		    call_user_func(fp, argcount, argvars, rettv,
 					       firstline, lastline,
 				  (fp->uf_flags & FC_DICT) ? selfdict : NULL);
-		    if (--fp->uf_calls <= 0 && (isdigit(*fp->uf_name)
-				|| STRNCMP(fp->uf_name, "<lambda>", 8) == 0)
+		    if (--fp->uf_calls <= 0 && is_unnamed_func(fp->uf_name)
 						      && fp->uf_refcount <= 0)
 			/* Function was unreferenced while being used, free it
 			 * now. */
@@ -1735,20 +1752,24 @@ ex_function(exarg_T *eap)
     int		saved_did_emsg;
     int		saved_wait_return = need_wait_return;
     char_u	*name = NULL;
+    char_u	*clname = NULL;
     char_u	*p;
     char_u	*arg;
     char_u	*line_arg = NULL;
+    char_u	*invarg = NULL;
     garray_T	newargs;
     garray_T	newlines;
     int		varargs = FALSE;
     int		flags = 0;
     ufunc_T	*fp;
+    ufunc_T	*fp_tofree = NULL;
     int		indent;
     int		nesting;
     char_u	*skip_until = NULL;
     dictitem_T	*v;
     funcdict_T	fudi;
     static int	func_nr = 0;	    /* number for nameless function */
+    static int	closure_nr = 0;	    /* number for closure function */
     int		paren;
     hashtab_T	*ht;
     int		todo;
@@ -1943,7 +1964,7 @@ ex_function(exarg_T *eap)
 						      : eval_isnamec(arg[j])))
 		++j;
 	    if (arg[j] != NUL)
-		emsg_funcname((char *)e_invarg2, arg);
+		invarg = arg;	/* Invalid argument: show the message later. */
 	}
 	/* Disallow using the g: dict. */
 	if (fudi.fd_dict != NULL && fudi.fd_dict->dv_scope == VAR_DEF_SCOPE)
@@ -1981,6 +2002,9 @@ ex_function(exarg_T *eap)
 	    break;
     }
 
+    if (invarg != NULL && (flags & FC_CLOSURE) == 0)
+	emsg_funcname((char *)e_invarg2, invarg);
+
     /* When there is a line break use what follows for the function body.
      * Makes 'exe "func Test()\n...\nendfunc"' work. */
     if (*p == '\n')
@@ -2177,15 +2201,46 @@ ex_function(exarg_T *eap)
      */
     if (fudi.fd_dict == NULL)
     {
+	if ((flags & FC_CLOSURE) != 0)
+	{
+	    char_u  *q = eap->arg;
+
+	    vim_free(name);
+	    name = trans_function_name(&q, eap->skip, TFN_NO_DEREF, NULL, NULL);
+	}
+
 	v = find_var(name, &ht, FALSE);
-	if (v != NULL && v->di_tv.v_type == VAR_FUNC)
+	if (v != NULL && v->di_tv.v_type == VAR_FUNC
+		&& ((flags & FC_CLOSURE) == 0 || !eap->forceit))
 	{
 	    emsg_funcname(N_("E707: Function name conflicts with variable: %s"),
 									name);
 	    goto erret;
 	}
 
-	fp = find_func(name);
+	if ((flags & FC_CLOSURE) != 0)
+	{
+	    char	numbuf[22];
+
+	    if ((name[0] == K_SPECIAL && name[1] == KS_EXTRA
+						   && name[2] == (int)KE_SNR)
+		    || ht != get_funccal_local_ht()
+		    || (eap->arg[0] != 'l' && eap->arg[1] == ':'))
+	    {
+		emsg_funcname(N_("Exxx: Closure function should be a local function: %s"),
+									name);
+		goto erret;
+	    }
+	    clname = name;
+	    sprintf(numbuf, "<closure>%d", ++closure_nr);
+	    name = vim_strsave((char_u *)numbuf);
+	    if (name == NULL)
+		goto erret;
+	    fp = NULL;
+	}
+	else
+	    fp = find_func(name);
+
 	if (fp != NULL)
 	{
 	    if (!eap->forceit)
@@ -2268,6 +2323,7 @@ ex_function(exarg_T *eap)
 	fp = (ufunc_T *)alloc((unsigned)(sizeof(ufunc_T) + STRLEN(name)));
 	if (fp == NULL)
 	    goto erret;
+	fp_tofree = fp;
 
 	if (fudi.fd_dict != NULL)
 	{
@@ -2276,14 +2332,10 @@ ex_function(exarg_T *eap)
 		/* add new dict entry */
 		fudi.fd_di = dictitem_alloc(fudi.fd_newkey);
 		if (fudi.fd_di == NULL)
-		{
-		    vim_free(fp);
 		    goto erret;
-		}
 		if (dict_add(fudi.fd_dict, fudi.fd_di) == FAIL)
 		{
 		    vim_free(fudi.fd_di);
-		    vim_free(fp);
 		    goto erret;
 		}
 	    }
@@ -2302,10 +2354,7 @@ ex_function(exarg_T *eap)
 	/* insert the new function in the function list */
 	STRCPY(fp->uf_name, name);
 	if (hash_add(&func_hashtab, UF2HIKEY(fp)) == FAIL)
-	{
-	    vim_free(fp);
 	    goto erret;
-	}
     }
     fp->uf_args = newargs;
     fp->uf_lines = newlines;
@@ -2317,6 +2366,7 @@ ex_function(exarg_T *eap)
 		    name);
 	    goto erret;
 	}
+	fp->uf_refcount = 1;
 	fp->uf_scoped = current_funccal;
 	current_funccal->fc_refcount++;
 	if (ga_grow(&current_funccal->fc_funcs, 1) == FAIL)
@@ -2324,6 +2374,15 @@ ex_function(exarg_T *eap)
 	((ufunc_T **)current_funccal->fc_funcs.ga_data)
 				[current_funccal->fc_funcs.ga_len++] = fp;
 	func_ref(current_funccal->func->uf_name);
+
+	if ((flags & FC_DICT) == 0)
+	{
+	    typval_T    tv;
+
+	    tv.v_type = VAR_FUNC;
+	    tv.vval.v_string = name;
+	    set_var(clname, &tv, TRUE);
+	}
     }
     else
 	fp->uf_scoped = NULL;
@@ -2346,10 +2405,12 @@ erret:
     ga_clear_strings(&newargs);
 errret_2:
     ga_clear_strings(&newlines);
+    vim_free(fp_tofree);
 ret_free:
     vim_free(skip_until);
     vim_free(fudi.fd_newkey);
     vim_free(name);
+    vim_free(clname);
     did_emsg |= saved_did_emsg;
     need_wait_return |= saved_wait_return;
 }
@@ -2670,8 +2731,7 @@ get_user_func_name(expand_T *xp, int idx
 	    ++hi;
 	fp = HI2UF(hi);
 
-	if ((fp->uf_flags & FC_DICT)
-				|| STRNCMP(fp->uf_name, "<lambda>", 8) == 0)
+	if ((fp->uf_flags & FC_DICT) || is_unnamed_func(fp->uf_name))
 	    return (char_u *)""; /* don't show dict and lambda functions */
 
 	if (STRLEN(fp->uf_name) + 4 >= IOSIZE)
@@ -2758,7 +2818,7 @@ func_unref(char_u *name)
 {
     ufunc_T *fp = NULL;
 
-    if (name == NULL)
+    if (!is_unnamed_func(name))
 	return;
     if (isdigit(*name))
     {
@@ -2771,7 +2831,7 @@ func_unref(char_u *name)
 		EMSG2(_(e_intern2), "func_unref()");
 	}
     }
-    else if (STRNCMP(name, "<lambda>", 8) == 0)
+    else
     {
 	/* fail silently, when lambda function isn't found. */
 	fp = find_func(name);
@@ -2793,7 +2853,7 @@ func_ref(char_u *name)
 {
     ufunc_T *fp;
 
-    if (name == NULL)
+    if (!is_unnamed_func(name))
 	return;
     else if (isdigit(*name))
     {
@@ -2803,7 +2863,7 @@ func_ref(char_u *name)
 	else
 	    ++fp->uf_refcount;
     }
-    else if (STRNCMP(name, "<lambda>", 8) == 0)
+    else
     {
 	/* fail silently, when lambda function isn't found. */
 	fp = find_func(name);
