# HG changeset patch
# Parent  805eacbc6d5b28b57a3734f4399bcbdf64d67353

diff --git a/src/os_win32.c b/src/os_win32.c
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -4879,7 +4879,7 @@ mch_start_job(char_u *cmd, job_T *job, j
 	jo = NULL;
     }
     ResumeThread(pi.hThread);
-    CloseHandle(job->jv_proc_info.hThread);
+    CloseHandle(pi.hThread);
     job->jv_proc_info = pi;
     job->jv_job_object = jo;
     job->jv_status = JOB_STARTED;
