# HG changeset patch
# Parent  76a5bf2cc8f890b05939285c6526da37f30ef10d

diff --git a/src/memline.c b/src/memline.c
--- a/src/memline.c
+++ b/src/memline.c
@@ -2142,7 +2142,23 @@ get_ctime(time_t thetime, int add_newlin
     if (curtime == NULL)
 	vim_strncpy((char_u *)buf, (char_u *)_("(Invalid)"), sizeof(buf) - 1);
     else
+    {
 	(void)strftime(buf, sizeof(buf) - 1, "%a %b %d %H:%M:%S %Y", curtime);
+# ifdef MSWIN
+	if (enc_codepage >= 0 && (int)GetACP() != enc_codepage)
+	{
+	    char_u	*to_free = NULL;
+	    int		maxlen;
+
+	    acp_to_enc((char_u *)buf, (int)strlen(buf), &to_free, &maxlen);
+	    if (to_free != NULL)
+	    {
+		STRCPY(buf, to_free);
+		vim_free(to_free);
+	    }
+	}
+# endif
+    }
 #else
     STRCPY(buf, "(unknown)");
 #endif
