# HG changeset patch
# Parent 486655e0c5a21469364d3cf895535137f09b3724

diff --git a/src/os_win32.c b/src/os_win32.c
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -2509,7 +2509,7 @@
     WCHAR		*porig, *porigPrev;
     int			flen;
     WIN32_FIND_DATAW	fb;
-    HANDLE		hFind;
+    HANDLE		hFind = INVALID_HANDLE_VALUE;
     int			c;
     int			slen;
 
@@ -2517,6 +2517,8 @@
     if (flen > _MAX_PATH)
 	return OK;
 
+    szTrueName[0] = NUL;
+
     /* slash_adjust(name) not needed, already adjusted by fname_case(). */
 
     /* Build the new name in szTrueName[] one component at a time. */
@@ -2664,6 +2666,8 @@
     if (flen > _MAX_PATH)
 	return;
 
+    szTrueName[0] = NUL;
+
     /* Build the new name in szTrueName[] one component at a time. */
     porig = name;
     ptrue = szTrueName;
@@ -6272,6 +6276,7 @@
 		    while (i > 0)
 			free(argv[--i]);
 		    free(argv);
+		    argv = NULL;
 		    argc = 0;
 		}
 	    }
