# HG changeset patch
# Parent 803c1d7e72973d3c43278ce018d8328f7fbb5d9a
diff --git a/src/screen.c b/src/screen.c
--- a/src/screen.c
+++ b/src/screen.c
@@ -3705,7 +3705,7 @@ win_line(wp, lnum, startrow, endrow, noc
 		    if (diff_hlf != (hlf_T)0)
 			char_attr = hl_attr(diff_hlf);
 #endif
-		    p_extra = NUL;
+		    p_extra = NULL;
 		    c_extra = ' ';
 		    n_extra = get_breakindent_win(wp,
 				       ml_get_buf(wp->w_buffer, lnum, FALSE));
