# HG changeset patch
# Parent 7818ca6de3d07439571bc1fe788a3efff6979531
diff --git a/src/os_win32.c b/src/os_win32.c
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -3830,7 +3830,7 @@
 
     static BOOL
 vim_create_process(
-    const char		*cmd,
+    char		*cmd,
     DWORD		flags,
     BOOL		inherit_handles,
     STARTUPINFO		*si,
@@ -3853,7 +3853,7 @@
 		flags,			/* Creation flags */
 		NULL,			/* Environment */
 		NULL,			/* Current directory */
-		si,			/* Startup information */
+		(LPSTARTUPINFOW)si,	/* Startup information */
 		pi);			/* Process information */
 	    vim_free(wcmd);
 	    return ret;
