# HG changeset patch
# Parent  cae39edf37777565a29a6bb9cf6507bfad571845

diff --git a/src/mbyte.c b/src/mbyte.c
--- a/src/mbyte.c
+++ b/src/mbyte.c
@@ -4904,8 +4904,20 @@ im_preedit_window_set_position(void)
     if (preedit_window == NULL)
 	return;
 
+#if GTK_CHECK_VERSION(3,22,0)
+    {
+	GdkDisplay *dpy = gtk_widget_get_display(preedit_window);
+	GdkWindow *win = gtk_widget_get_window(preedit_window);
+	GdkMonitor *monitor = gdk_display_get_monitor_at_window(dpy, win);
+	GdkRectangle geometry;
+	gdk_monitor_get_geometry(monitor, &geometry);
+	sw = geometry.width;
+	sh = geometry.height;
+    }
+#else
     sw = gdk_screen_get_width(gtk_widget_get_screen(preedit_window));
     sh = gdk_screen_get_height(gtk_widget_get_screen(preedit_window));
+#endif
 #if GTK_CHECK_VERSION(3,0,0)
     gdk_window_get_origin(gtk_widget_get_window(gui.drawarea), &x, &y);
 #else
