# HG changeset patch
# Parent ad2128196d42f41f140bc14bfdedcb32f55c3172

diff --git a/nsis/gvim.nsi b/nsis/gvim.nsi
--- a/nsis/gvim.nsi
+++ b/nsis/gvim.nsi
@@ -40,7 +40,7 @@
 ComponentText "This will install Vim ${VER_MAJOR}.${VER_MINOR} on your computer."
 DirText "Choose a directory to install Vim (must end in 'vim')"
 Icon icons\vim_16c.ico
-# NSIS2 uses a different strategy with six diferent images in a strip...
+# NSIS2 uses a different strategy with six different images in a strip...
 #EnabledBitmap icons\enabled.bmp
 #DisabledBitmap icons\disabled.bmp
 UninstallText "This will uninstall Vim ${VER_MAJOR}.${VER_MINOR} from your system."
@@ -250,7 +250,7 @@
 	    # Windows 95/98/ME: not supported
 	    Goto lbl_done
 	lbl_winnt:
-	    # Windows NT/2000/XT and later
+	    # Windows NT/2000/XP and later
 	    File /oname=vim.exe ${VIMSRC}\vimw32.exe
 	lbl_done:
 	StrCpy $2 "$2 vim view vimdiff"
