# HG changeset patch
# Parent  965c67c10f1b984252319372935058a2a06a9dd4

diff --git a/src/macros.h b/src/macros.h
--- a/src/macros.h
+++ b/src/macros.h
@@ -331,7 +331,7 @@
 #  if defined(WIN32)
 #   ifndef isnan
 #    define isnan(x) _isnan(x)
-     static inline int isinf(double x) { return !_finite(x) && !_isnan(x); }
+     static __inline int isinf(double x) { return !_finite(x) && !_isnan(x); }
 #   endif
 #  else
 #   ifndef HAVE_ISNAN
