# HG changeset patch
# Parent 5d91d4dbbb46799342f635146641ad925f86e151
diff --git a/src/spell.c b/src/spell.c
--- a/src/spell.c
+++ b/src/spell.c
@@ -4234,7 +4234,9 @@
     if (spl_copy == NULL)
 	goto theend;
 
+#ifdef FEAT_MBYTE
     wp->w_s->b_cjk = 0;
+#endif
 
     /* Loop over comma separated language names. */
     for (splp = spl_copy; *splp != NUL; )
@@ -4244,11 +4246,13 @@
 	region = NULL;
 	len = (int)STRLEN(lang);
 
+#ifdef FEAT_MBYTE
 	if (STRCMP(lang, "cjk") == 0)
 	{
 	    wp->w_s->b_cjk = 1;
 	    continue;
 	}
+#endif
 
 	/* If the name ends in ".spl" use it as the name of the spell file.
 	 * If there is a region name let "region" point to it and remove it
