# HG changeset patch
# Parent 42a3184767bcffce6333ea3b0b73266a76328185
diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -366,7 +366,11 @@ endif
 endif
 CC := $(CROSS_COMPILE)gcc
 CXX := $(CROSS_COMPILE)g++
+ifeq ($(UNDER_CYGWIN),yes)
 WINDRES := $(CROSS_COMPILE)windres
+else
+WINDRES := windres
+endif
 WINDRES_CC = $(CC)
 
 #>>>>> end of choices
