# HG changeset patch
# Parent 3de9e96d50c35c2664f657bb3847d62ec5be59bd

diff --git a/src/if_python.c b/src/if_python.c
--- a/src/if_python.c
+++ b/src/if_python.c
@@ -756,8 +756,10 @@
 	    goto fail;
 
 	/* The first python thread is vim's, release the lock. */
-	Python_SaveThread();
+	PyEval_SaveThread();
+#ifdef PY_CAN_RECURSE
 	pygilstate = PyGILState_Ensure();
+#endif
 
 	globals = PyModule_GetDict(PyImport_AddModule("__main__"));
 
@@ -767,7 +769,16 @@
 	 * the current directory in sys.path. */
 	PyRun_SimpleString("import sys; sys.path = filter(lambda x: x != '/must>not&exist', sys.path)");
 
+	/* lock is created and acquired in PyEval_InitThreads() and thread
+	 * state is created in Py_Initialize()
+	 * there _PyGILState_NoteThreadState() also sets gilcounter to 1
+	 * (python must have threads enabled!)
+	 * so the following does both: unlock GIL and save thread state in TLS
+	 * without deleting thread state
+	 */
+#ifdef PY_CAN_RECURSE
 	PyGILState_Release(pygilstate);
+#endif
 
 	initialised = 1;
     }
diff --git a/src/if_python3.c b/src/if_python3.c
--- a/src/if_python3.c
+++ b/src/if_python3.c
@@ -729,13 +729,11 @@
 #else
 	PyMac_Initialize();
 #endif
-	/* Initialise threads, and save the state using PyGILState_Ensure.
-	 * Without the call to PyGILState_Ensure, thread specific state (such
-	 * as the system trace hook), will be lost between invocations of
-	 * Python code. */
+	/* Initialise threads, and below save the state using
+	 * PyGILState_Ensure.  Without the call to PyGILState_Ensure, thread
+	 * specific state (such as the system trace hook), will be lost
+	 * between invocations of Python code. */
 	PyEval_InitThreads();
-	pygilstate = PyGILState_Ensure();
-
 #ifdef DYNAMIC_PYTHON3
 	get_py3_exceptions();
 #endif
@@ -743,6 +741,10 @@
 	if (PythonIO_Init())
 	    goto fail;
 
+	/* The first python thread is vim's, release the lock. */
+	PyEval_SaveThread();
+	pygilstate = PyGILState_Ensure();
+
 	globals = PyModule_GetDict(PyImport_AddModule("__main__"));
 
 	/* Remove the element from sys.path that was added because of our
@@ -754,12 +756,13 @@
 	 */
 	PyRun_SimpleString("import vim; import sys; sys.path = list(filter(lambda x: not x.endswith('must>not&exist'), sys.path))");
 
-	// lock is created and acquired in PyEval_InitThreads() and thread
-	// state is created in Py_Initialize()
-	// there _PyGILState_NoteThreadState() also sets gilcounter to 1
-	// (python must have threads enabled!)
-	// so the following does both: unlock GIL and save thread state in TLS
-	// without deleting thread state
+	/* lock is created and acquired in PyEval_InitThreads() and thread
+	 * state is created in Py_Initialize()
+	 * there _PyGILState_NoteThreadState() also sets gilcounter to 1
+	 * (python must have threads enabled!)
+	 * so the following does both: unlock GIL and save thread state in TLS
+	 * without deleting thread state
+	 */
 	PyGILState_Release(pygilstate);
 
 	py3initialised = 1;
