# HG changeset patch
# Parent 77fbb2dbb517de04021aeaec79135867b9d14328
diff --git a/src/gui_w32.c b/src/gui_w32.c
--- a/src/gui_w32.c
+++ b/src/gui_w32.c
@@ -122,7 +122,7 @@
 	    return FAIL;
     }
 
-    /* Enable DirectX/DirextWrite */
+    /* Enable DirectX/DirectWrite */
     if (dx_enable)
     {
 	if (!directx_enabled())
diff --git a/src/option.c b/src/option.c
--- a/src/option.c
+++ b/src/option.c
@@ -6965,7 +6965,7 @@
 #endif
 
 #if defined(FEAT_GUI) && defined(FEAT_RENDERING_OPTIONS)
-    else if (varp == &p_renopt && gui.in_use)
+    else if (varp == &p_renopt)
     {
 	if (!gui_mch_set_rendering_options(p_renopt))
 	    errmsg = e_invarg;
