# HG changeset patch
# Parent 1145b0eb0c4a98098f2e16b5bd30f2f7db274eab
diff --git a/src/ops.c b/src/ops.c
--- a/src/ops.c
+++ b/src/ops.c
@@ -2646,7 +2646,7 @@
 
 	/* The user may have moved the cursor before inserting something, try
 	 * to adjust the block for that. */
-	if (oap->start.lnum == curbuf->b_op_start.lnum)
+	if (oap->start.lnum == curbuf->b_op_start.lnum && !bd.is_MAX)
 	{
 	    if (oap->op_type == OP_INSERT
 		    && oap->start.col != curbuf->b_op_start.col)
