# HG changeset patch
# Parent  70fd8b5782c1f7b146235d52bb4163a996523f3d

diff --git a/src/if_py_both.h b/src/if_py_both.h
--- a/src/if_py_both.h
+++ b/src/if_py_both.h
@@ -3838,7 +3838,7 @@ get_firstwin(TabPageObject *tabObject)
 }
 static char *WindowAttrs[] = {
     "buffer", "cursor", "height", "vars", "options", "number", "row", "col",
-    "tabpage", "valid",
+    "tabpage", "valid", "width",
     NULL
 };
 
