# HG changeset patch
# Parent  03a44532fe4cc0b1f8e83814e5617e1cd0856eb2

diff --git a/src/if_ruby.c b/src/if_ruby.c
--- a/src/if_ruby.c
+++ b/src/if_ruby.c
@@ -327,8 +327,11 @@ static void ruby_vim_init(void);
 # define ruby_init			dll_ruby_init
 # define ruby_init_loadpath		dll_ruby_init_loadpath
 # ifdef WIN3264
-#  define NtInitialize			dll_NtInitialize
-#  define ruby_sysinit			dll_ruby_sysinit
+#  ifdef RUBY19_OR_LATER
+#   define ruby_sysinit			dll_ruby_sysinit
+#  else
+#   define NtInitialize			dll_NtInitialize
+#  endif
 #  if defined(DYNAMIC_RUBY_VER) && DYNAMIC_RUBY_VER >= 18
 #   define rb_w32_snprintf		dll_rb_w32_snprintf
 #  endif
@@ -439,8 +442,11 @@ static VALUE *dll_ruby_errinfo;
 static void (*dll_ruby_init) (void);
 static void (*dll_ruby_init_loadpath) (void);
 # ifdef WIN3264
+#  ifdef RUBY19_OR_LATER
+static void (*dll_ruby_sysinit) (int*, char***);
+#  else
 static void (*dll_NtInitialize) (int*, char***);
-static void (*dll_ruby_sysinit) (int*, char***);
+#  endif
 #  if defined(DYNAMIC_RUBY_VER) && DYNAMIC_RUBY_VER >= 18
 static int (*dll_rb_w32_snprintf)(char*, size_t, const char*, ...);
 #  endif
