# HG changeset patch
# Parent  9f798e77286a7f441c910b3c104c86e3746695fb

diff --git a/src/testdir/Makefile b/src/testdir/Makefile
--- a/src/testdir/Makefile
+++ b/src/testdir/Makefile
@@ -129,7 +129,7 @@ newtestssilent: $(NEW_TESTS)
 .vim.res: writevimcmd
 	@echo "$(VIMPROG)" > vimcmd
 	@echo "$(RUN_VIMTEST)" >> vimcmd
-	$(RUN_VIMTEST) $(NO_INITS) -S runtest.vim $*.vim > /dev/null
+	-$(RUN_VIMTEST) $(NO_INITS) -S runtest.vim $*.vim > /dev/null
 	@rm vimcmd
 
 test_gui.res: test_gui.vim
