# HG changeset patch
# Parent  b4aeb70a6b52839513e5a20a20ef14953ebe5157

diff --git a/src/ex_cmds2.c b/src/ex_cmds2.c
--- a/src/ex_cmds2.c
+++ b/src/ex_cmds2.c
@@ -1265,8 +1265,16 @@ set_ref_in_timer(int copyID)
 
     for (timer = first_timer; timer != NULL; timer = timer->tr_next)
     {
-	tv.v_type = VAR_PARTIAL;
-	tv.vval.v_partial = timer->tr_partial;
+	if (timer->tr_partial != NULL)
+	{
+	    tv.v_type = VAR_PARTIAL;
+	    tv.vval.v_partial = timer->tr_partial;
+	}
+	else
+	{
+	    tv.v_type = VAR_FUNC;
+	    tv.vval.v_string = timer->tr_callback;
+	}
 	abort = abort || set_ref_in_item(&tv, copyID, NULL, NULL);
     }
     return abort;
diff --git a/src/testdir/test_lambda.vim b/src/testdir/test_lambda.vim
--- a/src/testdir/test_lambda.vim
+++ b/src/testdir/test_lambda.vim
@@ -21,7 +21,7 @@ function! Test_lambda_with_timer()
   let s:timer_id = 0
   function! s:Foo()
     "let n = 0
-    let s:timer_id = timer_start(50, {-> execute("let s:n += 1 | echo s:n")}, {"repeat": -1})
+    let s:timer_id = timer_start(50, {-> execute("let s:n += 1 | echo s:n", "")}, {"repeat": -1})
   endfunction
 
   call s:Foo()
@@ -136,6 +136,18 @@ function! Test_lambda_call_lambda_from_l
   call assert_equal(1, l:F()())
 endfunction
 
+function! Test_lambda_delfunc()
+  function! s:gen()
+    let l:Foo = {-> 0}
+    let l:Bar = l:Foo
+    delfunction l:Foo
+    return l:Bar
+  endfunction
+
+  let l:F = s:gen()
+  call assert_fails(':call l:F()', 'E117:')
+endfunction
+
 function! Test_lambda_combination()
   call assert_equal(2, {x -> {x -> x}}(1)(2))
   call assert_equal(10, {y -> {x -> x(y)(10)}({y -> y})}({z -> z}))
diff --git a/src/userfunc.c b/src/userfunc.c
--- a/src/userfunc.c
+++ b/src/userfunc.c
@@ -261,6 +261,7 @@ get_lambda_tv(char_u **arg, typval_T *re
 {
     garray_T	newargs;
     garray_T	newlines;
+    garray_T	*pnewargs;
     ufunc_T	*fp = NULL;
     int		varargs;
     int		ret;
@@ -270,7 +271,7 @@ get_lambda_tv(char_u **arg, typval_T *re
     static int	lambda_no = 0;
     dict_T	*old_eval_lvars = eval_lvars_dict;
     dict_T	*old_eval_avars = eval_avars_dict;
-    dict_T	*d = NULL;
+    dict_T	*dl = NULL, *da = NULL;
 
     ga_init(&newargs);
     ga_init(&newlines);
@@ -281,23 +282,26 @@ get_lambda_tv(char_u **arg, typval_T *re
 	return NOTDONE;
 
     /* Parse the arguments again. */
+    if (evaluate)
+	pnewargs = &newargs;
+    else
+	pnewargs = NULL;
     *arg = skipwhite(*arg + 1);
-    ret = get_function_args(arg, '-', &newargs, &varargs, FALSE);
+    ret = get_function_args(arg, '-', pnewargs, &varargs, FALSE);
     if (ret == FAIL || **arg != '>')
 	goto errret;
 
+    /* Set up dictionaries for checking local variables and arguments. */
     if (evaluate)
     {
-	/* Don't use dict_alloc() here.  Avoid garbage collection.
-	 * These dicts have the same lifetime with "fp". */
-	d = (dict_T *)alloc(sizeof(dict_T) * 2);
-	if (d == NULL)
+	dl = dict_alloc();
+	if (dl == NULL)
 	    goto errret;
-
-	eval_lvars_dict = d;
-	eval_avars_dict = d + 1;
-	init_var_dict(eval_lvars_dict, NULL, 0);
-	init_var_dict(eval_avars_dict, NULL, 0);
+	da = dict_alloc();
+	if (da == NULL)
+	    goto errret;
+	eval_lvars_dict = dl;
+	eval_avars_dict = da;
     }
 
     /* Get the start and the end of the expression. */
@@ -317,6 +321,18 @@ get_lambda_tv(char_u **arg, typval_T *re
 	int	len;
 	char_u	*p;
 
+	/* Check if local variables and arguments were used. */
+	if (dl->dv_hashtab.ht_used == 0)
+	{
+	    dict_unref(dl);
+	    dl = NULL;
+	}
+	if (da->dv_hashtab.ht_used == 0)
+	{
+	    dict_unref(da);
+	    da = NULL;
+	}
+
 	sprintf((char*)name, "<lambda>%d", ++lambda_no);
 
 	fp = (ufunc_T *)alloc((unsigned)(sizeof(ufunc_T) + STRLEN(name)));
@@ -334,16 +350,15 @@ get_lambda_tv(char_u **arg, typval_T *re
 	    goto errret;
 	((char_u **)(newlines.ga_data))[newlines.ga_len++] = p;
 	STRCPY(p, "return ");
-	STRNCPY(p + 7, s, e - s);
-	p[7 + e - s] = NUL;
+	vim_strncpy(p + 7, s, e - s);
 
 	fp->uf_refcount = 1;
 	STRCPY(fp->uf_name, name);
 	hash_add(&func_hashtab, UF2HIKEY(fp));
 	fp->uf_args = newargs;
 	fp->uf_lines = newlines;
-	fp->uf_lvars = eval_lvars_dict;
-	fp->uf_avars = eval_avars_dict;
+	fp->uf_lvars = dl;
+	fp->uf_avars = da;
 
 #ifdef FEAT_PROFILE
 	fp->uf_tml_count = NULL;
@@ -361,10 +376,6 @@ get_lambda_tv(char_u **arg, typval_T *re
 	rettv->vval.v_string = vim_strsave(name);
 	rettv->v_type = VAR_FUNC;
     }
-    else
-    {
-	ga_clear_strings(&newargs);
-    }
 
     eval_lvars_dict = old_eval_lvars;
     eval_avars_dict = old_eval_avars;
@@ -374,12 +385,8 @@ errret:
     ga_clear_strings(&newargs);
     ga_clear_strings(&newlines);
     vim_free(fp);
-    if (d != NULL)
-    {
-	vars_clear(&d[0].dv_hashtab);
-	vars_clear(&d[1].dv_hashtab);
-	vim_free(d);
-    }
+    dict_unref(dl);
+    dict_unref(da);
     eval_lvars_dict = old_eval_lvars;
     eval_avars_dict = old_eval_avars;
     return FAIL;
@@ -1118,12 +1125,8 @@ func_free(ufunc_T *fp)
     /* clear this function */
     ga_clear_strings(&(fp->uf_args));
     ga_clear_strings(&(fp->uf_lines));
-    if (fp->uf_lvars != NULL)
-    {
-	vars_clear(&fp->uf_lvars->dv_hashtab);
-	vars_clear(&fp->uf_avars->dv_hashtab);
-	vim_free(fp->uf_lvars);
-    }
+    dict_unref(fp->uf_lvars);
+    dict_unref(fp->uf_avars);
 #ifdef FEAT_PROFILE
     vim_free(fp->uf_tml_count);
     vim_free(fp->uf_tml_total);
@@ -2775,6 +2778,7 @@ set_ref_in_func(
 {
     ufunc_T	*fp;
     int		abort = FALSE;
+    typval_T	dtv;
 
     if (name == NULL)
 	return abort;
@@ -2783,11 +2787,19 @@ set_ref_in_func(
     if (fp != NULL)
     {
 	if (fp->uf_lvars != NULL)
-	    abort = abort || set_ref_in_ht(&fp->uf_lvars->dv_hashtab,
-							copyID, list_stack);
+	{
+	    dtv.v_type = VAR_DICT;
+	    dtv.vval.v_dict = fp->uf_lvars;
+	    abort = abort ||
+			set_ref_in_item(&dtv, copyID, ht_stack, list_stack);
+	}
 	if (fp->uf_avars != NULL)
-	    abort = abort || set_ref_in_ht(&fp->uf_avars->dv_hashtab,
-							copyID, list_stack);
+	{
+	    dtv.v_type = VAR_DICT;
+	    dtv.vval.v_dict = fp->uf_avars;
+	    abort = abort ||
+			set_ref_in_item(&dtv, copyID, ht_stack, list_stack);
+	}
     }
     return abort;
 }
