# HG changeset patch
# Parent  0ccbc6536844a72fcc1ea7489746301a6bd3c40c

diff --git a/src/eval.c b/src/eval.c
--- a/src/eval.c
+++ b/src/eval.c
@@ -6828,6 +6828,7 @@ check_vars(
     {
 	if (find_var(name, NULL, TRUE) != NULL)
 	{
+	    TRACE("lvar: %s", varname);
 	    *eval_lavars_used = TRUE;
 	}
     }
@@ -6835,6 +6836,7 @@ check_vars(
     {
 	if (find_var(name, NULL, TRUE) != NULL)
 	{
+	    TRACE("avar: %s", varname);
 	    *eval_lavars_used = TRUE;
 	}
     }
diff --git a/src/userfunc.c b/src/userfunc.c
--- a/src/userfunc.c
+++ b/src/userfunc.c
@@ -336,6 +336,7 @@ get_lambda_tv(char_u **arg, typval_T *re
 	((char_u **)(newlines.ga_data))[newlines.ga_len++] = p;
 	STRCPY(p, "return ");
 	vim_strncpy(p + 7, s, e - s);
+	TRACE("lambda: la:%d, name: %s, expr: %s", eval_lavars, name, p + 7);
 
 	fp->uf_refcount = 1;
 	STRCPY(fp->uf_name, name);
@@ -352,6 +353,7 @@ get_lambda_tv(char_u **arg, typval_T *re
 	    ((ufunc_T **)current_funccal->fc_funcs.ga_data)
 				    [current_funccal->fc_funcs.ga_len++] = fp;
 	    func_ref(current_funccal->func->uf_name);
+	    TRACE("name:%s, fc_refcount:%d", current_funccal->func->uf_name, current_funccal->fc_refcount);
 	}
 	else
 	    fp->uf_scoped = NULL;
@@ -657,12 +659,17 @@ free_funccal(
     listitem_T	*li;
     int		i;
 
+    TRACE("free_funccal'ing: %s", fc->func->uf_name);
+
     for (i = 0; i < fc->fc_funcs.ga_len; ++i)
     {
 	ufunc_T	    *fp = ((ufunc_T **)(fc->fc_funcs.ga_data))[i];
 
 	if (fp != NULL)
+	{
+	    TRACE("free_funccal: detach: %s (%p)", fp->uf_name, fp->uf_name);
 	    fp->uf_scoped = NULL;
+	}
     }
 
     /* The a: variables typevals may not have been allocated, only free the
@@ -682,10 +689,14 @@ free_funccal(
 	ufunc_T	    *fp = ((ufunc_T **)(fc->fc_funcs.ga_data))[i];
 
 	if (fp != NULL)
+	{
+	    TRACE("free_funccal: detach2");
 	    func_unref(fc->func->uf_name);
+	}
     }
     ga_clear(&fc->fc_funcs);
 
+    TRACE("free_funccal'ed: %s", fc->func->uf_name);
     func_unref(fc->func->uf_name);
     vim_free(fc);
 }
@@ -1069,6 +1080,8 @@ call_user_func(
 	fc->caller = previous_funccal;
 	previous_funccal = fc;
 
+	TRACE("Add to previous_funccal: %s", fc->func->uf_name);
+
 	/* Make a copy of the a: variables, since we didn't do that above. */
 	todo = (int)fc->l_avars.dv_hashtab.ht_used;
 	for (hi = fc->l_avars.dv_hashtab.ht_array; todo > 0; ++hi)
@@ -1101,6 +1114,7 @@ funccal_unref(funccall_T *fc, ufunc_T *f
     if (fc == NULL)
 	return;
 
+    TRACE("funccal_unref'ing: %s, cnt:%d", fc->func->uf_name, fc->fc_refcount);
     if (--fc->fc_refcount <= 0)
     {
 	for (pfc = &previous_funccal; *pfc != NULL; )
@@ -1127,7 +1141,10 @@ funccal_unref(funccall_T *fc, ufunc_T *f
 	    for (i = 0; i < fc->fc_funcs.ga_len; ++i)
 	    {
 		if (((ufunc_T **)(fc->fc_funcs.ga_data))[i] == fp)
+		{
 		    ((ufunc_T **)(fc->fc_funcs.ga_data))[i] = NULL;
+		    TRACE("func_free: detach: %s", fp->uf_name);
+		}
 	    }
 	}
     }
@@ -1141,6 +1158,8 @@ func_free(ufunc_T *fp)
 {
     hashitem_T	*hi;
 
+    TRACE("func_free'ing: %s (%p)", fp->uf_name, fp->uf_name);
+
     /* clear this function */
     ga_clear_strings(&(fp->uf_args));
     ga_clear_strings(&(fp->uf_lines));
@@ -1159,6 +1178,8 @@ func_free(ufunc_T *fp)
 
     funccal_unref(fp->uf_scoped, fp);
 
+    TRACE("func_free'ed: %s (%p)", fp->uf_name, fp->uf_name);
+
     vim_free(fp);
 }
 
@@ -2326,6 +2347,7 @@ ex_function(exarg_T *eap)
 	((ufunc_T **)current_funccal->fc_funcs.ga_data)
 				[current_funccal->fc_funcs.ga_len++] = fp;
 	func_ref(current_funccal->func->uf_name);
+	TRACE("name:%s, fc_refcount:%d", current_funccal->func->uf_name, current_funccal->fc_refcount);
     }
     else
 	fp->uf_scoped = NULL;
@@ -2778,6 +2800,10 @@ func_unref(char_u *name)
 	/* fail silently, when lambda function isn't found. */
 	fp = find_func(name);
     }
+    if (fp != NULL)
+	TRACE("func_unref'ing: %s (%p), cnt:%d", name, name, fp->uf_refcount);
+    else
+	TRACE("func_unref: %s (%p) (not unref'ing)", name, name);
     if (fp != NULL && --fp->uf_refcount <= 0)
     {
 	/* Only delete it when it's not being used.  Otherwise it's done
@@ -2804,6 +2830,7 @@ func_ref(char_u *name)
 	    EMSG2(_(e_intern2), "func_ref()");
 	else
 	    ++fp->uf_refcount;
+	TRACE("func_ref: %s, cnt:%d", name, fp ? fp->uf_refcount : -1);
     }
     else if (STRNCMP(name, "<lambda>", 8) == 0)
     {
@@ -2811,6 +2838,7 @@ func_ref(char_u *name)
 	fp = find_func(name);
 	if (fp != NULL)
 	    ++fp->uf_refcount;
+	TRACE("func_ref: %s, cnt:%d", name, fp ? fp->uf_refcount : -1);
     }
 }
 
@@ -2821,6 +2849,8 @@ func_ref(char_u *name)
     static int
 can_free_funccal(funccall_T *fc, int copyID)
 {
+    TRACE("can_free_funccal: %s: copyID:%d lvl:%d ld:%d ad:%d fc:%d, ref:%d", fc->func ? fc->func->uf_name: NULL, copyID, fc->l_varlist.lv_copyID, fc->l_vars.dv_copyID, fc->l_avars.dv_copyID, fc->fc_copyID, fc->fc_refcount);
+
     return (fc->l_varlist.lv_copyID != copyID
 	    && fc->l_vars.dv_copyID != copyID
 	    && fc->l_avars.dv_copyID != copyID
@@ -3429,11 +3459,20 @@ free_unref_funccal(int copyID, int testi
     int		did_free = FALSE;
     int		did_free_funccal = FALSE;
     funccall_T	*fc, **pfc;
+    int i;
+
+    TRACE("free_unref_funccal()");
+    for (i = 0, pfc = &previous_funccal; *pfc != NULL && i < 100; i++)
+    {
+	TRACE("    previous_funccal chain: %p, %s (%p)", *pfc, (*pfc)->func->uf_name, (*pfc)->func->uf_name);
+	pfc = &(*pfc)->caller;
+    }
 
     for (pfc = &previous_funccal; *pfc != NULL; )
     {
 	if (can_free_funccal(*pfc, copyID))
 	{
+	    TRACE("free_unref_funccal: %p, %s", *pfc, (*pfc)->func->uf_name);
 	    fc = *pfc;
 	    *pfc = fc->caller;
 	    free_funccal(fc, TRUE);
@@ -3610,9 +3649,11 @@ set_ref_in_previous_funccal(int copyID)
     int		abort = FALSE;
     funccall_T	*fc;
 
+    TRACE("set_ref_in_previous_funccal()");
     for (fc = previous_funccal; fc != NULL; fc = fc->caller)
     {
 	fc->fc_copyID = copyID + 1;
+	TRACE("set_ref_in_previous_funccal: %s", fc->func->uf_name);
 	abort = abort || set_ref_in_ht(&fc->l_vars.dv_hashtab, copyID + 1,
 									NULL);
 	abort = abort || set_ref_in_ht(&fc->l_avars.dv_hashtab, copyID + 1,
@@ -3630,9 +3671,11 @@ set_ref_in_call_stack(int copyID)
     int		abort = FALSE;
     funccall_T	*fc;
 
+    TRACE("set_ref_in_call_stack()");
     for (fc = current_funccal; fc != NULL; fc = fc->caller)
     {
 	fc->fc_copyID = copyID;
+	TRACE("set_ref_in_call_stack: %s", fc->func->uf_name);
 	abort = abort || set_ref_in_ht(&fc->l_vars.dv_hashtab, copyID, NULL);
 	abort = abort || set_ref_in_ht(&fc->l_avars.dv_hashtab, copyID, NULL);
     }
@@ -3671,6 +3714,7 @@ set_ref_in_func(char_u *name, int copyID
     char_u	*tofree = NULL;
     char_u	*fname;
 
+    TRACE("set_ref_in_func(): %s", name);
     if (name == NULL)
 	return FALSE;
 
@@ -3678,11 +3722,13 @@ set_ref_in_func(char_u *name, int copyID
     fp = find_func(fname);
     if (fp != NULL)
     {
+	TRACE("set_ref_in_func: %s", fp->uf_name);
 	for (fc = fp->uf_scoped; fc != NULL; fc = fc->func->uf_scoped)
 	{
 	    if (fc->fc_copyID != copyID)
 	    {
 		fc->fc_copyID = copyID;
+		TRACE("set_ref_in_func: %s", fc->func->uf_name);
 		set_ref_in_ht(&fc->l_vars.dv_hashtab, copyID, NULL);
 		set_ref_in_ht(&fc->l_avars.dv_hashtab, copyID, NULL);
 	    }
