# HG changeset patch
# Parent  59621a1090036fa75d338d31cc369da074f69873

diff --git a/runtime/doc/eval.txt b/runtime/doc/eval.txt
--- a/runtime/doc/eval.txt
+++ b/runtime/doc/eval.txt
@@ -1219,10 +1219,19 @@ the following ways:
 
 1. The body of the lambda expression is an |expr1| and not a sequence of |Ex|
    commands.
-2. The prefix "a:" is optional for arguments.  E.g.: >
+2. The prefix "a:" should not be used for arguments.  E.g.: >
 	:let F = {arg1, arg2 -> arg1 - arg2}
 	:echo F(5, 2)
 <	3
+3. Lambda expressions can get the value of outer scope variables and arguments.
+   E.g.: >
+	:function Foo(arg)
+	:  let i = 3
+	:  return {x -> x + i - a:arg}
+	:endfunction
+	:let Bar = Foo(4)
+	:echo Bar(6)
+<	5
 
 The arguments are optional.  Example: >
 	:let F = {-> 'error function'}
diff --git a/src/dict.c b/src/dict.c
--- a/src/dict.c
+++ b/src/dict.c
@@ -188,7 +188,7 @@ dictitem_alloc(char_u *key)
 /*
  * Make a copy of a Dictionary item.
  */
-    static dictitem_T *
+    dictitem_T *
 dictitem_copy(dictitem_T *org)
 {
     dictitem_T *di;
diff --git a/src/eval.c b/src/eval.c
--- a/src/eval.c
+++ b/src/eval.c
@@ -237,6 +237,7 @@ static int get_env_tv(char_u **arg, typv
 
 static int get_env_len(char_u **arg);
 static char_u * make_expanded_name(char_u *in_start, char_u *expr_start, char_u *expr_end, char_u *in_end);
+static void check_vars(char_u *name, int len);
 static typval_T *alloc_string_tv(char_u *string);
 static hashtab_T *find_var_ht(char_u *name, char_u **varname);
 static void delete_var(hashtab_T *ht, hashitem_T *hi);
@@ -4332,6 +4333,9 @@ eval7(
 	    {
 		partial_T *partial;
 
+		if (!evaluate)
+		    check_vars(s, len);
+
 		/* If "s" is the name of a variable of type VAR_FUNC
 		 * use its contents. */
 		s = deref_func_name(s, &len, &partial, !evaluate);
@@ -4363,7 +4367,10 @@ eval7(
 	    else if (evaluate)
 		ret = get_var_tv(s, len, rettv, NULL, TRUE, FALSE);
 	    else
+	    {
+		check_vars(s, len);
 		ret = OK;
+	    }
 	}
 	vim_free(alias);
     }
@@ -5540,6 +5547,11 @@ set_ref_in_item(
 	    }
 	}
     }
+    else if (tv->v_type == VAR_FUNC)
+    {
+	abort = set_ref_in_func(tv->vval.v_string,
+						copyID, ht_stack, list_stack);
+    }
     else if (tv->v_type == VAR_PARTIAL)
     {
 	partial_T	*pt = tv->vval.v_partial;
@@ -5549,6 +5561,8 @@ set_ref_in_item(
 	 */
 	if (pt != NULL)
 	{
+	    abort = set_ref_in_func(pt->pt_name, copyID, ht_stack, list_stack);
+
 	    if (pt->pt_dict != NULL)
 	    {
 		typval_T dtv;
@@ -6791,6 +6805,57 @@ get_var_tv(
 }
 
 /*
+ * Check if variable "name" is a local variable or an argument.
+ * If it is a local variable, store it to the "eval_lvars_dict" dict.
+ * If it is an argument, store it to the "eval_avars_dict" dict.
+ */
+    static void
+check_vars(
+    char_u	*name,
+    int		len)		/* length of "name" */
+{
+    int		cc;
+    char_u	*varname;
+    hashtab_T	*ht;
+
+    /* truncate the name, so that we can use strcmp() */
+    cc = name[len];
+    name[len] = NUL;
+
+    ht = find_var_ht(name, &varname);
+    if ((ht == get_funccal_local_ht()) && (eval_lvars_dict != NULL))
+    {
+	dictitem_T  *di, *di2;
+
+	di = find_var_in_ht(ht, *name, varname, FALSE);
+	if (di != NULL
+		&& dict_find(eval_lvars_dict, DI2HIKEY(di), -1) == NULL)
+	{
+	    di2 = dictitem_copy(di);
+	    if (di2 != NULL)
+		dict_add(eval_lvars_dict, di2);
+	}
+    }
+    else if ((ht == get_funccal_args_ht()) && (eval_avars_dict != NULL))
+    {
+	dictitem_T  *di, *di2;
+
+	di = find_var_in_ht(ht, *name, varname, FALSE);
+	if (di != NULL
+		&& dict_find(eval_avars_dict, DI2HIKEY(di), -1) == NULL)
+	{
+	    di2 = dictitem_copy(di);
+	    if (di2 != NULL)
+		dict_add(eval_avars_dict, di2);
+	}
+    }
+
+    name[len] = cc;
+
+    return;
+}
+
+/*
  * Handle expr[expr], expr[expr:expr] subscript and .name lookup.
  * Also handle function call with Funcref variable: func(expr)
  * Can all be combined: dict.func(expr)[idx]['func'](expr)
@@ -7457,11 +7522,15 @@ init_var_dict(dict_T *dict, dictitem_T *
     dict->dv_scope = scope;
     dict->dv_refcount = DO_NOT_FREE_CNT;
     dict->dv_copyID = 0;
-    dict_var->di_tv.vval.v_dict = dict;
-    dict_var->di_tv.v_type = VAR_DICT;
-    dict_var->di_tv.v_lock = VAR_FIXED;
-    dict_var->di_flags = DI_FLAGS_RO | DI_FLAGS_FIX;
-    dict_var->di_key[0] = NUL;
+
+    if (dict_var != NULL)
+    {
+	dict_var->di_tv.vval.v_dict = dict;
+	dict_var->di_tv.v_type = VAR_DICT;
+	dict_var->di_tv.v_lock = VAR_FIXED;
+	dict_var->di_flags = DI_FLAGS_RO | DI_FLAGS_FIX;
+	dict_var->di_key[0] = NUL;
+    }
 }
 
 /*
diff --git a/src/globals.h b/src/globals.h
--- a/src/globals.h
+++ b/src/globals.h
@@ -1654,6 +1654,11 @@ EXTERN time_T time_for_testing INIT(= 0)
 
 /* Abort conversion to string after a recursion error. */
 EXTERN int  did_echo_string_emsg INIT(= FALSE);
+
+/* Dict for local variables used in a lambda expression. */
+EXTERN dict_T *eval_lvars_dict INIT(= NULL);
+/* Dict for arguments used in a lambda expression. */
+EXTERN dict_T *eval_avars_dict INIT(= NULL);
 #endif
 
 /*
diff --git a/src/proto/dict.pro b/src/proto/dict.pro
--- a/src/proto/dict.pro
+++ b/src/proto/dict.pro
@@ -5,6 +5,7 @@ void dict_unref(dict_T *d);
 int dict_free_nonref(int copyID);
 void dict_free_items(int copyID);
 dictitem_T *dictitem_alloc(char_u *key);
+dictitem_T *dictitem_copy(dictitem_T *org);
 void dictitem_remove(dict_T *dict, dictitem_T *item);
 void dictitem_free(dictitem_T *item);
 dict_T *dict_copy(dict_T *orig, int deep, int copyID);
diff --git a/src/proto/userfunc.pro b/src/proto/userfunc.pro
--- a/src/proto/userfunc.pro
+++ b/src/proto/userfunc.pro
@@ -18,6 +18,7 @@ char_u *get_user_func_name(expand_T *xp,
 void ex_delfunction(exarg_T *eap);
 void func_unref(char_u *name);
 void func_ref(char_u *name);
+int set_ref_in_func(char_u *name, int copyID, ht_stack_T **ht_stack, list_stack_T **list_stack);
 void ex_return(exarg_T *eap);
 void ex_call(exarg_T *eap);
 int do_return(exarg_T *eap, int reanimate, int is_cmd, void *rettv);
diff --git a/src/testdir/test_lambda.vim b/src/testdir/test_lambda.vim
--- a/src/testdir/test_lambda.vim
+++ b/src/testdir/test_lambda.vim
@@ -51,3 +51,101 @@ func Test_not_lamda()
   let x = {'>' : 'foo'}
   call assert_equal('foo', x['>'])
 endfunc
+
+function! Test_lambda_capture_by_value()
+  let v = 1
+  let l:F = {x -> x + v}
+  let v = 2
+  call assert_equal(11, l:F(10))
+endfunction
+
+function! Test_lambda_side_effect()
+  function! s:update_and_return(arr)
+    let a:arr[1] = 5
+    return a:arr
+  endfunction
+
+  function! s:foo(arr)
+    return {-> s:update_and_return(a:arr)}
+  endfunction
+
+  let arr = [3,2,1]
+  call assert_equal([3, 5, 1], s:foo(arr)())
+endfunction
+
+function! Test_lambda_refer_local_variable_from_other_scope()
+  function! s:foo(X)
+    return a:X() " refer l:x in s:bar()
+  endfunction
+
+  function! s:bar()
+    let x = 123
+    return s:foo({-> x})
+  endfunction
+
+  call assert_equal(123, s:bar())
+endfunction
+
+function! Test_lambda_do_not_share_local_variable()
+  function! s:define_funcs()
+    let l:One = {-> split(execute("let a = 'abc' | echo a"))[0]}
+    let l:Two = {-> exists("a") ? a : "no"}
+    return [l:One, l:Two]
+  endfunction
+
+  let l:F = s:define_funcs()
+
+  call assert_equal('no', l:F[1]())
+  call assert_equal('abc', l:F[0]())
+  call assert_equal('no', l:F[1]())
+endfunction
+
+function! Test_lambda_closure()
+  function! s:foo()
+    let x = [0]
+    return {-> [execute("let x[0] += 1"), x[0]][-1]}
+  endfunction
+
+  let l:F = s:foo()
+  call test_garbagecollect_now()
+  call assert_equal(1, l:F())
+  call assert_equal(2, l:F())
+  call assert_equal(3, l:F())
+  call assert_equal(4, l:F())
+endfunction
+
+function! Test_lambda_with_a_var()
+  function! s:foo()
+    let x = 2
+    return {... -> a:000 + [x]}
+  endfunction
+  function! s:bar()
+    return s:foo()(1)
+  endfunction
+
+  call assert_equal([1, 2], s:bar())
+endfunction
+
+function! Test_lambda_call_lambda_from_lambda()
+  function! s:foo(x)
+    let l:F1 = {-> {-> a:x}}
+    return {-> l:F1()}
+  endfunction
+
+  let l:F = s:foo(1)
+  call assert_equal(1, l:F()())
+endfunction
+
+function! Test_lambda_combination()
+  call assert_equal(2, {x -> {x -> x}}(1)(2))
+  call assert_equal(10, {y -> {x -> x(y)(10)}({y -> y})}({z -> z}))
+  call assert_equal(5.0, {x -> {y -> x / y}}(10)(2.0))
+
+  call assert_equal(6, {x -> {f -> f(x)}}(3)({x -> x * 2}))
+  call assert_equal(6, {f -> {x -> f(x)}}({x -> x * 2})(3))
+
+  " Z combinator
+  let Z = {f -> {x -> f({y -> x(x)(y)})}({x -> f({y -> x(x)(y)})})}
+  let Fact = {f -> {x -> x == 0 ? 1 : x * f(x - 1)}}
+  call assert_equal(120, Z(Fact)(5))
+endfunction
diff --git a/src/userfunc.c b/src/userfunc.c
--- a/src/userfunc.c
+++ b/src/userfunc.c
@@ -47,6 +47,8 @@ struct ufunc
     scid_T	uf_script_ID;	/* ID of script where function was defined,
 				   used for s: variables */
     int		uf_refcount;	/* for numbered function: reference count */
+    dict_T	*uf_lvars;	/* for lambda: outer scope variables */
+    dict_T	*uf_avars;	/* for lambda: outer scope arguments */
     char_u	uf_name[1];	/* name of function (actually longer); can
 				   start with <SNR>123_ (<SNR> is K_SPECIAL
 				   KS_EXTRA KE_SNR) */
@@ -266,6 +268,9 @@ get_lambda_tv(char_u **arg, typval_T *re
     char_u	*start = skipwhite(*arg + 1);
     char_u	*s, *e;
     static int	lambda_no = 0;
+    dict_T	*old_eval_lvars = eval_lvars_dict;
+    dict_T	*old_eval_avars = eval_avars_dict;
+    dict_T	*d = NULL;
 
     ga_init(&newargs);
     ga_init(&newlines);
@@ -281,6 +286,20 @@ get_lambda_tv(char_u **arg, typval_T *re
     if (ret == FAIL || **arg != '>')
 	goto errret;
 
+    if (evaluate)
+    {
+	/* Don't use dict_alloc() here.  Avoid garbage collection.
+	 * These dicts have the same lifetime with "fp". */
+	d = (dict_T *)alloc(sizeof(dict_T) * 2);
+	if (d == NULL)
+	    goto errret;
+
+	eval_lvars_dict = d;
+	eval_avars_dict = d + 1;
+	init_var_dict(eval_lvars_dict, NULL, 0);
+	init_var_dict(eval_avars_dict, NULL, 0);
+    }
+
     /* Get the start and the end of the expression. */
     *arg = skipwhite(*arg + 1);
     s = *arg;
@@ -298,18 +317,17 @@ get_lambda_tv(char_u **arg, typval_T *re
 	int	len;
 	char_u	*p;
 
-	fp = (ufunc_T *)alloc((unsigned)(sizeof(ufunc_T) + 20));
+	sprintf((char*)name, "<lambda>%d", ++lambda_no);
+
+	fp = (ufunc_T *)alloc((unsigned)(sizeof(ufunc_T) + STRLEN(name)));
 	if (fp == NULL)
 	    goto errret;
 
-	sprintf((char*)name, "<lambda>%d", ++lambda_no);
-
 	ga_init2(&newlines, (int)sizeof(char_u *), 1);
 	if (ga_grow(&newlines, 1) == FAIL)
 	    goto errret;
 
-	/* Add "return " before the expression.
-	 * TODO: Support multiple expressions.  */
+	/* Add "return " before the expression. */
 	len = 7 + e - s + 1;
 	p = (char_u *)alloc(len);
 	if (p == NULL)
@@ -324,6 +342,8 @@ get_lambda_tv(char_u **arg, typval_T *re
 	hash_add(&func_hashtab, UF2HIKEY(fp));
 	fp->uf_args = newargs;
 	fp->uf_lines = newlines;
+	fp->uf_lvars = eval_lvars_dict;
+	fp->uf_avars = eval_avars_dict;
 
 #ifdef FEAT_PROFILE
 	fp->uf_tml_count = NULL;
@@ -342,14 +362,26 @@ get_lambda_tv(char_u **arg, typval_T *re
 	rettv->v_type = VAR_FUNC;
     }
     else
+    {
 	ga_clear_strings(&newargs);
-
+    }
+
+    eval_lvars_dict = old_eval_lvars;
+    eval_avars_dict = old_eval_avars;
     return OK;
 
 errret:
     ga_clear_strings(&newargs);
     ga_clear_strings(&newlines);
     vim_free(fp);
+    if (d != NULL)
+    {
+	vars_clear(&d[0].dv_hashtab);
+	vars_clear(&d[1].dv_hashtab);
+	vim_free(d);
+    }
+    eval_lvars_dict = old_eval_lvars;
+    eval_avars_dict = old_eval_avars;
     return FAIL;
 }
 
@@ -640,6 +672,48 @@ free_funccal(
     vim_free(fc);
 }
 
+    static void
+copy_vars(dict_T *to, dict_T *from, int copyitem)
+{
+    dictitem_T	*di, *org;
+    int		todo;
+    hashitem_T	*hi;
+
+    if (from == NULL)
+	return;
+
+    todo = (int)from->dv_hashtab.ht_used;
+    for (hi = from->dv_hashtab.ht_array; todo > 0; ++hi)
+    {
+	if (!HASHITEM_EMPTY(hi))
+	{
+	    --todo;
+
+	    /* Skip if there is already a variable with the same name. */
+	    if (!HASHITEM_EMPTY(hash_find(&to->dv_hashtab, hi->hi_key)))
+		continue;
+
+	    org = HI2DI(hi);
+	    di = dictitem_alloc(org->di_key);
+	    if (di == NULL)
+		break;
+	    di->di_flags |= DI_FLAGS_RO | DI_FLAGS_FIX;
+	    if (copyitem)
+		copy_tv(&org->di_tv, &di->di_tv);
+	    else
+		di->di_tv = org->di_tv;
+
+	    if (dict_add(to, di) == FAIL)
+	    {
+		if (copyitem)
+		    clear_tv(&di->di_tv);
+		vim_free(di);
+		break;
+	    }
+	}
+    }
+}
+
 /*
  * Call a user function.
  */
@@ -758,7 +832,6 @@ call_user_func(
     for (i = 0; i < argcount; ++i)
     {
 	int	    addlocal = FALSE;
-	dictitem_T  *v2;
 
 	ai = i - fp->uf_args.ga_len;
 	if (ai < 0)
@@ -778,9 +851,6 @@ call_user_func(
 	{
 	    v = &fc->fixvar[fixvar_idx++].var;
 	    v->di_flags = DI_FLAGS_RO | DI_FLAGS_FIX;
-
-	    if (addlocal)
-		v2 = v;
 	}
 	else
 	{
@@ -789,36 +859,23 @@ call_user_func(
 	    if (v == NULL)
 		break;
 	    v->di_flags = DI_FLAGS_RO | DI_FLAGS_FIX | DI_FLAGS_ALLOC;
-
-	    if (addlocal)
-	    {
-		v2 = (dictitem_T *)alloc((unsigned)(sizeof(dictitem_T)
-							     + STRLEN(name)));
-		if (v2 == NULL)
-		{
-		    vim_free(v);
-		    break;
-		}
-		v2->di_flags = DI_FLAGS_RO | DI_FLAGS_FIX | DI_FLAGS_ALLOC;
-	    }
 	}
 	STRCPY(v->di_key, name);
-	hash_add(&fc->l_avars.dv_hashtab, DI2HIKEY(v));
 
 	/* Note: the values are copied directly to avoid alloc/free.
 	 * "argvars" must have VAR_FIXED for v_lock. */
 	v->di_tv = argvars[i];
 	v->di_tv.v_lock = VAR_FIXED;
 
-	/* Named arguments can be accessed without the "a:" prefix in lambda
-	 * expressions.  Add to the l: dict. */
 	if (addlocal)
 	{
-	    STRCPY(v2->di_key, name);
-	    copy_tv(&v->di_tv, &v2->di_tv);
-	    v2->di_tv.v_lock = VAR_FIXED;
-	    hash_add(&fc->l_vars.dv_hashtab, DI2HIKEY(v2));
+	    /* Named arguments should be accessed without the "a:" prefix in
+	     * lambda expressions.  Add to the l: dict. */
+	    copy_tv(&v->di_tv, &v->di_tv);
+	    hash_add(&fc->l_vars.dv_hashtab, DI2HIKEY(v));
 	}
+	else
+	    hash_add(&fc->l_avars.dv_hashtab, DI2HIKEY(v));
 
 	if (ai >= 0 && ai < MAX_FUNC_ARGS)
 	{
@@ -827,6 +884,8 @@ call_user_func(
 	    fc->l_listitems[ai].li_tv.v_lock = VAR_FIXED;
 	}
     }
+    copy_vars(&fc->l_vars, fp->uf_lvars, TRUE);
+    copy_vars(&fc->l_avars, fp->uf_avars, FALSE);
 
     /* Don't redraw while executing the function. */
     ++RedrawingDisabled;
@@ -1059,6 +1118,12 @@ func_free(ufunc_T *fp)
     /* clear this function */
     ga_clear_strings(&(fp->uf_args));
     ga_clear_strings(&(fp->uf_lines));
+    if (fp->uf_lvars != NULL)
+    {
+	vars_clear(&fp->uf_lvars->dv_hashtab);
+	vars_clear(&fp->uf_avars->dv_hashtab);
+	vim_free(fp->uf_lvars);
+    }
 #ifdef FEAT_PROFILE
     vim_free(fp->uf_tml_count);
     vim_free(fp->uf_tml_total);
@@ -2206,6 +2271,8 @@ ex_function(exarg_T *eap)
     }
     fp->uf_args = newargs;
     fp->uf_lines = newlines;
+    fp->uf_lvars = NULL;
+    fp->uf_avars = NULL;
 #ifdef FEAT_PROFILE
     fp->uf_tml_count = NULL;
     fp->uf_tml_total = NULL;
@@ -2693,6 +2760,39 @@ func_ref(char_u *name)
 }
 
 /*
+ * Mark all lists and dicts referenced through function "name" with "copyID".
+ * "list_stack" is used to add lists to be marked.  Can be NULL.
+ * "ht_stack" is used to add hashtabs to be marked.  Can be NULL.
+ *
+ * Returns TRUE if setting references failed somehow.
+ */
+    int
+set_ref_in_func(
+    char_u	    *name,
+    int		    copyID,
+    ht_stack_T	    **ht_stack,
+    list_stack_T    **list_stack)
+{
+    ufunc_T	*fp;
+    int		abort = FALSE;
+
+    if (name == NULL)
+	return abort;
+
+    fp = find_func(name);
+    if (fp != NULL)
+    {
+	if (fp->uf_lvars != NULL)
+	    abort = abort || set_ref_in_ht(&fp->uf_lvars->dv_hashtab,
+							copyID, list_stack);
+	if (fp->uf_avars != NULL)
+	    abort = abort || set_ref_in_ht(&fp->uf_avars->dv_hashtab,
+							copyID, list_stack);
+    }
+    return abort;
+}
+
+/*
  * Return TRUE if items in "fc" do not have "copyID".  That means they are not
  * referenced from anywhere that is in use.
  */
