# HG changeset patch
# Parent  42bec9aba9a8c58e269d7f531d22aea20489d556

diff --git a/src/dict.c b/src/dict.c
--- a/src/dict.c
+++ b/src/dict.c
@@ -109,7 +109,7 @@ dict_free_dict(dict_T *d)
     vim_free(d);
 }
 
-    static void
+    void
 dict_free(dict_T *d)
 {
     if (!in_free_unref_items)
@@ -188,7 +188,7 @@ dictitem_alloc(char_u *key)
 /*
  * Make a copy of a Dictionary item.
  */
-    static dictitem_T *
+    dictitem_T *
 dictitem_copy(dictitem_T *org)
 {
     dictitem_T *di;
diff --git a/src/eval.c b/src/eval.c
--- a/src/eval.c
+++ b/src/eval.c
@@ -237,6 +237,7 @@ static int get_env_tv(char_u **arg, typv
 
 static int get_env_len(char_u **arg);
 static char_u * make_expanded_name(char_u *in_start, char_u *expr_start, char_u *expr_end, char_u *in_end);
+static void check_lvars(char_u *name, int len);
 static typval_T *alloc_string_tv(char_u *string);
 static hashtab_T *find_var_ht(char_u *name, char_u **varname);
 static void delete_var(hashtab_T *ht, hashitem_T *hi);
@@ -4363,7 +4364,10 @@ eval7(
 	    else if (evaluate)
 		ret = get_var_tv(s, len, rettv, NULL, TRUE, FALSE);
 	    else
+	    {
+		check_lvars(s, len);
 		ret = OK;
+	    }
 	}
 	vim_free(alias);
     }
@@ -6791,6 +6795,42 @@ get_var_tv(
 }
 
 /*
+ * Check if variable "name" is a local variable.
+ * If it is a local variable, store it to the "eval_lvars_dict" dict.
+ */
+    static void
+check_lvars(
+    char_u	*name,
+    int		len)		/* length of "name" */
+{
+    int		cc;
+    char_u	*varname;
+    hashtab_T	*ht;
+
+    /* truncate the name, so that we can use strcmp() */
+    cc = name[len];
+    name[len] = NUL;
+
+    ht = find_var_ht(name, &varname);
+    if ((ht == get_funccal_local_ht()) && (eval_lvars_dict != NULL))
+    {
+	dictitem_T  *di, *di2;
+
+	di = find_var_in_ht(ht, *name, varname, FALSE);
+	if (di != NULL)
+	{
+	    di2 = dictitem_copy(di);
+	    if (di2 != NULL)
+		dict_add(eval_lvars_dict, di2);
+	}
+    }
+
+    name[len] = cc;
+
+    return;
+}
+
+/*
  * Handle expr[expr], expr[expr:expr] subscript and .name lookup.
  * Also handle function call with Funcref variable: func(expr)
  * Can all be combined: dict.func(expr)[idx]['func'](expr)
diff --git a/src/globals.h b/src/globals.h
--- a/src/globals.h
+++ b/src/globals.h
@@ -1654,6 +1654,9 @@ EXTERN time_T time_for_testing INIT(= 0)
 
 /* Abort conversion to string after a recursion error. */
 EXTERN int  did_echo_string_emsg INIT(= FALSE);
+
+/* Dict for local variables used in a lambda expression. */
+EXTERN dict_T *eval_lvars_dict INIT(= NULL);
 #endif
 
 /*
diff --git a/src/proto/dict.pro b/src/proto/dict.pro
--- a/src/proto/dict.pro
+++ b/src/proto/dict.pro
@@ -1,10 +1,12 @@
 /* dict.c */
 dict_T *dict_alloc(void);
 int rettv_dict_alloc(typval_T *rettv);
+void dict_free(dict_T *d);
 void dict_unref(dict_T *d);
 int dict_free_nonref(int copyID);
 void dict_free_items(int copyID);
 dictitem_T *dictitem_alloc(char_u *key);
+dictitem_T *dictitem_copy(dictitem_T *org);
 void dictitem_remove(dict_T *dict, dictitem_T *item);
 void dictitem_free(dictitem_T *item);
 dict_T *dict_copy(dict_T *orig, int deep, int copyID);
diff --git a/src/userfunc.c b/src/userfunc.c
--- a/src/userfunc.c
+++ b/src/userfunc.c
@@ -47,6 +47,7 @@ struct ufunc
     scid_T	uf_script_ID;	/* ID of script where function was defined,
 				   used for s: variables */
     int		uf_refcount;	/* for numbered function: reference count */
+    dict_T	*uf_lvars;	/* for lambda: outer scope variables */
     char_u	uf_name[1];	/* name of function (actually longer); can
 				   start with <SNR>123_ (<SNR> is K_SPECIAL
 				   KS_EXTRA KE_SNR) */
@@ -260,6 +261,8 @@ get_lambda_tv(char_u **arg, typval_T *re
     char_u	*start = skipwhite(*arg + 1);
     char_u	*s, *e;
     static int	lambda_no = 0;
+    dict_T	*old_eval_lvars = eval_lvars_dict;
+    dict_T	*d = NULL;
 
     ga_init(&newargs);
     ga_init(&newlines);
@@ -275,6 +278,20 @@ get_lambda_tv(char_u **arg, typval_T *re
     if (ret == FAIL || **arg != '>')
 	goto errret;
 
+    /* Prepare local vars dict.  Not using dict_alloc() here, because
+     * don't want to grabage collected the dict. */
+    d = (dict_T *)alloc(sizeof(dict_T));
+    if (d == NULL)
+	goto errret;
+    d->dv_used_next = NULL;
+    d->dv_used_prev = NULL;
+    hash_init(&d->dv_hashtab);
+    d->dv_lock = 0;
+    d->dv_scope = 0;
+    d->dv_refcount = DO_NOT_FREE_CNT;
+    d->dv_copyID = 0;
+    eval_lvars_dict = d;
+
     /* Get the start and the end of the expression. */
     *arg = skipwhite(*arg + 1);
     s = *arg;
@@ -318,6 +335,7 @@ get_lambda_tv(char_u **arg, typval_T *re
 	hash_add(&func_hashtab, UF2HIKEY(fp));
 	fp->uf_args = newargs;
 	fp->uf_lines = newlines;
+	fp->uf_lvars = d;
 
 #ifdef FEAT_PROFILE
 	fp->uf_tml_count = NULL;
@@ -336,14 +354,21 @@ get_lambda_tv(char_u **arg, typval_T *re
 	rettv->v_type = VAR_FUNC;
     }
     else
+    {
 	ga_clear_strings(&newargs);
-
+	dict_free(d);
+    }
+
+    eval_lvars_dict = old_eval_lvars;
     return OK;
 
 errret:
     ga_clear_strings(&newargs);
     ga_clear_strings(&newlines);
     vim_free(fp);
+    if (d != NULL)
+	dict_free(d);
+    eval_lvars_dict = old_eval_lvars;
     return FAIL;
 }
 
@@ -634,6 +659,37 @@ free_funccal(
     vim_free(fc);
 }
 
+    static void
+copy_lvars(dict_T *to, dict_T *from)
+{
+    dictitem_T	*di;
+    int		todo;
+    hashitem_T	*hi;
+
+    if (from == NULL)
+	return;
+
+    todo = (int)from->dv_hashtab.ht_used;
+    for (hi = from->dv_hashtab.ht_array; todo > 0; ++hi)
+    {
+	if (!HASHITEM_EMPTY(hi))
+	{
+	    --todo;
+
+	    if (!HASHITEM_EMPTY(hash_find(&to->dv_hashtab, hi->hi_key)))
+		continue;
+	    di = dictitem_copy(HI2DI(hi));
+	    if (di == NULL)
+		break;
+	    if (dict_add(to, di) == FAIL)
+	    {
+		dictitem_free(di);
+		break;
+	    }
+	}
+    }
+}
+
 /*
  * Call a user function.
  */
@@ -821,6 +877,7 @@ call_user_func(
 	    fc->l_listitems[ai].li_tv.v_lock = VAR_FIXED;
 	}
     }
+    copy_lvars(&fc->l_vars, fp->uf_lvars);
 
     /* Don't redraw while executing the function. */
     ++RedrawingDisabled;
@@ -1053,6 +1110,8 @@ func_free(ufunc_T *fp)
     /* clear this function */
     ga_clear_strings(&(fp->uf_args));
     ga_clear_strings(&(fp->uf_lines));
+    if (fp->uf_lvars != NULL)
+	dict_free(fp->uf_lvars);
 #ifdef FEAT_PROFILE
     vim_free(fp->uf_tml_count);
     vim_free(fp->uf_tml_total);
@@ -2200,6 +2259,7 @@ ex_function(exarg_T *eap)
     }
     fp->uf_args = newargs;
     fp->uf_lines = newlines;
+    fp->uf_lvars = NULL;
 #ifdef FEAT_PROFILE
     fp->uf_tml_count = NULL;
     fp->uf_tml_total = NULL;
