# HG changeset patch
# Parent  bf2d6782347280d403922136e5ce0258ffec29f1

diff --git a/src/eval.c b/src/eval.c
--- a/src/eval.c
+++ b/src/eval.c
@@ -6828,6 +6828,7 @@ check_vars(
     {
 	if (find_var(name, NULL, TRUE) != NULL)
 	{
+	    TRACE("lvar: %s", varname);
 	    *eval_lavars_used = TRUE;
 	}
     }
@@ -6835,6 +6836,7 @@ check_vars(
     {
 	if (find_var(name, NULL, TRUE) != NULL)
 	{
+	    TRACE("avar: %s", varname);
 	    *eval_lavars_used = TRUE;
 	}
     }
diff --git a/src/userfunc.c b/src/userfunc.c
--- a/src/userfunc.c
+++ b/src/userfunc.c
@@ -336,6 +336,7 @@ get_lambda_tv(char_u **arg, typval_T *re
 	((char_u **)(newlines.ga_data))[newlines.ga_len++] = p;
 	STRCPY(p, "return ");
 	vim_strncpy(p + 7, s, e - s);
+	TRACE("lambda: la:%d, name: %s, expr: %s", eval_lavars, name, p + 7);
 
 	fp->uf_refcount = 1;
 	STRCPY(fp->uf_name, name);
@@ -352,6 +353,7 @@ get_lambda_tv(char_u **arg, typval_T *re
 	    ((ufunc_T **)current_funccal->fc_funcs.ga_data)
 				    [current_funccal->fc_funcs.ga_len++] = fp;
 	    func_ref(current_funccal->func->uf_name);
+	    TRACE("name:%s, fc_refcount:%d", current_funccal->func->uf_name, current_funccal->fc_refcount);
 	}
 	else
 	    fp->uf_scoped = NULL;
@@ -678,6 +680,7 @@ free_funccal(
 	for (li = fc->l_varlist.lv_first; li != NULL; li = li->li_next)
 	    clear_tv(&li->li_tv);
 
+    TRACE("free_funccal: %s", fc->func->uf_name);
     vim_free(fc);
 }
 
@@ -1059,6 +1062,8 @@ call_user_func(
 	fc->caller = previous_funccal;
 	previous_funccal = fc;
 
+	TRACE("Add to previous_funccal: %s", fc->func->uf_name);
+
 	/* Make a copy of the a: variables, since we didn't do that above. */
 	todo = (int)fc->l_avars.dv_hashtab.ht_used;
 	for (hi = fc->l_avars.dv_hashtab.ht_array; todo > 0; ++hi)
@@ -1085,6 +1090,7 @@ funccal_unref(funccall_T *fc)
     if (fc == NULL)
 	return;
 
+    TRACE("funccal_unref'ing: %s, cnt:%d", fc->func->uf_name, fc->fc_refcount);
     func_unref(fc->func->uf_name);
     if (--fc->fc_refcount > 0)
 	return;
@@ -1130,6 +1136,8 @@ func_free(ufunc_T *fp)
 
     funccal_unref(fp->uf_scoped);
 
+    TRACE("func_free: %s", fp->uf_name);
+
     vim_free(fp);
 }
 
@@ -2297,6 +2305,7 @@ ex_function(exarg_T *eap)
 	((ufunc_T **)current_funccal->fc_funcs.ga_data)
 				[current_funccal->fc_funcs.ga_len++] = fp;
 	func_ref(current_funccal->func->uf_name);
+	TRACE("name:%s, fc_refcount:%d", current_funccal->func->uf_name, current_funccal->fc_refcount);
     }
     else
 	fp->uf_scoped = NULL;
@@ -2749,6 +2758,10 @@ func_unref(char_u *name)
 	/* fail silently, when lambda function isn't found. */
 	fp = find_func(name);
     }
+    if (fp != NULL)
+	TRACE("func_unref'ing: %s, cnt:%d", name, fp->uf_refcount);
+    else
+	TRACE("func_unref: %s (not unref'ing)", name);
     if (fp != NULL && --fp->uf_refcount <= 0)
     {
 	/* Only delete it when it's not being used.  Otherwise it's done
@@ -2775,6 +2788,7 @@ func_ref(char_u *name)
 	    EMSG2(_(e_intern2), "func_ref()");
 	else
 	    ++fp->uf_refcount;
+	TRACE("func_ref: %s, cnt:%d", name, fp ? fp->uf_refcount : -1);
     }
     else if (STRNCMP(name, "<lambda>", 8) == 0)
     {
@@ -2782,6 +2796,7 @@ func_ref(char_u *name)
 	fp = find_func(name);
 	if (fp != NULL)
 	    ++fp->uf_refcount;
+	TRACE("func_ref: %s, cnt:%d", name, fp ? fp->uf_refcount : -1);
     }
 }
 
@@ -2792,6 +2807,8 @@ func_ref(char_u *name)
     static int
 can_free_funccal(funccall_T *fc, int copyID)
 {
+    TRACE("can_free_funccal: %s: copyID:%d lvl:%d ld:%d ad:%d fc:%d, ref:%d", fc->func ? fc->func->uf_name: NULL, copyID, fc->l_varlist.lv_copyID, fc->l_vars.dv_copyID, fc->l_avars.dv_copyID, fc->fc_copyID, fc->fc_refcount);
+
     return (fc->l_varlist.lv_copyID != copyID
 	    && fc->l_vars.dv_copyID != copyID
 	    && fc->l_avars.dv_copyID != copyID
@@ -3402,10 +3419,12 @@ free_unref_funccal(int copyID, int testi
     funccall_T	*fc, **pfc;
     ufunc_T	*fp;
 
+    TRACE("free_unref_funccal()");
     for (pfc = &previous_funccal; *pfc != NULL; )
     {
 	if (can_free_funccal(*pfc, copyID))
 	{
+	    TRACE("free_unref_funccal: %s", (*pfc)->func->uf_name);
 	    fc = *pfc;
 	    *pfc = fc->caller;
 	    fp = fc->func;
@@ -3585,9 +3604,11 @@ set_ref_in_previous_funccal(int copyID)
     int		abort = FALSE;
     funccall_T	*fc;
 
+    TRACE("set_ref_in_previous_funccal()");
     for (fc = previous_funccal; fc != NULL; fc = fc->caller)
     {
 	fc->fc_copyID = copyID + 1;
+	TRACE("set_ref_in_previous_funccal: %s", fc->func->uf_name);
 	abort = abort || set_ref_in_ht(&fc->l_vars.dv_hashtab, copyID + 1,
 									NULL);
 	abort = abort || set_ref_in_ht(&fc->l_avars.dv_hashtab, copyID + 1,
@@ -3605,9 +3626,11 @@ set_ref_in_call_stack(int copyID)
     int		abort = FALSE;
     funccall_T	*fc;
 
+    TRACE("set_ref_in_call_stack()");
     for (fc = current_funccal; fc != NULL; fc = fc->caller)
     {
 	fc->fc_copyID = copyID;
+	TRACE("set_ref_in_call_stack: %s", fc->func->uf_name);
 	abort = abort || set_ref_in_ht(&fc->l_vars.dv_hashtab, copyID, NULL);
 	abort = abort || set_ref_in_ht(&fc->l_avars.dv_hashtab, copyID, NULL);
     }
@@ -3646,6 +3669,7 @@ set_ref_in_func(char_u *name, int copyID
     char_u	*tofree = NULL;
     char_u	*fname;
 
+    TRACE("set_ref_in_func(): %s", name);
     if (name == NULL)
 	return FALSE;
 
@@ -3653,11 +3677,13 @@ set_ref_in_func(char_u *name, int copyID
     fp = find_func(fname);
     if (fp != NULL)
     {
+	TRACE("set_ref_in_func: %s", fp->uf_name);
 	for (fc = fp->uf_scoped; fc != NULL; fc = fc->func->uf_scoped)
 	{
 	    if (fc->fc_copyID != copyID)
 	    {
 		fc->fc_copyID = copyID;
+		TRACE("set_ref_in_func: %s", fc->func->uf_name);
 		set_ref_in_ht(&fc->l_vars.dv_hashtab, copyID, NULL);
 		set_ref_in_ht(&fc->l_avars.dv_hashtab, copyID, NULL);
 	    }
