# HG changeset patch
# Parent  e645a4c5020356d0621ea999c9e1083c2e92943b

diff --git a/runtime/macros/matchit.vim b/runtime/macros/matchit.vim
--- a/runtime/macros/matchit.vim
+++ b/runtime/macros/matchit.vim
@@ -303,7 +303,7 @@ fun! s:CleanUp(options, mode, startline,
       let regexp = s:Wholematch(matchline, a:1, currcol-1)
       let endcol = matchend(matchline, regexp)
       if endcol > currcol  " This is NOT off by one!
-	execute "normal!" . (endcol - currcol) . "l"
+	call cursor(0, endcol)
       endif
     endif " a:0
   endif " a:mode != "o" && etc.
