# HG changeset patch
# Parent  03d5d7ec99df5fc5f530015051b63ba68dffcf98

diff --git a/src/GvimExt/Make_ming.mak b/src/GvimExt/Make_ming.mak
--- a/src/GvimExt/Make_ming.mak
+++ b/src/GvimExt/Make_ming.mak
@@ -29,7 +29,7 @@ LDFLAGS += -static-libgcc -static-libstd
 endif
 
 ifeq ($(CROSS),yes)
-DEL = rm
+DEL = rm -f
 ifeq ($(MINGWOLD),yes)
 CXXFLAGS := -O2 -fvtable-thunks
 else
@@ -38,7 +38,7 @@ endif
 else
 CXXFLAGS := -O2
 ifneq (sh.exe, $(SHELL))
-DEL = rm
+DEL = rm -f
 else
 DEL = del
 endif
diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -124,13 +124,13 @@ ifndef CTAGS
 CTAGS = ctags -I INIT+,INIT2+,INIT3+,INIT4+,INIT5+ --fields=+S
 endif
 
+
 # Link against the shared version of libstdc++ by default.  Set
 # STATIC_STDCPLUS to "yes" to link against static version instead.
 ifndef STATIC_STDCPLUS
 STATIC_STDCPLUS=no
 endif
 
-
 # Link against the shared version of libwinpthread by default.  Set
 # STATIC_WINPTHREAD to "yes" to link against static version instead.
 ifndef STATIC_WINPTHREAD
@@ -140,6 +140,13 @@ endif
 # This is used when STATIC_STDCPLUS=yes.
 HAS_GCC_EH=yes
 
+# Reduce the size of the executables by using the --gc-sections linker
+# option.  Set USE_GC_SECTIONS to "no" if you see any issues with this.
+ifndef USE_GC_SECTIONS
+USE_GC_SECTIONS=yes
+endif
+
+
 # If the user doesn't want gettext, undefine it.
 ifeq (no, $(GETTEXT))
 GETTEXT=
@@ -179,7 +186,7 @@ ifeq ($(CROSS),yes)
  ifndef CROSS_COMPILE
 CROSS_COMPILE = i586-pc-mingw32msvc-
  endif
-DEL = rm
+DEL = rm -f
 MKDIR = mkdir -p
 DIRSLASH = /
 else
@@ -207,7 +214,7 @@ CROSS_COMPILE =
 # In this case, unix-like commands can be used.
 #
  ifneq (sh.exe, $(SHELL))
-DEL = rm
+DEL = rm -f
 MKDIR = mkdir -p
 DIRSLASH = /
  else
@@ -1095,6 +1102,16 @@ LIB += -lgcc_eh
 LIB += -Wl,-Bstatic -lwinpthread -Wl,-Bdynamic
 endif
 
+# To reduce the file size
+ifeq (yes, $(USE_GC_SECTIONS))
+CFLAGS += -ffunction-sections -fno-asynchronous-unwind-tables
+CXXFLAGS += -fasynchronous-unwind-tables
+LFLAGS += -Wl,--gc-sections
+ ifeq (yes, $(VIMDLL))
+EXELFLAGS += -Wl,--gc-sections
+ endif
+endif
+
 ifeq (yes, $(MAP))
 LFLAGS += -Wl,-Map=$(TARGET).map
 endif
diff --git a/src/xxd/Make_ming.mak b/src/xxd/Make_ming.mak
--- a/src/xxd/Make_ming.mak
+++ b/src/xxd/Make_ming.mak
@@ -16,7 +16,7 @@ CC = gcc
 CFLAGS = -O2 -Wall -DWIN32 $(DEFINES)
 
 ifneq (sh.exe, $(SHELL))
-DEL = rm
+DEL = rm -f
 else
 DEL = del
 endif
