# HG changeset patch
# Parent  87cada7913f7bbe806b648b66c4fa61be396aa61

diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -513,8 +513,8 @@ endif
 #>>>>> end of choices
 ###########################################################################
 
-CFLAGS = -I. -Iproto $(DEFINES) -pipe -march=$(ARCH) -Wall
-CXXFLAGS = -std=gnu++11
+CFLAGS = -I. -Iproto $(DEFINES) -pipe -march=$(ARCH) -Wall -ffunction-sections -fno-asynchronous-unwind-tables
+CXXFLAGS = -fasynchronous-unwind-tables -std=gnu++11
 # This used to have --preprocessor, but it's no longer supported
 WINDRES_FLAGS =
 EXTRA_LIBS =
@@ -1042,6 +1042,8 @@ LIB += -lgcc_eh
 LIB += -Wl,-Bstatic -lwinpthread -Wl,-Bdynamic
 endif
 
+LFLAGS += -Wl,--gc-sections
+
 ifeq (yes, $(MAP))
 LFLAGS += -Wl,-Map=$(TARGET).map
 endif
