# HG changeset patch
# Parent  c529374cdfdced0cc6cc24148157bd23f326efd1

diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -525,10 +525,10 @@ endif
 #>>>>> end of choices
 ###########################################################################
 
-CFLAGS = -I. -Iproto $(DEFINES) -pipe -march=$(ARCH) -Wall
+CFLAGS = -I. -Iproto $(DEFINES) -pipe -march=$(ARCH) -Wall -ffunction-sections -fno-asynchronous-unwind-tables
 # To get additional compiler warnings
 #CFLAGS += -Wextra -pedantic
-CXXFLAGS = -std=gnu++11
+CXXFLAGS = -fasynchronous-unwind-tables -std=gnu++11
 # This used to have --preprocessor, but it's no longer supported
 WINDRES_FLAGS =
 EXTRA_LIBS =
@@ -1095,6 +1095,8 @@ LIB += -lgcc_eh
 LIB += -Wl,-Bstatic -lwinpthread -Wl,-Bdynamic
 endif
 
+LFLAGS += -Wl,--gc-sections
+
 ifeq (yes, $(MAP))
 LFLAGS += -Wl,-Map=$(TARGET).map
 endif
