# HG changeset patch
# Parent 7f1899e893992cf6fab414285ece04dda8dc8d74
diff --git a/src/Make_cyg.mak b/src/Make_cyg.mak
--- a/src/Make_cyg.mak
+++ b/src/Make_cyg.mak
@@ -365,7 +365,7 @@ endif
 ##############################
 ifdef LUA
 DEFINES += -DFEAT_LUA
-INCLUDES += -I$(LUA)/include
+INCLUDES += -I$(LUA)/include -I$(LUA)
 EXTRA_OBJS += $(OUTDIR)/if_lua.o
 
 ifndef DYNAMIC_LUA
diff --git a/src/Make_ming.mak b/src/Make_ming.mak
--- a/src/Make_ming.mak
+++ b/src/Make_ming.mak
@@ -402,7 +402,7 @@ endif
 endif
 
 ifdef LUA
-CFLAGS += -I$(LUA)/include -DFEAT_LUA
+CFLAGS += -I$(LUA)/include -I$(LUA) -DFEAT_LUA
 ifeq (yes, $(DYNAMIC_LUA))
 CFLAGS += -DDYNAMIC_LUA -DDYNAMIC_LUA_DLL=\"lua$(LUA_VER).dll\"
 endif
