# HG changeset patch
# Parent  29f16c55c42efada128f0dff20be2f16b5812d6e

diff --git a/src/INSTALLpc.txt b/src/INSTALLpc.txt
--- a/src/INSTALLpc.txt
+++ b/src/INSTALLpc.txt
@@ -164,6 +164,11 @@ options:
   msvc2015
 For x64 builds run this with the "x86_amd64" option:
   msvc2015 x86_amd64
+This enables x86_x64 cross compiler. This works on any editions including
+Express edition.
+If you use Community (or Professional) edition, you can enable the x64 native
+compiler by using the "x64" option:
+  msvc2015 x64
 
 The following Visual C++ team blog can serve as a reference page:
     http://blogs.msdn.com/b/vcblog/archive/2012/10/08/windows-xp-targeting-with-c-in-visual-studio-2012.aspx
diff --git a/src/msvc2015.bat b/src/msvc2015.bat
--- a/src/msvc2015.bat
+++ b/src/msvc2015.bat
@@ -7,6 +7,10 @@ rem   For x86 builds run this without op
 rem     msvc2015
 rem   For x64 builds run this with "x86_amd64" option:
 rem     msvc2015 x86_amd64
+rem   This works on any editions including Express edition.
+rem   If you use Community (or Professional) edition, you can also use "x64"
+rem   option:
+rem     msvc2015 x64
 @echo on
 
 call "%VS140COMNTOOLS%..\..\VC\vcvarsall.bat" %*
@@ -23,7 +27,7 @@ if not exist "%WinSdk71%" (
 )
 
 set INCLUDE=%WinSdk71%\Include;%INCLUDE%
-if "%Platform%"=="x64" (
+if /i "%Platform%"=="x64" (
 	set "LIB=%WinSdk71%\Lib\x64;%LIB%"
 	set SUBSYSTEM_VER=5.02
 ) else (
