# HG changeset patch
# Parent  a8086327b8b390d1e3a48cee96201c589eb9d7eb

diff --git a/src/msvc-latest.bat b/src/msvc-latest.bat
--- a/src/msvc-latest.bat
+++ b/src/msvc-latest.bat
@@ -61,6 +61,11 @@ if exist "%InstallDir%\VC\Auxiliary\Buil
 )
 
 :done
+if defined VCToolsVersion (
+	if "%VSVEROPT%"=="-latest" (
+		title VCTools %VCToolsVersion:~0,5% %VSCMD_ARG_TGT_ARCH%
+	)
+)
 if "%VSWHERE_SET%"=="yes" (
 	set VSWHERE=
 	set VSWHERE_SET=
diff --git a/src/msvc2015.bat b/src/msvc2015.bat
--- a/src/msvc2015.bat
+++ b/src/msvc2015.bat
@@ -12,4 +12,16 @@ rem   If you use Community (or Professio
 rem   option:
 rem     msvc2015 x64
 
+set Platform=
+if not exist "%VS140COMNTOOLS%..\..\VC\vcvarsall.bat" (
+	echo Error: vcvarsall.bat not found.
+	exit /b 1
+)
 call "%VS140COMNTOOLS%..\..\VC\vcvarsall.bat" %*
+if defined VisualStudioVersion (
+	if defined Platform (
+		title VS 2015 %Platform%
+	) else (
+		title VS 2015 x86
+	)
+)
diff --git a/src/msvc2017.bat b/src/msvc2017.bat
--- a/src/msvc2017.bat
+++ b/src/msvc2017.bat
@@ -11,3 +11,6 @@ rem     msvc2017 x86_amd64
 set "VSVEROPT=-version [15.0^,16.0^)"
 call "%~dp0msvc-latest.bat" %*
 set VSVEROPT=
+if defined VCToolsVersion (
+	title VS 2017 %VSCMD_ARG_TGT_ARCH%
+)
diff --git a/src/msvc2019.bat b/src/msvc2019.bat
--- a/src/msvc2019.bat
+++ b/src/msvc2019.bat
@@ -11,3 +11,6 @@ rem     msvc2019 x64
 set "VSVEROPT=-version [16.0^,17.0^)"
 call "%~dp0msvc-latest.bat" %*
 set VSVEROPT=
+if defined VCToolsVersion (
+	title VS 2019 %VSCMD_ARG_TGT_ARCH%
+)
diff --git a/src/msvc2022.bat b/src/msvc2022.bat
--- a/src/msvc2022.bat
+++ b/src/msvc2022.bat
@@ -11,3 +11,6 @@ rem     msvc2022 x64
 set "VSVEROPT=-version [17.0^,18.0^)"
 call "%~dp0msvc-latest.bat" %*
 set VSVEROPT=
+if defined VCToolsVersion (
+	title VS 2022 %VSCMD_ARG_TGT_ARCH%
+)
