# HG changeset patch
# Parent  2dc5f88e91553fc1bf5ca4b5f971a5ea42a7b730

diff --git a/src/Make_mvc.mak b/src/Make_mvc.mak
--- a/src/Make_mvc.mak
+++ b/src/Make_mvc.mak
@@ -158,14 +158,9 @@
 #	nmake -f Make_mvc.mvc "DEFINES=-DEMACS_TAGS"
 
 RM =		del /f /q
-PS =		powershell.exe
 
-PSFLAGS =	-NoLogo -NoProfile -Command
-
-!IF ![$(PS) $(PSFLAGS) try{Out-File -FilePath '.\major.tmp' -InputObject \
-	\"MAJOR=$$(((Select-String -Pattern 'VIM_VERSION_MAJOR\s+\d{1,2}' \
-	-Path '.\version.h').Line[-2..-1^]-join '').Trim())\"} \
-	catch{exit 1}]
+!IF ![for /f "tokens=2,3" %I in (version.h) do \
+	@if "%I"=="VIM_VERSION_MAJOR" echo MAJOR=%J>.\major.tmp]
 ! INCLUDE .\major.tmp
 ! IF [$(RM) .\major.tmp]
 ! ENDIF
@@ -174,10 +169,8 @@ PSFLAGS =	-NoLogo -NoProfile -Command
 MAJOR =		9
 !ENDIF
 
-!IF ![$(PS) $(PSFLAGS) try{Out-File -FilePath '.\minor.tmp' -InputObject \
-	\"MINOR=$$(((Select-String -Pattern 'VIM_VERSION_MINOR\s+\d{1,2}' \
-	-Path '.\version.h').Line[-2..-1^]-join '').Trim())\"} \
-	catch{exit 1}]
+!IF ![for /f "tokens=2,3" %I in (version.h) do \
+	@if "%I"=="VIM_VERSION_MINOR" echo MINOR=%J>.\minor.tmp]
 ! INCLUDE .\minor.tmp
 ! IF [$(RM) .\minor.tmp]
 ! ENDIF
@@ -186,21 +179,18 @@ MAJOR =		9
 MINOR =		1
 !ENDIF
 
-!IF ![$(PS) $(PSFLAGS) try{Out-File -FilePath '.\patchlvl.tmp' -InputObject \
-	\"PATCHLEVEL=$$([decimal^]((Get-Content -Path '.\version.c' \
-	-TotalCount ((Select-String -Pattern 'static int included_patches' \
-	-Path '.\version.c').LineNumber+3))[-1^]).Trim().TrimEnd(','))\"} \
-	catch{exit 1}]
-! INCLUDE .\patchlvl.tmp
-! IF [$(RM) .\patchlvl.tmp]
+!IF ![cmd /V:ON /C "echo off && set LINE=0&& set FIND=0&& set LL=&& \
+	for /f "tokens=1,3 delims=, " %I in (version.c) do \
+	( set /A LINE+=1 > NUL && set LL=%J&& \
+	if "!LL:~0,-2!"=="included_patches" ( set /A FIND=LINE+3 > NUL ) \
+	else if "!LINE!"=="!FIND!" ( echo PATCHLEVEL=%I>.\patchlvl.tmp ) )"]
+! IF EXIST(.\patchlvl.tmp)
+!  INCLUDE .\patchlvl.tmp
+!  IF [$(RM) .\patchlvl.tmp]
+!  ENDIF
 ! ENDIF
 !ENDIF
 
-
-# Build on Windows NT/XP
-
-TARGETOS = WINNT
-
 !IFDEF PATCHLEVEL
 RCFLAGS =	-DVIM_VERSION_PATCHLEVEL=$(PATCHLEVEL)
 !ENDIF
