# HG changeset patch
# Parent  85c0590cddfa0e2fea8b1451b409551822475a7a

diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -258,17 +258,25 @@ endif
 ifndef PYTHON3_VER
 PYTHON3_VER=31
 endif
+ifndef DYNAMIC_PYTHON3_DLL
+DYNAMIC_PYTHON3_DLL=python$(PYTHON3_VER).dll
+endif
+ifdef PYTHON3_HOME
+PYTHON3_HOME_DEF=-DPYTHON3_HOME=\"$(PYTHON3_HOME)\"
+endif
 
 ifeq (no,$(DYNAMIC_PYTHON3))
-PYTHON3LIB=-L$(PYTHON3)/libs -lPYTHON$(PYTHON3_VER)
+PYTHON3LIB=-L$(PYTHON3)/libs -lpython$(PYTHON3_VER)
 endif
 
+ifndef PYTHON3INC
 ifeq ($(CROSS),no)
 PYTHON3INC=-I $(PYTHON3)/include
 else
 PYTHON3INC=-I $(PYTHON3)/win32inc
 endif
 endif
+endif
 
 #	TCL interface:
 #	  TCL=[Path to TCL directory] (Set inside Make_cyg.mak or Make_ming.mak)
@@ -482,7 +490,7 @@ endif
 ifdef PYTHON3 
 CFLAGS += -DFEAT_PYTHON3 
 ifeq (yes, $(DYNAMIC_PYTHON3))
-CFLAGS += -DDYNAMIC_PYTHON3 -DDYNAMIC_PYTHON3_DLL=\"PYTHON$(PYTHON3_VER).dll\"
+CFLAGS += -DDYNAMIC_PYTHON3 -DDYNAMIC_PYTHON3_DLL=\"$(DYNAMIC_PYTHON3_DLL)\"
 endif
 endif
 
@@ -814,7 +822,7 @@ INCL = vim.h feature.h os_win32.h os_dos
 	$(CC) -c $(CFLAGS) $(PYTHONINC) $(PYTHON_HOME_DEF) $< -o $@
 
 $(OUTDIR)/if_python3.o : if_python3.c if_py_both.h $(INCL)
-	$(CC) -c $(CFLAGS) $(PYTHON3INC) $< -o $@
+	$(CC) -c $(CFLAGS) $(PYTHON3INC) $(PYTHON3_HOME_DEF) $< -o $@
 
 $(OUTDIR)/%.o : %.c $(INCL)
 	$(CC) -c $(CFLAGS) $< -o $@
