# HG changeset patch
# Parent 935ebda7488e31e6532809af01fede6ad9eb8238

diff --git a/src/Makefile b/src/Makefile
--- a/src/Makefile
+++ b/src/Makefile
@@ -351,7 +351,7 @@
 # If the selected GUI isn't found, the GUI is disabled automatically
 #CONF_OPT_GUI = --enable-gui=gtk2
 #CONF_OPT_GUI = --enable-gui=gtk2 --disable-gtktest
-#CONF_OPT_GUI = --enable-gui=gnome2
+CONF_OPT_GUI = --enable-gui=gnome2
 #CONF_OPT_GUI = --enable-gui=gnome2 --disable-gtktest
 #CONF_OPT_GUI = --enable-gui=motif
 #CONF_OPT_GUI = --enable-gui=motif --with-motif-lib="-static -lXm -shared"
@@ -434,7 +434,7 @@
 
 # TCL
 # Uncomment this when you want to include the Tcl interface.
-#CONF_OPT_TCL = --enable-tclinterp
+CONF_OPT_TCL = --enable-tclinterp
 #CONF_OPT_TCL = --enable-tclinterp --with-tclsh=tclsh8.4
 
 # CSCOPE
@@ -915,7 +915,7 @@
 #
 # Uncomment the next line to install Vim in your home directory.
 #prefix = $(HOME)
-prefix = /usr
+prefix = /usr/local
 
 ### exec_prefix	is the top directory for the executable (default $(prefix))
 #
