# HG changeset patch
# Parent  e0117ece79634edb55f6884eb4c9bf82fdfb3319

diff --git a/src/Makefile b/src/Makefile
--- a/src/Makefile
+++ b/src/Makefile
@@ -548,7 +548,7 @@ CONF_OPT_FEAT = --with-features=huge
 # again.
 #CC = cc
 #CC = gcc
-#CC = clang
+CC = clang
 
 # COMPILER FLAGS - change as you please. Either before running {{{1
 # configure or afterwards. For examples see below.
@@ -557,6 +557,7 @@ CONF_OPT_FEAT = --with-features=huge
 # When not defined, configure will try to use -O2 -g for gcc and -O for cc.
 #CFLAGS = -g
 #CFLAGS = -O
+CFLAGS = -g -O2 -Wall -Wmissing-prototypes -D_INTL_REDIRECT_MACROS
 
 # Optimization limits - depends on the compiler.  Automatic check in configure
 # doesn't work very well, because many compilers only give a warning for
