# HG changeset patch
# Parent  e13a1bf0c8473042e162d75dec5731aefd9b5613

diff --git a/src/gui.c b/src/gui.c
--- a/src/gui.c
+++ b/src/gui.c
@@ -55,7 +55,7 @@ enum {
 static void gui_attempt_start(void);
 
 static int can_update_cursor = TRUE; /* can display the cursor */
-static int cursor_updating = FALSE;
+static int disable_flush = 0;	/* If > 0, gui_mch_flush() is disabled. */
 
 /*
  * The Athena scrollbars can move the thumb to after the end of the scrollbar,
@@ -1977,8 +1977,7 @@ gui_write(
     gui.dragged_sb = SBAR_NONE;
 #endif
 
-    if (!cursor_updating)
-	gui_mch_flush();	    /* In case vim decides to take a nap */
+    gui_may_flush();		    /* In case vim decides to take a nap */
 }
 
 /*
@@ -2007,21 +2006,31 @@ gui_can_update_cursor(void)
 }
 
 /*
- * Start updating the cursor.  No need to call gui_mch_flush() yet.
+ * Disable issuing gui_mch_flush().
  */
     void
-gui_start_updating_cursor(void)
+gui_disable_flush(void)
 {
-    cursor_updating = TRUE;
+    ++disable_flush;
 }
 
 /*
- * End updating the cursor.  gui_mch_flush() can be called.
+ * Enable issuing gui_mch_flush().
  */
     void
-gui_end_updating_cursor(void)
+gui_enable_flush(void)
 {
-    cursor_updating = FALSE;
+    --disable_flush;
+}
+
+/*
+ * Issue gui_mch_flush() if it is not disabled.
+ */
+    void
+gui_may_flush(void)
+{
+    if (disable_flush == 0)
+	gui_mch_flush();
 }
 
     static void
@@ -3691,7 +3700,6 @@ gui_update_tabline(void)
 	/* Updating the tabline uses direct GUI commands, flush
 	 * outstanding instructions first. (esp. clear screen) */
 	out_flush();
-	gui_mch_flush();
 
 	if (!showit != !shown)
 	    gui_mch_show_tabline(showit);
@@ -4494,7 +4502,9 @@ gui_do_scroll(void)
 	 * disappear when losing focus after a scrollbar drag. */
 	if (wp->w_redr_type < type)
 	    wp->w_redr_type = type;
+	gui_disable_flush();
 	updateWindow(wp);   /* update window, status line, and cmdline */
+	gui_enable_flush();
     }
 
 #ifdef FEAT_INS_EXPAND
diff --git a/src/main.c b/src/main.c
--- a/src/main.c
+++ b/src/main.c
@@ -1242,7 +1242,16 @@ main_loop(
 	    if (VIsual_active)
 		update_curbuf(INVERTED);/* update inverted part */
 	    else if (must_redraw)
+	    {
+#ifdef FEAT_GUI
+		/* Stop issuing gui_mch_flush(). */
+		gui_disable_flush();
+#endif
 		update_screen(0);
+#ifdef FEAT_GUI
+		gui_enable_flush();
+#endif
+	    }
 	    else if (redraw_cmdline || clear_cmdline)
 		showmode();
 	    redraw_statuslines();
@@ -1285,7 +1294,7 @@ main_loop(
 	    {
 #  ifdef FEAT_GUI
 		/* Stop issuing gui_mch_flush(). */
-		gui_start_updating_cursor();
+		gui_disable_flush();
 #  endif
 		if (conceal_old_cursor_line != conceal_new_cursor_line
 			&& conceal_old_cursor_line
@@ -1293,7 +1302,7 @@ main_loop(
 		    update_single_line(curwin, conceal_old_cursor_line);
 		update_single_line(curwin, conceal_new_cursor_line);
 #  ifdef FEAT_GUI
-		gui_end_updating_cursor();
+		gui_enable_flush();
 #  endif
 		curwin->w_valid &= ~VALID_CROW;
 	    }
diff --git a/src/proto/gui.pro b/src/proto/gui.pro
--- a/src/proto/gui.pro
+++ b/src/proto/gui.pro
@@ -25,8 +25,9 @@ void gui_update_cursor_later(void);
 void gui_write(char_u *s, int len);
 void gui_dont_update_cursor(int undraw);
 void gui_can_update_cursor(void);
-void gui_start_updating_cursor(void);
-void gui_end_updating_cursor(void);
+void gui_disable_flush(void);
+void gui_enable_flush(void);
+void gui_may_flush(void);
 int gui_outstr_nowrap(char_u *s, int len, int flags, guicolor_T fg, guicolor_T bg, int back);
 void gui_undraw_cursor(void);
 void gui_redraw(int x, int y, int w, int h);
diff --git a/src/screen.c b/src/screen.c
--- a/src/screen.c
+++ b/src/screen.c
@@ -800,9 +800,9 @@ update_screen(int type_arg)
     {
 	if (did_undraw && !gui_mch_is_blink_off())
 	{
-	    gui_start_updating_cursor();
+	    gui_disable_flush();
 	    out_flush();	/* required before updating the cursor */
-	    gui_end_updating_cursor();
+	    gui_enable_flush();
 
 	    /* Put the GUI position where the cursor was, gui_update_cursor()
 	     * uses that. */
@@ -812,7 +812,7 @@ update_screen(int type_arg)
 	    gui.col = mb_fix_col(gui.col, gui.row);
 # endif
 	    gui_update_cursor(FALSE, FALSE);
-	    gui_mch_flush();
+	    gui_may_flush();
 	    screen_cur_col = gui.col;
 	    screen_cur_row = gui.row;
 	}
diff --git a/src/term.c b/src/term.c
--- a/src/term.c
+++ b/src/term.c
@@ -2514,15 +2514,15 @@ out_flush_cursor(
     int		clear_selection)/* clear selection under cursor */
 {
 #ifdef FEAT_GUI
-    gui_start_updating_cursor();
+    gui_disable_flush();
 #endif
     out_flush();
 #ifdef FEAT_GUI
-    gui_end_updating_cursor();
+    gui_enable_flush();
     if (gui.in_use)
     {
 	gui_update_cursor(force, clear_selection);
-	gui_mch_flush();
+	gui_may_flush();
     }
 #endif
 }
