# HG changeset patch
# Parent  8e4a857a37653164162ce0ca1dabd4ca8df8711c

diff --git a/src/userfunc.c b/src/userfunc.c
--- a/src/userfunc.c
+++ b/src/userfunc.c
@@ -355,7 +355,7 @@ get_lambda_tv(char_u **arg, typval_T *re
 
     if (evaluate)
     {
-	int	    len, flags = 0;
+	int	    flags = 0;
 	char_u	    *p;
 	char_u	    name[20];
 
@@ -373,14 +373,11 @@ get_lambda_tv(char_u **arg, typval_T *re
 	if (ga_grow(&newlines, 1) == FAIL)
 	    goto errret;
 
-	// Add "return " before the expression.
-	len = 7 + e - s + 1;
-	p = alloc(len);
+	// Store the expression.
+	p = vim_strnsave(s, (int)(e - s));
 	if (p == NULL)
 	    goto errret;
 	((char_u **)(newlines.ga_data))[newlines.ga_len++] = p;
-	STRCPY(p, "return ");
-	vim_strncpy(p + 7, s, e - s);
 
 	fp->uf_refcount = 1;
 	set_ufunc_name(fp, name);
@@ -1293,6 +1290,11 @@ call_user_func(
 
     if (default_arg_err && (fp->uf_flags & FC_ABORT))
 	did_emsg = TRUE;
+    else if (islambda)
+    {
+	char_u *p = *(char_u**)fp->uf_lines.ga_data;
+	eval1(&p, rettv, TRUE);
+    }
     else
 	// call do_cmdline() to execute the lines
 	do_cmdline(NULL, get_func_line, (void *)fc,
