# HG changeset patch
# Parent fffe7b85b76c6ba6af14f260cfecf3bab2226e24
diff --git a/src/version.c b/src/version.c
--- a/src/version.c
+++ b/src/version.c
@@ -1392,11 +1392,11 @@
 	    if (isalpha((int)vers[3]))
 	    {
 		int len = (isalpha((int)vers[4])) ? 5 : 4;
-		sprintf((char *)vers + len, ".%d%s", highest_patch(),
+		sprintf((char *)vers + len, ".%03d%s", highest_patch(),
 							 mediumVersion + len);
 	    }
 	    else
-		sprintf((char *)vers + 3, ".%d", highest_patch());
+		sprintf((char *)vers + 3, ".%03d", highest_patch());
 	}
 	col += (int)STRLEN(vers);
     }
