# HG changeset patch
# Parent  a5e0e2c27c58246bde170438098486caefde9d20

diff --git a/src/terminal.c b/src/terminal.c
--- a/src/terminal.c
+++ b/src/terminal.c
@@ -1912,6 +1912,11 @@ handle_moverect(VTermRect dest, VTermRec
 	VTermColor  fg, bg;
 	VTermScreenCellAttrs attr;
 	int	    clear_attr;
+	int	    disable_redraw = p_tf
+#ifdef FEAT_GUI
+					    || gui.in_use
+#endif
+					    ;
 
 	/* Set the color to clear lines with. */
 	vterm_state_get_default_colors(vterm_obtain_state(term->tl_vterm),
@@ -1919,6 +1924,8 @@ handle_moverect(VTermRect dest, VTermRec
 	vim_memset(&attr, 0, sizeof(attr));
 	clear_attr = cell2attr(attr, fg, bg);
 
+	if (disable_redraw)
+	    ++RedrawingDisabled;
 	FOR_ALL_WINDOWS(wp)
 	{
 	    if (wp->w_buffer == term->tl_buffer)
@@ -1926,6 +1933,8 @@ handle_moverect(VTermRect dest, VTermRec
 				 src.start_row - dest.start_row, FALSE, FALSE,
 				 clear_attr);
 	}
+	if (disable_redraw)
+	    --RedrawingDisabled;
     }
 
     term->tl_dirty_row_start = MIN(term->tl_dirty_row_start, dest.start_row);
