# HG changeset patch
# Parent  b366794c67551cd403263586cbcc929a74a62b32

diff --git a/src/message.c b/src/message.c
--- a/src/message.c
+++ b/src/message.c
@@ -4205,6 +4205,17 @@ vim_vsnprintf(
 		default: break;
 	    }
 
+# if defined(FEAT_EVAL) && defined(FEAT_NUM64)
+	    switch (fmt_spec)
+	    {
+	    case 'd': case 'u': case 'o': case 'x': case 'X':
+		if (tvs != NULL && length_modifier == '\0')
+		    length_modifier = 'L';
+	    default:
+		break;
+	    }
+# endif
+
 	    /* get parameter value, do initial processing */
 	    switch (fmt_spec)
 	    {
