Redraw after undo

From: K.Takata <kentkt@csc.jp>

Signed-off-by: K.Takata <kentkt@csc.jp>
---
 src/undo.c |    2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/undo.c b/src/undo.c
index 06efd9230..b54dcffac 100644
--- a/src/undo.c
+++ b/src/undo.c
@@ -3077,6 +3077,8 @@ u_undo_end(
     if (VIsual_active)
 	check_pos(curbuf, &VIsual);
 
+    redraw_later(UPD_NOT_VALID);
+
     smsg_attr_keep(0, _("%ld %s; %s #%ld  %s"),
 	    u_oldcount < 0 ? -u_oldcount : u_oldcount,
 	    _(msgstr),
