# HG changeset patch
# Parent 15d21ff11851ad1625e9931760510cb06e91d57e

diff --git a/src/gui_w32.c b/src/gui_w32.c
--- a/src/gui_w32.c
+++ b/src/gui_w32.c
@@ -579,6 +579,27 @@
     return (os_version.dwPlatformId == VER_PLATFORM_WIN32s);
 }
 
+/*
+ * Return TRUE when Visual Style is enabled.
+ */
+    static int
+is_visual_style_enabled(void)
+{
+    HANDLE hUxtheme;
+    static BOOL (WINAPI *pIsThemeActive)(void) = NULL;
+    static BOOL loaded = FALSE;
+
+    if (!loaded) {
+	hUxtheme = GetModuleHandle("uxtheme.dll");
+	if (hUxtheme != NULL)
+	    pIsThemeActive = (void*)GetProcAddress(hUxtheme, "IsThemeActive");
+	loaded = TRUE;
+    }
+    if (pIsThemeActive)
+	return pIsThemeActive();
+    return FALSE;
+}
+
 #ifdef FEAT_MENU
 /*
  * Figure out how high the menu bar is at the moment.
@@ -1655,7 +1676,7 @@
 	    return FAIL;
     }
     s_textArea = CreateWindowEx(
-	WS_EX_CLIENTEDGE,
+	is_visual_style_enabled() ? 0 : WS_EX_CLIENTEDGE,
 	szTextAreaClass, "Vim text area",
 	WS_CHILD | WS_VISIBLE, 0, 0,
 	100,				/* Any value will do for now */
@@ -1704,7 +1725,9 @@
     /*
      * Start out by adding the configured border width into the border offset
      */
-    gui.border_offset = gui.border_width + 2;	/*CLIENT EDGE*/
+    gui.border_offset = gui.border_width;
+    if (!is_visual_style_enabled())
+	gui.border_offset += 2;	    /*CLIENT EDGE*/
 
     /*
      * Set up for Intellimouse processing
