# HG changeset patch
# Parent  6473c0b7c03e713c43156af6d0ce9d1150bb40d0

diff --git a/src/os_win32.c b/src/os_win32.c
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -6289,7 +6289,6 @@ mch_total_mem(int special UNUSED)
 {
     MEMORYSTATUSEX  ms;
 
-    PlatformId();
     /* Need to use GlobalMemoryStatusEx() when there is more memory than
      * what fits in 32 bits. But it's not always available. */
     ms.dwLength = sizeof(MEMORYSTATUSEX);
@@ -6427,8 +6426,6 @@ mch_rename(
     char *
 default_shell(void)
 {
-    PlatformId();
-
     return "cmd.exe";
 }
 
@@ -6724,7 +6721,6 @@ copy_infostreams(char_u *from, char_u *t
 mch_copy_file_attribute(char_u *from, char_u *to)
 {
     /* File streams only work on Windows NT and later. */
-    PlatformId();
     copy_infostreams(from, to);
     return 0;
 }
@@ -6755,8 +6751,6 @@ myresetstkoflw(void)
     DWORD	nPageSize;
     DWORD	dummy;
 
-    PlatformId();
-
     /* We need to know the system page size. */
     GetSystemInfo(&si);
     nPageSize = si.dwPageSize;
