# HG changeset patch
# Parent  95c701d0a6e1eb3d91ef7430da807272ef3e725f

diff --git a/src/evalfunc.c b/src/evalfunc.c
--- a/src/evalfunc.c
+++ b/src/evalfunc.c
@@ -6231,7 +6231,7 @@ f_has(typval_T *argvars, typval_T *rettv
 #ifdef MSWIN
 	"win32",
 #endif
-#if defined(UNIX) && (defined(__CYGWIN32__) || defined(__CYGWIN__))
+#if defined(UNIX) && defined(__CYGWIN__)
 	"win32unix",
 #endif
 #ifdef _WIN64
diff --git a/src/os_unix.c b/src/os_unix.c
--- a/src/os_unix.c
+++ b/src/os_unix.c
@@ -1605,7 +1605,7 @@ x_IOerror_check(Display *dpy UNUSED)
 {
     /* This function should not return, it causes exit().  Longjump instead. */
     LONGJMP(lc_jump_env, 1);
-#  if defined(VMS) || defined(__CYGWIN__) || defined(__CYGWIN32__)
+#  if defined(VMS) || defined(__CYGWIN__)
     return 0;  /* avoid the compiler complains about missing return value */
 #  endif
 }
@@ -1627,7 +1627,7 @@ x_IOerror_handler(Display *dpy UNUSED)
 
     /* This function should not return, it causes exit().  Longjump instead. */
     LONGJMP(x_jump_env, 1);
-# if defined(VMS) || defined(__CYGWIN__) || defined(__CYGWIN32__)
+# if defined(VMS) || defined(__CYGWIN__)
     return 0;  /* avoid the compiler complains about missing return value */
 # endif
 }
@@ -6713,7 +6713,7 @@ mch_expand_wildcards(
     }
     vim_free(tempname);
 
-# if defined(__CYGWIN__) || defined(__CYGWIN32__)
+# ifdef __CYGWIN__
     /* Translate <CR><NL> into <NL>.  Caution, buffer may contain NUL. */
     p = buffer;
     for (i = 0; i < (int)len; ++i)
diff --git a/src/os_win32.c b/src/os_win32.c
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -2046,8 +2044,8 @@ executable_exists(char *name, char_u **p
     return TRUE;
 }
 
-#if ((defined(__MINGW32__) || defined (__CYGWIN32__)) && \
-       __MSVCRT_VERSION__ >= 0x800) || (defined(_MSC_VER) && _MSC_VER >= 1400)
+#if (defined(__MINGW32__) && __MSVCRT_VERSION__ >= 0x800) || \
+	(defined(_MSC_VER) && _MSC_VER >= 1400)
 /*
  * Bad parameter handler.
  *
