# HG changeset patch
# Parent b8538e19d4267758867af94ae48eab74e13e2814
diff --git a/src/testdir/Make_dos.mak b/src/testdir/Make_dos.mak
--- a/src/testdir/Make_dos.mak
+++ b/src/testdir/Make_dos.mak
@@ -53,13 +53,16 @@
 	-$(VIMPROG) -u dos.vim --noplugin "+argdo set ff=unix|upd" +q \
 		dotest.in test60.ok test71.ok test74.ok
 
+restoreff:
+	-$(VIMPROG) -u dos.vim --noplugin "+argdo set ff=unix|upd" +q *.in *.ok
+
 report:
 	@echo ""
 	@echo Test results:
 	@IF EXIST test.log ( type test.log & echo TEST FAILURE & exit /b 1 ) \
 		ELSE ( ECHO ALL DONE )
 
-clean:
+clean:	restoreff
 	-del *.out
 	-del *.failed
 	-if exist test.ok del test.ok
