# HG changeset patch
# Parent  f8bb15c83cac90024b3e9357e62a632f53b5cc19

diff --git a/src/time.c b/src/time.c
--- a/src/time.c
+++ b/src/time.c
@@ -284,31 +284,10 @@ f_strftime(typval_T *argvars, typval_T *
 	vim_free(wp);
 # else
 	char_u	    result_buf[256];
-	vimconv_T   conv;
-	char_u	    *enc;
 
-	conv.vc_type = CONV_NONE;
-	enc = enc_locale();
-	convert_setup(&conv, p_enc, enc);
-	if (conv.vc_type != CONV_NONE)
-	    p = string_convert(&conv, p, NULL);
-	if (p != NULL)
-	    (void)strftime((char *)result_buf, sizeof(result_buf),
+	(void)strftime((char *)result_buf, sizeof(result_buf),
 							  (char *)p, curtime);
-	else
-	    result_buf[0] = NUL;
-
-	if (conv.vc_type != CONV_NONE)
-	    vim_free(p);
-	convert_setup(&conv, enc, p_enc);
-	if (conv.vc_type != CONV_NONE)
-	    rettv->vval.v_string = string_convert(&conv, result_buf, NULL);
-	else
-	    rettv->vval.v_string = vim_strsave(result_buf);
-
-	// Release conversion descriptors
-	convert_setup(&conv, NULL, NULL);
-	vim_free(enc);
+	rettv->vval.v_string = vim_strsave(result_buf);
 # endif
     }
 }
