# HG changeset patch
# Parent  0b8132f937a29711e8051a96d68eb69cad6d4aad

diff --git a/src/testdir/test_undo.vim b/src/testdir/test_undo.vim
--- a/src/testdir/test_undo.vim
+++ b/src/testdir/test_undo.vim
@@ -328,6 +328,9 @@ func Test_insert_expr()
 endfunc
 
 func Test_undofile_earlier()
+  if has('win32') && $GITHUB_ACTIONS == 'true'
+    throw 'Skipped: does not work on Windows on GitHub Actions by unknown reasons'
+  endif
   " Issue #1254
   " create undofile with timestamps older than Vim startup time.
   let t0 = localtime() - 43200
