# HG changeset patch
# Parent  529e20bd06d7a891cfa826a26f0a46f275630c1e

diff --git a/runtime/doc/eval.txt b/runtime/doc/eval.txt
--- a/runtime/doc/eval.txt
+++ b/runtime/doc/eval.txt
@@ -7711,7 +7711,7 @@ readdirex({directory} [, {expr}])			*rea
 		This is much faster than calling |readdir()| then calling
 		|getfperm()|, |getfsize()|, |getftime()| and |getftype()| for
 		each file and directory especially on MS-Windows.
-		NOTE: unlike |readdir()|, the list may be unsorted.
+		The list will be sorted by name (case sensitive).
 
 		The Dictionary for file and directory information has the
 		following items:
@@ -7747,9 +7747,6 @@ readdirex({directory} [, {expr}])			*rea
 		When {expr} is a function the entry is passed as the argument.
 		For example, to get a list of files ending in ".txt": >
 		  readdirex(dirname, {e -> e.name =~ '.txt$'})
-<		To sort the result by name (case sensitive): >
-		  readdirex(dirname)->sort({e1, e2 -> e1.name ==# e2.name ?
-			\ 0 : e1.name ># e2.name ? 1 : -1})
 <
 		Can also be used as a |method|: >
 			GetDirName()->readdirex()
diff --git a/src/fileio.c b/src/fileio.c
--- a/src/fileio.c
+++ b/src/fileio.c
@@ -4746,6 +4746,16 @@ theend:
 }
 # endif
 
+    static int
+compare_readdirex_item(const void *p1, const void *p2)
+{
+    char_u  *name1, *name2;
+
+    name1 = dict_get_string(*(dict_T**)p1, "name", FALSE);
+    name2 = dict_get_string(*(dict_T**)p2, "name", FALSE);
+    return STRCMP(name1, name2);
+}
+
 /*
  * Core part of "readdirex()" function.
  * Retrieve the list of files/directories of "path" into "list".
@@ -4759,13 +4769,19 @@ readdirex_core(
     int		(*checkitem)(void *context, dict_T *item))
 {
     int			failed = FALSE;
+    int			i;
+    garray_T		ga;
 # ifdef MSWIN
     char_u		*buf, *p;
     int			ok;
     HANDLE		hFind = INVALID_HANDLE_VALUE;
     WIN32_FIND_DATAW    wfd;
     WCHAR		*wn = NULL;	// UTF-16 name, NULL when not used.
-
+# endif
+
+    ga_init2(&ga, (int)sizeof(dict_T *), 20);
+
+# ifdef MSWIN
     buf = alloc(MAXPATHL);
     if (buf == NULL)
 	return FAIL;
@@ -4814,14 +4830,17 @@ readdirex_core(
 
 	    if (!ignore)
 	    {
-		if (list_append_dict(list, item) == FAIL)
+		if (ga_grow(&ga, 1) == OK)
+		    ((dict_T**)ga.ga_data)[ga.ga_len++] = item;
+		else
 		{
 		    failed = TRUE;
 		    dict_unref(item);
 		    break;
 		}
 	    }
-	    dict_unref(item);
+	    else
+		dict_unref(item);
 
 	    ok = FindNextFileW(hFind, &wfd);
 	}
@@ -4874,20 +4893,35 @@ readdirex_core(
 
 	    if (!ignore)
 	    {
-		if (list_append_dict(list, item) == FAIL)
+		if (ga_grow(&ga, 1) == OK)
+		    ((dict_T**)ga.ga_data)[ga.ga_len++] = item;
+		else
 		{
 		    failed = TRUE;
 		    dict_unref(item);
 		    break;
 		}
 	    }
-	    dict_unref(item);
+	    else
+		dict_unref(item);
 	}
 
 	closedir(dirp);
     }
 # endif
 
+    if (!failed && ga.ga_len > 0)
+    {
+	qsort((void*)ga.ga_data, (size_t)ga.ga_len, sizeof(dict_T*),
+		compare_readdirex_item);
+	for (i = 0; i < ga.ga_len; i++)
+	{
+	    list_append_dict(list, ((dict_T**)ga.ga_data)[i]);
+	    dict_unref(((dict_T**)ga.ga_data)[i]);
+	}
+    }
+    ga_clear(&ga);
+
     return failed ? FAIL : OK;
 }
 #endif
