# HG changeset patch
# Parent  c4c97646331087a88e79d85a68d86e500d59ec25

diff --git a/src/evalfunc.c b/src/evalfunc.c
--- a/src/evalfunc.c
+++ b/src/evalfunc.c
@@ -11097,7 +11097,6 @@ f_str2nr(typval_T *argvars, typval_T *re
     char_u	*p;
     varnumber_T	n;
     int		what;
-    int		isneg;
 
     if (argvars[1].v_type != VAR_UNKNOWN)
     {
@@ -11110,8 +11109,7 @@ f_str2nr(typval_T *argvars, typval_T *re
     }
 
     p = skipwhite(get_tv_string(&argvars[0]));
-    isneg = (*p == '-');
-    if (*p == '+' || *p == '-')
+    if (*p == '+')
 	p = skipwhite(p + 1);
     switch (base)
     {
@@ -11121,11 +11119,7 @@ f_str2nr(typval_T *argvars, typval_T *re
 	default: what = 0;
     }
     vim_str2nr(p, NULL, NULL, what, &n, NULL, 0);
-    if (isneg)
-	rettv->vval.v_number = -n;
-    else
-	rettv->vval.v_number = n;
-
+    rettv->vval.v_number = n;
 }
 
 #ifdef HAVE_STRFTIME
