# HG changeset patch
# Parent  7dd3dce73b35f7846b891f59a2215cc59ffb0557

diff --git a/nsis/gvim.nsi b/nsis/gvim.nsi
--- a/nsis/gvim.nsi
+++ b/nsis/gvim.nsi
@@ -215,6 +215,28 @@ Section "Vim executables and runtime fil
 	SetOutPath $0\macros
 	File ${VIMRT}\macros\*.*
 
+	SetOutPath $0\pack\dist\opt\dvorak\dvorak
+	File ${VIMRT}\pack\dist\opt\dvorak\dvorak\*.*
+	SetOutPath $0\pack\dist\opt\dvorak\plugin
+	File ${VIMRT}\pack\dist\opt\dvorak\plugin\*.*
+
+	SetOutPath $0\pack\dist\opt\editexisting\plugin
+	File ${VIMRT}\pack\dist\opt\editexisting\plugin\*.*
+
+	SetOutPath $0\pack\dist\opt\justify\plugin
+	File ${VIMRT}\pack\dist\opt\justify\plugin\*.*
+
+	SetOutPath $0\pack\dist\opt\matchit\doc
+	File ${VIMRT}\pack\dist\opt\matchit\doc\*.*
+	SetOutPath $0\pack\dist\opt\matchit\plugin
+	File ${VIMRT}\pack\dist\opt\matchit\plugin\*.*
+
+	SetOutPath $0\pack\dist\opt\shellmenu\plugin
+	File ${VIMRT}\pack\dist\opt\shellmenu\plugin\*.*
+
+	SetOutPath $0\pack\dist\opt\swapmouse\plugin
+	File ${VIMRT}\pack\dist\opt\swapmouse\plugin\*.*
+
 	SetOutPath $0\plugin
 	File ${VIMRT}\plugin\*.*
 
