# HG changeset patch
# Parent  fae369f94e3f200a2830118d1ec8fa1159879b5c

diff --git a/src/auto/configure b/src/auto/configure
--- a/src/auto/configure
+++ b/src/auto/configure
@@ -9516,7 +9516,7 @@ fi
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking for CYGWIN environment" >&5
 $as_echo_n "checking for CYGWIN environment... " >&6; }
 case `uname` in
-    CYGWIN*)    CYGWIN=yes; { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
+    CYGWIN*|MSYS*)    CYGWIN=yes; { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
 $as_echo "yes" >&6; }
                 { $as_echo "$as_me:${as_lineno-$LINENO}: checking for CYGWIN clipboard support" >&5
 $as_echo_n "checking for CYGWIN clipboard support... " >&6; }
diff --git a/src/configure.in b/src/configure.in
--- a/src/configure.in
+++ b/src/configure.in
@@ -2805,7 +2805,7 @@ dnl ------------------------------------
 dnl Check for Cygwin, which needs an extra source file if not using X11
 AC_MSG_CHECKING(for CYGWIN environment)
 case `uname` in
-    CYGWIN*)    CYGWIN=yes; AC_MSG_RESULT(yes)
+    CYGWIN*|MSYS*)    CYGWIN=yes; AC_MSG_RESULT(yes)
                 AC_MSG_CHECKING(for CYGWIN clipboard support)
                 if test "x$with_x" = "xno" ; then
                   OS_EXTRA_SRC=winclip.c; OS_EXTRA_OBJ=objects/winclip.o
