# HG changeset patch
# Parent  1bc7646598ff382e222dbc0dac7d396797e0f7ff

diff --git a/runtime/doc/options.txt b/runtime/doc/options.txt
--- a/runtime/doc/options.txt
+++ b/runtime/doc/options.txt
@@ -6629,10 +6629,14 @@ A jump table for the options with a shor
 	The name of the temporary file can be represented by "%s" if necessary
 	(the file name is appended automatically if no %s appears in the value
 	of this option).
-	For the Amiga the default is ">".  For MS-Windows using powershell the
-	default is "2>&1 | Out-File -Encoding default", otherwise the default
-	is ">%s 2>&1".  The output is directly saved in a file and not echoed
-	to the screen.
+	For the Amiga the default is ">".  The output is directly saved in a
+	file and not echoed to the screen.
+	For MS-Windows using powershell the default is "2>&1 | Out-File
+	-Encoding default", otherwise if "tee" command is not found in the
+	PATH the default is ">%s 2>&1".  The output is directly saved in a
+	file and not echoed to the screen.  If "tee" command is found in the
+	PATH the default becomes "2>&1| tee" or "|& tee" depends on the
+	'shell' option.  The output is also echoed to the screen.
 	For Unix the default is "| tee".  The stdout of the compiler is saved
 	in a file and echoed to the screen.  If the 'shell' option is "csh" or
 	"tcsh" after initializations, the default becomes "|& tee".  If the
diff --git a/runtime/doc/quickfix.txt b/runtime/doc/quickfix.txt
--- a/runtime/doc/quickfix.txt
+++ b/runtime/doc/quickfix.txt
@@ -944,11 +944,13 @@ or simpler >
 "$*" can be given multiple times, for example: >
    :set makeprg=gcc\ -o\ $*\ $*
 
-The 'shellpipe' option defaults to ">" for the Amiga and ">%s 2>&1" for Win32.
-This means that the output of the compiler is saved in a file and not shown on
-the screen directly.  For Unix "| tee" is used.  The compiler output is shown
-on the screen and saved in a file the same time.  Depending on the shell used
-"|& tee" or "2>&1| tee" is the default, so stderr output will be included.
+The 'shellpipe' option defaults to ">" for the Amiga and ">%s 2>&1" Win32 if
+"tee" command is not found in the PATH.  This means that the output of the
+compiler is saved in a file and not shown on the screen directly.  If "tee"
+command is found on Win32 the default becomes "2>&1| tee" and the output is
+also echoed to the screen.  For Unix "| tee" is used.  The compiler output is
+shown on the screen and saved in a file the same time.  Depending on the shell
+used "|& tee" or "2>&1| tee" is the default, so stderr output will be included.
 
 If 'shellpipe' is empty, the {errorfile} part will be omitted.  This is useful
 for compilers that write to an errorfile themselves (e.g., Manx's Amiga C).
diff --git a/src/option.c b/src/option.c
--- a/src/option.c
+++ b/src/option.c
@@ -919,10 +919,11 @@ set_init_3(void)
 	    if (do_sp)
 	    {
 #  ifdef MSWIN
-		p_sp = (char_u *)">&";
-#  else
-		p_sp = (char_u *)"|& tee";
+		if (!mch_can_exe((char_u *)"tee", NULL, TRUE))
+		    p_sp = (char_u *)">&";
+		else
 #  endif
+		    p_sp = (char_u *)"|& tee";
 		options[idx_sp].def_val[VI_DEFAULT] = p_sp;
 	    }
 # endif
@@ -983,10 +984,11 @@ set_init_3(void)
 		if (do_sp)
 		{
 #  ifdef MSWIN
-		    p_sp = (char_u *)">%s 2>&1";
-#  else
-		    p_sp = (char_u *)"2>&1| tee";
+		    if (!mch_can_exe((char_u *)"tee", NULL, TRUE))
+			p_sp = (char_u *)">%s 2>&1";
+		    else
 #  endif
+			p_sp = (char_u *)"2>&1| tee";
 		    options[idx_sp].def_val[VI_DEFAULT] = p_sp;
 		}
 # endif
diff --git a/src/testdir/test_shell.vim b/src/testdir/test_shell.vim
--- a/src/testdir/test_shell.vim
+++ b/src/testdir/test_shell.vim
@@ -22,22 +22,22 @@ func Test_shell_options()
           \ ['tcsh', '-c', '|& tee', '', '>&', '', '']]
   endif
   if has('win32')
-    let shells += [['cmd', '/c', '>%s 2>&1', '', '>%s 2>&1', '"&|<>()@^', ''],
-          \ ['cmd.exe', '/c', '>%s 2>&1', '', '>%s 2>&1', '"&|<>()@^', '('],
+    let shells += [['cmd', '/c', '2>&1| tee', '', '>%s 2>&1', '"&|<>()@^', ''],
+          \ ['cmd.exe', '/c', '2>&1| tee', '', '>%s 2>&1', '"&|<>()@^', '('],
           \ ['powershell.exe', '-Command', '2>&1 | Out-File -Encoding default',
           \           '', '2>&1 | Out-File -Encoding default', '"&|<>()@^', '"'],
           \ ['powershell', '-Command', '2>&1 | Out-File -Encoding default', '',
           \               '2>&1 | Out-File -Encoding default', '"&|<>()@^', '"'],
-          \ ['sh.exe', '-c', '>%s 2>&1', '', '>%s 2>&1', '"&|<>()@^', '"'],
-          \ ['ksh.exe', '-c', '>%s 2>&1', '', '>%s 2>&1', '"&|<>()@^', '"'],
-          \ ['mksh.exe', '-c', '>%s 2>&1', '', '>%s 2>&1', '"&|<>()@^', '"'],
-          \ ['pdksh.exe', '-c', '>%s 2>&1', '', '>%s 2>&1', '"&|<>()@^', '"'],
-          \ ['zsh.exe', '-c', '>%s 2>&1', '', '>%s 2>&1', '"&|<>()@^', '"'],
-          \ ['zsh-beta.exe', '-c', '>%s 2>&1', '', '>%s 2>&1', '"&|<>()@^', '"'],
-          \ ['bash.exe', '-c', '>%s 2>&1', '', '>%s 2>&1', '"&|<>()@^', '"'],
-          \ ['dash.exe', '-c', '>%s 2>&1', '', '>%s 2>&1', '"&|<>()@^', '"'],
-          \ ['csh.exe', '-c', '>&', '', '>&', '"&|<>()@^', '"'],
-          \ ['tcsh.exe', '-c', '>&', '', '>&', '"&|<>()@^', '"']]
+          \ ['sh.exe', '-c', '2>&1| tee', '', '>%s 2>&1', '"&|<>()@^', '"'],
+          \ ['ksh.exe', '-c', '2>&1| tee', '', '>%s 2>&1', '"&|<>()@^', '"'],
+          \ ['mksh.exe', '-c', '2>&1| tee', '', '>%s 2>&1', '"&|<>()@^', '"'],
+          \ ['pdksh.exe', '-c', '2>&1| tee', '', '>%s 2>&1', '"&|<>()@^', '"'],
+          \ ['zsh.exe', '-c', '2>&1| tee', '', '>%s 2>&1', '"&|<>()@^', '"'],
+          \ ['zsh-beta.exe', '-c', '2>&1| tee', '', '>%s 2>&1', '"&|<>()@^', '"'],
+          \ ['bash.exe', '-c', '2>&1| tee', '', '>%s 2>&1', '"&|<>()@^', '"'],
+          \ ['dash.exe', '-c', '2>&1| tee', '', '>%s 2>&1', '"&|<>()@^', '"'],
+          \ ['csh.exe', '-c', '|& tee', '', '>&', '"&|<>()@^', '"'],
+          \ ['tcsh.exe', '-c', '|& tee', '', '>&', '"&|<>()@^', '"']]
   endif
 
   " start a new Vim instance with 'shell' set to each of the supported shells
