# HG changeset patch
# Parent bf06863be498f6339cd3c5f885b1654149f11d07
diff --git a/src/if_tcl.c b/src/if_tcl.c
--- a/src/if_tcl.c
+++ b/src/if_tcl.c
@@ -156,6 +156,11 @@
 #  define DYNAMIC_TCL_VER "8.3"
 # endif
 
+# ifndef Tcl_FindExecutable /* work around for Tcl 8.6 */
+#  define Tcl_FindExecutable \
+	(tclStubsPtr->tcl_FindExecutable)
+# endif
+
 # ifndef  DYNAMIC_TCL /* Just generating prototypes */
 typedef int HANDLE;
 # endif
