# HG changeset patch
# Parent  ec9380ef504bb4af016cac8e9909550589b0c535

diff --git a/src/Make_mvc.mak b/src/Make_mvc.mak
--- a/src/Make_mvc.mak
+++ b/src/Make_mvc.mak
@@ -215,11 +215,18 @@ ASSEMBLY_ARCHITECTURE=$(CPU)
 ASSEMBLY_ARCHITECTURE = x86
 !  endif
 ! else
+!  ifdef PLATFORM
+CPU = $(PLATFORM)
+!  else
 CPU = $(PROCESSOR_ARCHITECTURE)
-ASSEMBLY_ARCHITECTURE = $(PROCESSOR_ARCHITECTURE)
+!  endif
+ASSEMBLY_ARCHITECTURE = $(CPU)
 !  if ("$(CPU)" == "x86") || ("$(CPU)" == "X86")
 CPU = i386
 !  endif
+!  if ("$(CPU)" == "x64") || ("$(CPU)" == "X64")
+CPU = AMD64
+!  endif
 ! endif
 !else  # !PROCESSOR_ARCHITECTURE
 # We're on Windows 95
