# HG changeset patch
# Parent  daafc613df026e623e0b0d4685866884248a307b

diff --git a/src/if_perl.xs b/src/if_perl.xs
--- a/src/if_perl.xs
+++ b/src/if_perl.xs
@@ -44,6 +44,11 @@
 # define __builtin_expect(expr, val) (expr)
 // Work around for using MSVC and Strawberry Perl 5.32.
 # define NO_THREAD_SAFE_LOCALE
+
+# ifdef DYNAMIC_PERL
+// Work around for suppressing warning C4273 (inconsistent DLL linkage).
+#  define PERL_EXT_RE_BUILD
+# endif
 #endif
 
 #ifdef __GNUC__
@@ -1483,7 +1488,7 @@ vim_IOLayer_init(void)
 
 # if (PERL_REVISION == 5) && (PERL_VERSION >= 18)
 #  undef Perl_sv_free2
-void Perl_sv_free2(pTHX_ SV* sv, const U32 refcnt)
+void Perl_sv_free2(pTHX_ SV *const sv, const U32 refcnt)
 {
     (*dll_Perl_sv_free2)(aTHX_ sv, refcnt);
 }
@@ -1519,7 +1524,7 @@ NV Perl_sv_2nv_flags(pTHX_ SV *const sv,
 
 # ifdef PERL589_OR_LATER
 #  undef Perl_sv_2iv_flags
-IV Perl_sv_2iv_flags(pTHX_ SV* sv, I32 flags)
+IV Perl_sv_2iv_flags(pTHX_ SV *const sv, const I32 flags)
 {
     return (*dll_Perl_sv_2iv_flags)(aTHX_ sv, flags);
 }
