# HG changeset patch
# Parent 3eb577298089bae473d02784fba62fad02fbaf11
diff --git a/src/os_mswin.c b/src/os_mswin.c
--- a/src/os_mswin.c
+++ b/src/os_mswin.c
@@ -612,6 +612,7 @@
     char	buf[_MAX_PATH + 1];
 #endif
     char	*p;
+int n;
 
     vim_strncpy((char_u *)buf, (char_u *)name, sizeof(buf) - 1);
     p = buf + strlen(buf);
@@ -634,6 +635,11 @@
 	{
 	    n = wstat_symlink_aware(wp, (struct _stat *)stp);
 	    vim_free(wp);
+    {
+	char buf[256];
+	sprintf(buf, "W:%s: %d %08x\n", name, n, stp->st_mode);
+	OutputDebugString(buf);
+    }
 	    if (n >= 0)
 		return n;
 	    /* Retry with non-wide function (for Windows 98). Can't use
@@ -642,7 +648,13 @@
 	}
     }
 #endif
-    return stat_symlink_aware(buf, stp);
+    n = stat_symlink_aware(buf, stp);
+    {
+	char buf[256];
+	sprintf(buf, "A:%s: %d %08x\n", name, n, stp->st_mode);
+	OutputDebugString(buf);
+    }
+    return n;
 }
 
 #if defined(FEAT_GUI_MSWIN) || defined(PROTO)
diff --git a/src/os_win32.c b/src/os_win32.c
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -2894,6 +2894,11 @@
 	/* UNC path */
 	return (long)win32_getattrs(name);
     n = mch_stat(name, &st);
+    if (0) {
+	char buf[256];
+	sprintf(buf, "%s: %d %08x\n", name, n, st.st_mode);
+	OutputDebugString(buf);
+    }
     return n == 0 ? (long)st.st_mode : -1L;
 }
 
@@ -2908,6 +2913,7 @@
 {
     long	n = -1;
 
+    OutputDebugString("mch_setperm");
 #ifdef FEAT_MBYTE
     if (enc_codepage >= 0 && (int)GetACP() != enc_codepage)
     {
@@ -2940,6 +2946,7 @@
 mch_hide(char_u *name)
 {
     int attrs = win32_getattrs(name);
+    OutputDebugString("mch_hide");
     if (attrs == -1)
 	return;
 
@@ -2955,6 +2962,7 @@
 mch_isdir(char_u *name)
 {
     int f = win32_getattrs(name);
+    OutputDebugString("mch_isdir");
 
     if (f == -1)
 	return FALSE;		    /* file does not exist at all */
@@ -2993,6 +3001,7 @@
 mch_is_hard_link(char_u *fname)
 {
     BY_HANDLE_FILE_INFORMATION info;
+    OutputDebugString("mch_is_hard_link");
 
     return win32_fileinfo(fname, &info) == FILEINFO_OK
 						   && info.nNumberOfLinks > 1;
@@ -3011,6 +3020,7 @@
 #ifdef FEAT_MBYTE
     WCHAR		*wn = NULL;
     WIN32_FIND_DATAW	findDataW;
+    OutputDebugString("mch_is_symbolic_link");
 
     if (enc_codepage >= 0 && (int)GetACP() != enc_codepage)
 	wn = enc_to_utf16(fname, NULL);
@@ -3162,6 +3172,11 @@
 #ifdef FEAT_MBYTE
     vim_free(p);
 #endif
+    {
+	char buf[256];
+	sprintf(buf, "win32_getattrs:%s: %d\n", name, attr);
+	OutputDebugString(buf);
+    }
     return attr;
 }
 
@@ -3209,6 +3224,7 @@
 win32_set_archive(char_u *name)
 {
     int attrs = win32_getattrs(name);
+    OutputDebugString("win32_set_archive");
     if (attrs == -1)
 	return -1;
 
@@ -3225,6 +3241,7 @@
 mch_writable(char_u *name)
 {
     int attrs = win32_getattrs(name);
+    OutputDebugString("mch_writable");
 
     return (attrs != -1 && (!(attrs & FILE_ATTRIBUTE_READONLY)
 			  || (attrs & FILE_ATTRIBUTE_DIRECTORY)));
