# HG changeset patch
# Parent  962e42239e7eb386f0a43134245bb840cfd81d80

diff --git a/src/terminal.c b/src/terminal.c
--- a/src/terminal.c
+++ b/src/terminal.c
@@ -385,10 +385,6 @@ term_start(typval_T *argvar, jobopt_T *o
      * the job finished. */
     curbuf->b_p_ma = FALSE;
 
-    /* Set 'bufhidden' to "hide": allow closing the window. */
-    set_string_option_direct((char_u *)"bufhidden", -1,
-				      (char_u *)"hide", OPT_FREE|OPT_LOCAL, 0);
-
     set_term_and_win_size(term);
     setup_job_options(opt, term->tl_rows, term->tl_cols);
 
