# HG changeset patch
# Parent  8c110d08b87f6194fb8a28b903583f4c9b759079

diff --git a/src/testdir/Makefile b/src/testdir/Makefile
--- a/src/testdir/Makefile
+++ b/src/testdir/Makefile
@@ -59,7 +59,7 @@ clean:
 
 test1.out: test1.in
 	-rm -rf $*.failed $(RM_ON_RUN) $(RM_ON_START) wrongtermsize
-	$(RUN_VIM) $*.in
+	$(RUN_VIM) $*.in > /dev/null
 	@/bin/sh -c "if test -f wrongtermsize; \
 		then echo; \
 		echo test1 FAILED - terminal size must be 80x24 or larger; \
@@ -78,14 +78,14 @@ test1.out: test1.in
 	# 200 msec is sufficient, but only modern sleep supports a fraction of
 	# a second, fall back to a second if it fails.
 	@-/bin/sh -c "sleep .2 > /dev/null 2>&1 || sleep 1"
-	-$(RUN_VIM) $*.in
+	-$(RUN_VIM) $*.in > /dev/null
 
 	# For flaky tests retry one time.  No tests at the moment.
 	#@/bin/sh -c "if test -f test.out -a $* = test61; then \
 	#	  if diff test.out $*.ok; \
 	#	  then echo flaky test ok first time; \
 	#	  else rm -rf $*.failed $(RM_ON_RUN); \
-	#		$(RUN_VIM) $*.in; \
+	#		$(RUN_VIM) $*.in > /dev/null; \
 	#	  fi \
 	#	fi"
 
@@ -108,7 +108,7 @@ bench_re_freeze.out: bench_re_freeze.vim
 	# 200 msec is sufficient, but only modern sleep supports a fraction of
 	# a second, fall back to a second if it fails.
 	@-/bin/sh -c "sleep .2 > /dev/null 2>&1 || sleep 1"
-	-$(RUN_VIM) $*.in
+	-$(RUN_VIM) $*.in > /dev/null
 	@/bin/sh -c "if test -f benchmark.out; then cat benchmark.out; fi"
 
 nolog:
@@ -127,4 +127,4 @@ newtestssilent: $(NEW_TESTS)
 
 
 .vim.res:
-	$(RUN_VIMTEST) -u NONE -U NONE -S runtest.vim $*.vim
+	$(RUN_VIMTEST) -u NONE -U NONE -S runtest.vim $*.vim > /dev/null
