# HG changeset patch
# Parent  ce68b096106c59edd7d07e168f6edde07f83fa35

diff --git a/src/testdir/Makefile b/src/testdir/Makefile
--- a/src/testdir/Makefile
+++ b/src/testdir/Makefile
@@ -40,7 +40,7 @@ tiny:	nolog tinytests report
 benchmark: $(SCRIPTS_BENCH)
 
 report:
-	# without the +eval feature test_result.log is a copy of test.log
+	@# without the +eval feature test_result.log is a copy of test.log
 	@/bin/sh -c "if test -f test.log; \
 		then cp test.log test_result.log; \
 		else echo No failures reported > test_result.log; \
