# HG changeset patch
# Parent  27f652f3521a4d318acbc55c879ffb0a7d3d5c40

diff --git a/src/syntax.c b/src/syntax.c
--- a/src/syntax.c
+++ b/src/syntax.c
@@ -7795,13 +7795,7 @@ do_highlight(
 # if defined(FEAT_GUI) || defined(FEAT_TERMGUICOLORS)
 		/* In GUI guifg colors are only used when recognized */
 		i = color_name2handle(arg);
-		if (i != INVALCOLOR || STRCMP(arg, "NONE") == 0
-#  ifdef FEAT_GUI
-			|| !(USE_24BIT)
-#  else
-			|| !p_tgc
-#  endif
-		   )
+		if (i != INVALCOLOR || STRCMP(arg, "NONE") == 0 || !USE_24BIT)
 		{
 		    HL_TABLE()[idx].sg_gui_fg = i;
 # endif
