# HG changeset patch
# Parent  1396a87f62b73fb5bf322ad1809993cffca4296e

diff --git a/src/ex_getln.c b/src/ex_getln.c
--- a/src/ex_getln.c
+++ b/src/ex_getln.c
@@ -1183,6 +1183,7 @@ cmdline_insert_reg(int *gotesc UNUSED)
 {
     int		i;
     int		c;
+    int		literally = FALSE;
 #ifdef FEAT_EVAL
     int		save_new_cmdpos = new_cmdpos;
 #endif
@@ -1220,7 +1221,8 @@ cmdline_insert_reg(int *gotesc UNUSED)
 #endif
     if (c != ESC)	    // use ESC to cancel inserting register
     {
-	cmdline_paste(c, i == Ctrl_R, FALSE);
+	literally = i == Ctrl_R;
+	cmdline_paste(c, literally, FALSE);
 
 #ifdef FEAT_EVAL
 	// When there was a serious error abort getting the
@@ -1252,7 +1254,7 @@ cmdline_insert_reg(int *gotesc UNUSED)
     redrawcmd();
 
     // The text has been stuffed, the command line didn't change yet.
-    return CMDLINE_NOT_CHANGED;
+    return literally ? CMDLINE_NOT_CHANGED : CMDLINE_CHANGED;
 }
 
 /*
@@ -2082,7 +2084,12 @@ getcmdline_int(
 	case Ctrl_R:			// insert register
 		res = cmdline_insert_reg(&gotesc);
 		if (res == CMDLINE_NOT_CHANGED)
+		{
+#ifdef FEAT_SEARCH_EXTRA
+		    is_state.incsearch_postponed = TRUE;
+#endif
 		    goto cmdline_not_changed;
+		}
 		else if (res == GOTO_NORMAL_MODE)
 		    goto returncmd;
 		goto cmdline_changed;
