# HG changeset patch
# Parent  f8fe6495a290331be16b44f4b1f0f197f4485b4e

diff --git a/src/testdir/test_lambda.vim b/src/testdir/test_lambda.vim
--- a/src/testdir/test_lambda.vim
+++ b/src/testdir/test_lambda.vim
@@ -235,12 +235,15 @@ function! Test_lambda_with_timer()
   endfunction
 
   call s:Foo()
-  sleep 100ms
+  sleep 200ms
   " do not collect lambda nor l:n in lambda
-  call garbagecollect_for_testing()
-  sleep 100ms
+  call test_garbagecollect_now()
+  let m = s:n
+  sleep 200ms
   call timer_stop(s:timer_id)
-  call assert_true(s:n > 3)
+  call assert_true(m > 1)
+  call assert_true(s:n > m + 1)
+  call assert_true(s:n < 9)
 endfunction
 
 function! Test_lambda_with_partial()
