# HG changeset patch
# Parent  4cd6877122a8478aadda47712c604437e7fa1b92

diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -987,7 +987,7 @@ GvimExt/gvimext.dll: GvimExt/gvimext.cpp
 	$(MAKE) -C GvimExt -f Make_ming.mak CROSS=$(CROSS) CROSS_COMPILE=$(CROSS_COMPILE) CXX='$(CXX)' STATIC_STDCPLUS=$(STATIC_STDCPLUS)
 
 tags: notags
-	$(CTAGS) *.c *.cpp *.h
+	$(CTAGS) *.c *.cpp libvterm/src/*.c xdiff/*.c *.h libvterm/include/*.h libvterm/src/*.h xdiff/*.h
 ifdef PERL
 	$(CTAGS) --append=yes auto$(DIRSLASH)if_perl.c
 endif
diff --git a/src/Make_mvc.mak b/src/Make_mvc.mak
--- a/src/Make_mvc.mak
+++ b/src/Make_mvc.mak
@@ -1272,7 +1272,7 @@ GvimExt/gvimext.dll: GvimExt/gvimext.cpp
 
 
 tags: notags
-	$(CTAGS) *.c *.cpp *.h
+	$(CTAGS) *.c *.cpp libvterm/src/*.c xdiff/*.c *.h libvterm/include/*.h libvterm/src/*.h xdiff/*.h
 	if exist auto\if_perl.c $(CTAGS) --append=yes auto\if_perl.c
 
 notags:
