# HG changeset patch
# Parent 7533dabf15ef212e18c9666fa197b7f3e408e056

diff --git a/src/version.h b/src/version.h
--- a/src/version.h
+++ b/src/version.h
@@ -38,3 +38,5 @@
 #define VIM_VERSION_MEDIUM	"7.4a BETA"
 #define VIM_VERSION_LONG	"VIM - Vi IMproved 7.4a BETA (2013 Jul 6)"
 #define VIM_VERSION_LONG_DATE	"VIM - Vi IMproved 7.4a BETA (2013 Jul 6, compiled "
+
+#define VIM_VERSION_COPYRIGHT_YEAR	"1996-2013"
diff --git a/src/vim.rc b/src/vim.rc
--- a/src/vim.rc
+++ b/src/vim.rc
@@ -91,7 +91,7 @@
 	VALUE "FileDescription",	"Vi Improved - A Text Editor\0"
 	VALUE "FileVersion",		VIM_VERSION_MAJOR_STR ", " VIM_VERSION_MINOR_STR ", " VIM_VERSION_BUILD_STR ", " VIM_VERSION_PATCHLEVEL_STR  "\0"
 	VALUE "InternalName",		"VIM\0"
-	VALUE "LegalCopyright",		"Copyright \251 1996-2008\0"
+	VALUE "LegalCopyright",		"Copyright \251 " VIM_VERSION_COPYRIGHT_YEAR "\0"
 	VALUE "LegalTrademarks",	"Vim\0"
 	VALUE "OriginalFilename",	"VIM.EXE\0"
 	VALUE "ProductName",		"Vim\0"
diff --git a/src/vim16.rc b/src/vim16.rc
--- a/src/vim16.rc
+++ b/src/vim16.rc
@@ -67,7 +67,7 @@
 	VALUE "FileDescription",	"Vi Improved - A Text Editor\0"
 	VALUE "FileVersion",		VIM_VERSION_MAJOR_STR ", " VIM_VERSION_MINOR_STR ", " VIM_VERSION_BUILD_STR ", " VIM_VERSION_PATCHLEVEL_STR  "\0"
 	VALUE "InternalName",		"VIM\0"
-	VALUE "LegalCopyright",		"Copyright \251 1996-2005\0"
+	VALUE "LegalCopyright",		"Copyright \251 " VIM_VERSION_COPYRIGHT_YEAR "\0"
 	VALUE "LegalTrademarks",	"Vim\0"
 	VALUE "OriginalFilename",	"VIM.EXE\0"
 	VALUE "ProductName",		"Vim\0"
