# HG changeset patch
# Parent  57f4e4f7840802cfb41c02c0dc255ba535ad4d9f

diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -156,10 +156,10 @@ endif
 #	Perl interface:
 #	  PERL=[Path to Perl directory] (Set inside Make_cyg.mak or Make_ming.mak)
 #	  DYNAMIC_PERL=yes (to load the Perl DLL dynamically)
-#	  PERL_VER=[Perl version, eg 56, 58, 510] (default is 56)
+#	  PERL_VER=[Perl version, eg 56, 58, 510] (default is 524)
 ifdef PERL
 ifndef PERL_VER
-PERL_VER=56
+PERL_VER=524
 endif
 ifndef DYNAMIC_PERL
 DYNAMIC_PERL=yes
@@ -188,14 +188,14 @@ endif
 #	Lua interface:
 #	  LUA=[Path to Lua directory] (Set inside Make_cyg.mak or Make_ming.mak)
 #	  DYNAMIC_LUA=yes (to load the Lua DLL dynamically)
-#	  LUA_VER=[Lua version, eg 51, 52] (default is 51)
+#	  LUA_VER=[Lua version, eg 51, 52] (default is 53)
 ifdef LUA
 ifndef DYNAMIC_LUA
 DYNAMIC_LUA=yes
 endif
 
 ifndef LUA_VER
-LUA_VER=51
+LUA_VER=53
 endif
 
 ifeq (no,$(DYNAMIC_LUA))
@@ -207,7 +207,7 @@ endif
 #	MzScheme interface:
 #	  MZSCHEME=[Path to MzScheme directory] (Set inside Make_cyg.mak or Make_ming.mak)
 #	  DYNAMIC_MZSCHEME=yes (to load the MzScheme DLL dynamically)
-#	  MZSCHEME_VER=[MzScheme version] (default is 205_000)
+#	  MZSCHEME_VER=[MzScheme version] (default is 3m_a0solc (6.6))
 #	  MZSCHEME_DEBUG=no
 ifdef MZSCHEME
 ifndef DYNAMIC_MZSCHEME
@@ -215,7 +215,7 @@ DYNAMIC_MZSCHEME=yes
 endif
 
 ifndef MZSCHEME_VER
-MZSCHEME_VER=205_000
+MZSCHEME_VER=3m_a0solc
 endif
 
 # for version 4.x we need to generate byte-code for Scheme base
@@ -263,7 +263,7 @@ endif
 #	Python interface:
 #	  PYTHON=[Path to Python directory] (Set inside Make_cyg.mak or Make_ming.mak)
 #	  DYNAMIC_PYTHON=yes (to load the Python DLL dynamically)
-#	  PYTHON_VER=[Python version, eg 22, 23, ..., 27] (default is 22)
+#	  PYTHON_VER=[Python version, eg 22, 23, ..., 27] (default is 27)
 ifdef PYTHON
 ifndef DYNAMIC_PYTHON
 DYNAMIC_PYTHON=yes
@@ -296,14 +296,14 @@ endif
 #	Python3 interface:
 #	  PYTHON3=[Path to Python3 directory] (Set inside Make_cyg.mak or Make_ming.mak)
 #	  DYNAMIC_PYTHON3=yes (to load the Python3 DLL dynamically)
-#	  PYTHON3_VER=[Python3 version, eg 31, 32] (default is 31)
+#	  PYTHON3_VER=[Python3 version, eg 31, 32] (default is 35)
 ifdef PYTHON3
 ifndef DYNAMIC_PYTHON3
 DYNAMIC_PYTHON3=yes
 endif
 
 ifndef PYTHON3_VER
-PYTHON3_VER=31
+PYTHON3_VER=35
 endif
 ifndef DYNAMIC_PYTHON3_DLL
 DYNAMIC_PYTHON3_DLL=python$(PYTHON3_VER).dll
@@ -328,18 +328,18 @@ endif
 #	TCL interface:
 #	  TCL=[Path to TCL directory] (Set inside Make_cyg.mak or Make_ming.mak)
 #	  DYNAMIC_TCL=yes (to load the TCL DLL dynamically)
-#	  TCL_VER=[TCL version, eg 83, 84] (default is 83)
-#	  TCL_VER_LONG=[Tcl version, eg 8.3] (default is 8.3)
+#	  TCL_VER=[TCL version, eg 83, 84] (default is 86)
+#	  TCL_VER_LONG=[Tcl version, eg 8.3] (default is 8.6)
 #	    You must set TCL_VER_LONG when you set TCL_VER.
 ifdef TCL
 ifndef DYNAMIC_TCL
 DYNAMIC_TCL=yes
 endif
 ifndef TCL_VER
-TCL_VER = 83
+TCL_VER = 86
 endif
 ifndef TCL_VER_LONG
-TCL_VER_LONG = 8.3
+TCL_VER_LONG = 8.6
 endif
 TCLINC += -I$(TCL)/include
 endif
diff --git a/src/Make_mvc.mak b/src/Make_mvc.mak
--- a/src/Make_mvc.mak
+++ b/src/Make_mvc.mak
@@ -39,12 +39,12 @@
 #	Lua interface:
 #	  LUA=[Path to Lua directory]
 #	  DYNAMIC_LUA=yes (to load the Lua DLL dynamically)
-#	  LUA_VER=[Lua version]  (default is 51)
+#	  LUA_VER=[Lua version]  (default is 53)
 #
 #	MzScheme interface:
 #	  MZSCHEME=[Path to MzScheme directory]
 #	  DYNAMIC_MZSCHEME=yes (to load the MzScheme DLLs dynamically)
-#	  MZSCHEME_VER=[version, 205_000, ...]
+#	  MZSCHEME_VER=[version, 205_000, ...] (default is 3m_a0solc (6.6))
 #	  MZSCHEME_DEBUG=no
 #
 #	Perl interface:
@@ -52,17 +52,17 @@
 #	  DYNAMIC_PERL=yes (to load the Perl DLL dynamically)
 #	  PERL_VER=[Perl version, in the form 55 (5.005), 56 (5.6.x),
 #		    510 (5.10.x), etc]
-#	  (default is 56)
+#	  (default is 524)
 #
 #	Python interface:
 #	  PYTHON=[Path to Python directory]
 #	  DYNAMIC_PYTHON=yes (to load the Python DLL dynamically)
-#	  PYTHON_VER=[Python version, eg 22, 23, ..., 27]  (default is 22)
+#	  PYTHON_VER=[Python version, eg 22, 23, ..., 27]  (default is 27)
 #
 #	Python3 interface:
 #	  PYTHON3=[Path to Python3 directory]
 #	  DYNAMIC_PYTHON3=yes (to load the Python3 DLL dynamically)
-#	  PYTHON3_VER=[Python3 version, eg 30, 31]  (default is 31)
+#	  PYTHON3_VER=[Python3 version, eg 30, 31]  (default is 35)
 #
 #	Ruby interface:
 #	  RUBY=[Path to Ruby directory]
@@ -78,8 +78,8 @@
 #	Tcl interface:
 #	  TCL=[Path to Tcl directory]
 #	  DYNAMIC_TCL=yes (to load the Tcl DLL dynamically)
-#	  TCL_VER=[Tcl version, e.g. 80, 83]  (default is 83)
-#	  TCL_VER_LONG=[Tcl version, eg 8.3] (default is 8.3)
+#	  TCL_VER=[Tcl version, e.g. 80, 83]  (default is 86)
+#	  TCL_VER_LONG=[Tcl version, eg 8.3] (default is 8.6)
 #	    You must set TCL_VER_LONG when you set TCL_VER.
 #
 #	Cscope support: CSCOPE=yes
@@ -694,8 +694,8 @@ CFLAGS = $(CFLAGS) -DDYNAMIC_GETTEXT
 # TCL interface
 !ifdef TCL
 !ifndef TCL_VER
-TCL_VER = 83
-TCL_VER_LONG = 8.3
+TCL_VER = 86
+TCL_VER_LONG = 8.6
 !endif
 !message Tcl requested (version $(TCL_VER)) - root dir is "$(TCL)"
 !if "$(DYNAMIC_TCL)" == "yes"
@@ -717,7 +717,7 @@ TCL_LIB = $(TCL)\lib\tcl$(TCL_VER)vc.lib
 # Lua interface
 !ifdef LUA
 !ifndef LUA_VER
-LUA_VER = 51
+LUA_VER = 53
 !endif
 !message Lua requested (version $(LUA_VER)) - root dir is "$(LUA)"
 !if "$(DYNAMIC_LUA)" == "yes"
@@ -745,7 +745,7 @@ DYNAMIC_PYTHON3=yes
 # PYTHON interface
 !ifdef PYTHON
 !ifndef PYTHON_VER
-PYTHON_VER = 22
+PYTHON_VER = 27
 !endif
 !message Python requested (version $(PYTHON_VER)) - root dir is "$(PYTHON)"
 !if "$(DYNAMIC_PYTHON)" == "yes"
@@ -766,7 +766,7 @@ PYTHON_LIB = $(PYTHON)\libs\python$(PYTH
 # PYTHON3 interface
 !ifdef PYTHON3
 !ifndef PYTHON3_VER
-PYTHON3_VER = 31
+PYTHON3_VER = 35
 !endif
 !message Python3 requested (version $(PYTHON3_VER)) - root dir is "$(PYTHON3)"
 !if "$(DYNAMIC_PYTHON3)" == "yes"
@@ -788,7 +788,7 @@ PYTHON3_LIB = $(PYTHON3)\libs\python$(PY
 !ifdef MZSCHEME
 !message MzScheme requested - root dir is "$(MZSCHEME)"
 !ifndef MZSCHEME_VER
-MZSCHEME_VER = 205_000
+MZSCHEME_VER = 3m_a0solc
 !endif
 !ifndef MZSCHEME_COLLECTS
 MZSCHEME_COLLECTS=$(MZSCHEME)\collects
@@ -846,7 +846,7 @@ MZSCHEME_LIB = $(MZSCHEME_LIB) /STACK:83
 # Perl interface
 !ifdef PERL
 !ifndef PERL_VER
-PERL_VER = 56
+PERL_VER = 524
 !endif
 !message Perl requested (version $(PERL_VER)) - root dir is "$(PERL)"
 !if "$(DYNAMIC_PERL)" == "yes"
