# HG changeset patch
# Parent  313df90b2130b7a0219c053e3cf059a208641da9

diff --git a/src/po/Make_cyg.mak b/src/po/Make_cyg.mak
--- a/src/po/Make_cyg.mak
+++ b/src/po/Make_cyg.mak
@@ -12,7 +12,8 @@
 VIMRUNTIME = ../../runtime
 endif
 
-LANGUAGES =	af \
+LANGUAGES = \
+		af \
 		ca \
 		cs \
 		cs.cp1250 \
@@ -25,12 +26,16 @@
 		ga \
 		it \
 		ja \
+		ja.euc-jp \
 		ja.sjis \
 		ko \
 		ko.UTF-8 \
+		nb \
+		nl \
 		no \
 		pl \
 		pl.cp1250 \
+		pl.UTF-8 \
 		pt_BR \
 		ru \
 		ru.cp1251 \
@@ -41,12 +46,13 @@
 		uk.cp1251 \
 		vi \
 		zh_CN \
+		zh_CN.cp936 \
 		zh_CN.UTF-8 \
-		zh_CN.cp936 \
 		zh_TW \
 		zh_TW.UTF-8 \
 
-MOFILES =	af.mo \
+MOFILES = \
+		af.mo \
 		ca.mo \
 		cs.cp1250.mo \
 		cs.mo \
@@ -58,13 +64,17 @@
 		fr.mo \
 		ga.mo \
 		it.mo \
+		ja.euc-jp.mo \
 		ja.mo \
 		ja.sjis.mo \
 		ko.mo \
 		ko.UTF-8.mo \
+		nb.mo \
+		nl.mo \
 		no.mo \
 		pl.cp1250.mo \
 		pl.mo \
+		pl.UTF-8.mo \
 		pt_BR.mo \
 		ru.cp1251.mo \
 		ru.mo \
@@ -74,11 +84,11 @@
 		uk.cp1251.mo \
 		uk.mo \
 		vi.mo \
+		zh_CN.mo \
+		zh_CN.cp936.mo \
 		zh_CN.UTF-8.mo \
-		zh_CN.cp936.mo \
-		zh_CN.mo \
+		zh_TW.mo \
 		zh_TW.UTF-8.mo \
-		zh_TW.mo \
 
 PACKAGE = vim
 
@@ -135,6 +145,8 @@
 		$(CP) $$TARGET.mo $(VIMRUNTIME)/lang/$$TARGET/LC_MESSAGES/$(PACKAGE).mo ; \
 	done
 
+install-all: install
+
 clean:
 	$(RM) *.mo
 	$(RM) *.pot
diff --git a/src/po/Make_ming.mak b/src/po/Make_ming.mak
--- a/src/po/Make_ming.mak
+++ b/src/po/Make_ming.mak
@@ -10,10 +10,15 @@
 # language (xx) and add it to the next three lines.
 #
 
+ifndef VIMRUNTIME
+VIMRUNTIME = ..\..\runtime
+endif
+
 LANGUAGES = \
 		af \
 		ca \
 		cs \
+		cs.cp1250 \
 		de \
 		en_GB \
 		eo \
@@ -23,23 +28,35 @@
 		ga \
 		it \
 		ja \
+		ja.euc-jp \
+		ja.sjis \
 		ko \
+		ko.UTF-8 \
+		nb \
+		nl \
 		no \
 		pl \
+		pl.cp1250 \
+		pl.UTF-8 \
 		pt_BR \
 		ru \
+		ru.cp1251 \
 		sk \
+		sk.cp1250 \
 		sv \
 		uk \
+		uk.cp1251 \
 		vi \
 		zh_CN \
-		zh_CN.UTF-8\
+		zh_CN.cp936 \
+		zh_CN.UTF-8 \
 		zh_TW \
 		zh_TW.UTF-8 \
 
 MOFILES = \
 		af.mo \
 		ca.mo \
+		cs.cp1250.mo \
 		cs.mo \
 		de.mo \
 		en_GB.mo \
@@ -49,20 +66,31 @@
 		fr.mo \
 		ga.mo \
 		it.mo \
+		ja.euc-jp.mo \
 		ja.mo \
+		ja.sjis.mo \
 		ko.mo \
+		ko.UTF-8.mo \
+		nb.mo \
+		nl.mo \
 		no.mo \
+		pl.cp1250.mo \
 		pl.mo \
+		pl.UTF-8.mo \
 		pt_BR.mo \
+		ru.cp1251.mo \
 		ru.mo \
+		sk.cp1250.mo \
 		sk.mo \
 		sv.mo \
+		uk.cp1251.mo \
 		uk.mo \
 		vi.mo \
+		zh_CN.mo \
+		zh_CN.cp936.mo \
 		zh_CN.UTF-8.mo \
-		zh_CN.mo \
+		zh_TW.mo \
 		zh_TW.UTF-8.mo \
-		zh_TW.mo \
 
 PACKAGE = vim
 
@@ -72,9 +100,9 @@
 #GETTEXT_PATH = C:/gettext-0.10.35-w32/win32/Release/
 #GETTEXT_PATH = C:/cygwin/bin/
 
-MSGFMT = $(GETTEXT_PATH)msgfmt
-XGETTEXT = $(GETTEXT_PATH)xgettext
-MSGMERGE = $(GETTEXT_PATH)msgmerge
+MSGFMT = set OLD_PO_FILE_INPUT=yes && $(GETTEXT_PATH)msgfmt -v
+XGETTEXT = set OLD_PO_FILE_INPUT=yes && set OLD_PO_FILE_OUTPUT=yes && $(GETTEXT_PATH)xgettext
+MSGMERGE = set OLD_PO_FILE_INPUT=yes && set OLD_PO_FILE_OUTPUT=yes && $(GETTEXT_PATH)msgmerge
 
 MV = move
 CP = copy
@@ -108,6 +136,11 @@
 	$(MKD) $(VIMRUNTIME)\lang\$(LANGUAGE)\LC_MESSAGES
 	$(CP) $(LANGUAGE).mo $(VIMRUNTIME)\lang\$(LANGUAGE)\LC_MESSAGES\$(PACKAGE).mo
 
+install-all: all
+	FOR %%l IN ($(LANGUAGES)) DO @IF NOT EXIST $(VIMRUNTIME)\lang\%%l $(MKD) $(VIMRUNTIME)\lang\%%l
+	FOR %%l IN ($(LANGUAGES)) DO @IF NOT EXIST $(VIMRUNTIME)\lang\%%l\LC_MESSAGES $(MKD) $(VIMRUNTIME)\lang\%%l\LC_MESSAGES
+	FOR %%l IN ($(LANGUAGES)) DO @$(CP) %%l.mo $(VIMRUNTIME)\lang\%%l\LC_MESSAGES\$(PACKAGE).mo
+
 clean:
 	$(RM) *.mo
 	$(RM) *.pot
diff --git a/src/po/Make_mvc.mak b/src/po/Make_mvc.mak
--- a/src/po/Make_mvc.mak
+++ b/src/po/Make_mvc.mak
@@ -6,10 +6,15 @@
 # Please read README_mvc.txt before using this file.
 #
 
+!ifndef VIMRUNTIME
+VIMRUNTIME = ..\..\runtime
+!endif
+
 LANGUAGES = \
 		af \
 		ca \
 		cs \
+		cs.cp1250 \
 		de \
 		en_GB \
 		eo \
@@ -19,16 +24,27 @@
 		ga \
 		it \
 		ja \
+		ja.euc-jp \
+		ja.sjis \
 		ko \
+		ko.UTF-8 \
+		nb \
+		nl \
 		no \
 		pl \
+		pl.cp1250 \
+		pl.UTF-8 \
 		pt_BR \
 		ru \
+		ru.cp1251 \
 		sk \
+		sk.cp1250 \
 		sv \
 		uk \
+		uk.cp1251 \
 		vi \
 		zh_CN \
+		zh_CN.cp936 \
 		zh_CN.UTF-8 \
 		zh_TW \
 		zh_TW.UTF-8 \
@@ -36,6 +52,7 @@
 MOFILES = \
 		af.mo \
 		ca.mo \
+		cs.cp1250.mo \
 		cs.mo \
 		de.mo \
 		en_GB.mo \
@@ -45,27 +62,38 @@
 		fr.mo \
 		ga.mo \
 		it.mo \
+		ja.euc-jp.mo \
 		ja.mo \
+		ja.sjis.mo \
 		ko.mo \
+		ko.UTF-8.mo \
+		nb.mo \
+		nl.mo \
 		no.mo \
+		pl.cp1250.mo \
 		pl.mo \
+		pl.UTF-8.mo \
 		pt_BR.mo \
+		ru.cp1251.mo \
 		ru.mo \
+		sk.cp1250.mo \
 		sk.mo \
 		sv.mo \
+		uk.cp1251.mo \
 		uk.mo \
 		vi.mo \
+		zh_CN.mo \
+		zh_CN.cp936.mo \
 		zh_CN.UTF-8.mo \
-		zh_CN.mo \
+		zh_TW.mo \
 		zh_TW.UTF-8.mo \
-		zh_TW.mo \
 
 PACKAGE = vim
 
 # Correct the following line for the directory where gettext et al is installed
 GETTEXT_PATH = H:\gettext.0.14.4\bin
 
-MSGFMT = $(GETTEXT_PATH)\msgfmt
+MSGFMT = $(GETTEXT_PATH)\msgfmt -v
 XGETTEXT = $(GETTEXT_PATH)\xgettext
 MSGMERGE = $(GETTEXT_PATH)\msgmerge
 
@@ -83,6 +111,7 @@
 .SUFFIXES: .po .mo .pot
 
 .po.mo:
+	set OLD_PO_FILE_INPUT=yes
 	$(MSGFMT) -o $@ $<
 
 all: $(MOFILES)
@@ -91,9 +120,13 @@
 	$(LS) $(LSFLAGS) ..\*.c ..\if_perl.xs ..\globals.h > .\files
 
 first_time: files
+	set OLD_PO_FILE_INPUT=yes
+	set OLD_PO_FILE_OUTPUT=yes
 	$(XGETTEXT) --default-domain=$(LANGUAGE) --add-comments --keyword=_ --keyword=N_ --files-from=.\files
 
 $(LANGUAGES): files
+	set OLD_PO_FILE_INPUT=yes
+	set OLD_PO_FILE_OUTPUT=yes
 	$(XGETTEXT) --default-domain=$(PACKAGE) --add-comments --keyword=_ --keyword=N_ --files-from=.\files
 	$(MV) $(PACKAGE).po $(PACKAGE).pot
 	$(CP) $@.po $@.po.orig
@@ -105,6 +138,10 @@
 	if not exist $(INSTALLDIR) $(MKD) $(INSTALLDIR)
 	$(CP) $(LANGUAGE).mo $(INSTALLDIR)\$(PACKAGE).mo
 
+install-all: all
+	FOR %%l IN ($(LANGUAGES)) DO @IF NOT EXIST $(VIMRUNTIME)\lang\%%l\LC_MESSAGES $(MKD) $(VIMRUNTIME)\lang\%%l\LC_MESSAGES
+	FOR %%l IN ($(LANGUAGES)) DO @$(CP) %%l.mo $(VIMRUNTIME)\lang\%%l\LC_MESSAGES\$(PACKAGE).mo
+
 clean:
 	$(RM) *.mo
 	$(RM) *.pot
diff --git a/src/po/README_mingw.txt b/src/po/README_mingw.txt
--- a/src/po/README_mingw.txt
+++ b/src/po/README_mingw.txt
@@ -86,6 +86,9 @@
 
     This will create the xx\LC_MESSAGES directory (if it does not exist) and
     will copy vim.po to it.
+    You can also use the following command to install all languages:
+
+	make -f Make_ming.mak install-all
 
 (6) Another extra step ;-)). The command:
 
diff --git a/src/po/README_mvc.txt b/src/po/README_mvc.txt
--- a/src/po/README_mvc.txt
+++ b/src/po/README_mvc.txt
@@ -111,6 +111,9 @@
 
 This will create the xx\LC_MESSAGES directory in runtime\lang if it does not
 already exist.
+You can also use the following command to install all languages:
+
+	nmake -f make_mvc.mak install-all
 
 
 vim:tw=78:
