# HG changeset patch
# Parent  b24f11865a5d39e4659ad0a417e928b461d45022

diff --git a/runtime/indent/teraterm.vim b/runtime/indent/teraterm.vim
--- a/runtime/indent/teraterm.vim
+++ b/runtime/indent/teraterm.vim
@@ -1,9 +1,9 @@
 " Vim indent file
 " Language:	Tera Term Language (TTL)
-"		Based on Tera Term Version 4.86
+"		Based on Tera Term Version 4.92
 " Maintainer:	Ken Takata
 " URL:		https://github.com/k-takata/vim-teraterm
-" Last Change:	2015 Jun 4
+" Last Change:	2016 Aug 17
 " Filenames:	*.ttl
 " License:	VIM License
 
@@ -25,9 +25,7 @@ endif
 " The shiftwidth() function is relatively new.
 " Don't require it to exist.
 if exists('*shiftwidth')
-  function s:sw() abort
-    return shiftwidth()
-  endfunction
+  let s:sw = function('shiftwidth')
 else
   function s:sw() abort
     return &shiftwidth
@@ -48,7 +46,7 @@ function! GetTeraTermIndent(lnum)
 
   let l:ind = l:previ
 
-  if l:prevl =~ '^\s*if\>.*\<then\s*$'
+  if l:prevl =~ '^\s*if\>.*\<then\>'
     " previous line opened a block
     let l:ind += s:sw()
   endif
diff --git a/runtime/syntax/teraterm.vim b/runtime/syntax/teraterm.vim
--- a/runtime/syntax/teraterm.vim
+++ b/runtime/syntax/teraterm.vim
@@ -1,9 +1,9 @@
 " Vim syntax file
 " Language:	Tera Term Language (TTL)
-"		Based on Tera Term Version 4.86
+"		Based on Tera Term Version 4.92
 " Maintainer:	Ken Takata
 " URL:		https://github.com/k-takata/vim-teraterm
-" Last Change:	2015 Jun 24
+" Last Change:	2016 Aug 17
 " Filenames:	*.ttl
 " License:	VIM License
 
@@ -33,7 +33,7 @@ syn keyword ttlOperator	and or xor not
 
 syn match ttlVar	"\<groupmatchstr\d\>"
 syn match ttlVar	"\<param\d\>"
-syn keyword ttlVar	inputstr matchstr paramcnt result timeout mtimeout
+syn keyword ttlVar	inputstr matchstr paramcnt params result timeout mtimeout
 
 
 syn match ttlLine nextgroup=ttlStatement "^"
