# HG changeset patch
# Parent 7ac5bdfc9ac4edc2e3828579fc05d7f0d4a4f253
diff --git a/src/gui_w48.c b/src/gui_w48.c
--- a/src/gui_w48.c
+++ b/src/gui_w48.c
@@ -3476,10 +3476,9 @@
     filterp = convert_filterW(filter);
 
     vim_memset(&fileStruct, 0, sizeof(OPENFILENAMEW));
-#ifdef OPENFILENAME_SIZE_VERSION_400
+#ifdef OPENFILENAME_SIZE_VERSION_400W
     /* be compatible with Windows NT 4.0 */
-    /* TODO: what to use for OPENFILENAMEW??? */
-    fileStruct.lStructSize = OPENFILENAME_SIZE_VERSION_400;
+    fileStruct.lStructSize = OPENFILENAME_SIZE_VERSION_400W;
 #else
     fileStruct.lStructSize = sizeof(fileStruct);
 #endif
