# HG changeset patch
# Parent 703feace0ed867dc5d45a31f088ee387a0e0bfa9
diff --git a/src/gui_w32.c b/src/gui_w32.c
--- a/src/gui_w32.c
+++ b/src/gui_w32.c
@@ -50,7 +50,7 @@
 }
 #endif
 
-#ifdef FEAT_RENDERING_OPTIONS
+#if defined(FEAT_RENDERING_OPTIONS) || defined(PROTO)
     int
 gui_mch_set_rendering_options(char_u *s)
 {
diff --git a/src/option.c b/src/option.c
--- a/src/option.c
+++ b/src/option.c
@@ -6996,8 +6996,6 @@
 #if defined(FEAT_GUI) && defined(FEAT_RENDERING_OPTIONS)
     else if (varp == &p_renopt && gui.in_use)
     {
-	int gui_mch_set_rendering_options(char_u *);
-
 	if (!gui_mch_set_rendering_options(p_renopt))
 	    errmsg = e_invarg;
     }
diff --git a/src/proto/gui_w32.pro b/src/proto/gui_w32.pro
--- a/src/proto/gui_w32.pro
+++ b/src/proto/gui_w32.pro
@@ -1,4 +1,5 @@
 /* gui_w32.c */
+int gui_mch_set_rendering_options __ARGS((char_u *s));
 void gui_mch_set_blinking __ARGS((long wait, long on, long off));
 void gui_mch_stop_blink __ARGS((void));
 void gui_mch_start_blink __ARGS((void));
