# HG changeset patch
# Parent  1701a0eefce10947d99b6d35d3a44f4d5fe26e3d

diff --git a/src/os_win32.c b/src/os_win32.c
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -5399,6 +5399,56 @@ write_chars(
     return written;
 }
 
+#ifdef FEAT_MBYTE
+    static BOOL
+write_charsW(
+    LPWSTR pchBuf,
+    DWORD  cchToWrite)
+{
+    COORD coord = g_coord;
+    DWORD written;
+    DWORD n;
+    int ch, ch2;
+
+    FillConsoleOutputAttribute(g_hConOut, g_attrCurrent, cchToWrite,
+				coord, &written);
+    /* When writing fails or didn't write a single character, pretend one
+     * character was written, otherwise we get stuck. */
+    if (WriteConsoleOutputCharacterW(g_hConOut, pchBuf, cchToWrite,
+				coord, &written) == 0
+	    || written == 0)
+	written = 1;
+
+    for (n = 0; n < written; n++, pchBuf++)
+    {
+	ch = *pchBuf;
+	if (ch >= 0xD800 && ch <= 0xDBFF && written - n > 1)
+	{
+	    /* surrogate pairs handling */
+	    ch2 = pchBuf[1];
+	    if (ch2 >= 0xDC00 && ch2 <= 0xDFFF)
+	    {
+		ch = ((ch - 0xD800) << 10) + (ch2 & 0x3FF) + 0x10000;
+		pchBuf++;
+		n++;
+	    }
+	}
+	g_coord.X += (SHORT) utf_char2cells(ch);
+    }
+
+    while (g_coord.X > g_srScrollRegion.Right)
+    {
+	g_coord.X -= (SHORT) Columns;
+	if (g_coord.Y < g_srScrollRegion.Bottom)
+	    g_coord.Y++;
+    }
+
+    gotoxy(g_coord.X + 1, g_coord.Y + 1);
+
+    return written;
+}
+#endif
+
 
 /*
  * mch_write(): write the output buffer to the screen, translating ESC
@@ -5435,7 +5485,21 @@ mch_write(
 	{
 	    DWORD nWritten;
 
-	    nWritten = write_chars(s, prefix);
+#ifdef FEAT_MBYTE
+	    if (enc_utf8)
+	    {
+		WCHAR	*ret;
+		int	length;
+		MultiByteToWideChar_alloc(CP_UTF8, 0, (LPCSTR)s, prefix,
+						           &ret, &length);
+
+		write_charsW(ret, length);
+		nWritten = prefix;
+		vim_free(ret);
+	    }
+	    else
+#endif
+		nWritten = write_chars(s, prefix);
 #ifdef MCH_WRITE_DUMP
 	    if (fdDump)
 	    {
@@ -5663,7 +5727,20 @@ mch_write(
 	    /* Write a single character */
 	    DWORD nWritten;
 
-	    nWritten = write_chars(s, 1);
+#ifdef FEAT_MBYTE
+	    if (enc_utf8)
+	    {
+		WCHAR	*ret;
+		int	length;
+		MultiByteToWideChar_alloc(CP_UTF8, 0, (LPCSTR)s, 1,
+						           &ret, &length);
+
+		nWritten = write_charsW(ret, 1);
+		vim_free(ret);
+	    }
+	    else
+#endif
+		nWritten = write_chars(s, 1);
 #ifdef MCH_WRITE_DUMP
 	    if (fdDump)
 	    {
