# HG changeset patch
# Parent  65c55a798d2b662e118a4d017464e79193ad28e2

diff --git a/src/os_win32.c b/src/os_win32.c
--- a/src/os_win32.c
+++ b/src/os_win32.c
@@ -6385,15 +6385,13 @@ mch_write(
     char_u  *s,
     int	    len)
 {
+    char_u  *end = s + len;
+
 # ifdef VIMDLL
     if (gui.in_use)
 	return;
 # endif
 
-    // Avoid writing to a string literal.
-    if (s[len] != NUL)
-	s[len] = NUL;
-
     if (!term_console)
     {
 	write(1, s, (unsigned)len);
@@ -6414,10 +6412,13 @@ mch_write(
 	    return;
 	}
 
-	while ((ch = s[++prefix]))
+	while (s + ++prefix < end)
+	{
+	    ch = s[prefix];
 	    if (ch <= 0x1e && !(ch != '\n' && ch != '\r' && ch != '\b'
 						&& ch != '\a' && ch != '\033'))
 		break;
+	}
 
 	if (p_wd)
 	{
