# HG changeset patch
# Parent  d49f3d183891da8042315398f83943527c5b7db0

diff --git a/src/Make_cyg_ming.mak b/src/Make_cyg_ming.mak
--- a/src/Make_cyg_ming.mak
+++ b/src/Make_cyg_ming.mak
@@ -74,7 +74,11 @@ else
 CHANNEL=$(GUI)
 endif
 # Set to yes to enable terminal support.
+ifeq (HUGE, $(FEATURES))
+TERMINAL=yes
+else
 TERMINAL=no
+endif
 
 ifndef CTAGS
 # this assumes ctags is Exuberant ctags
diff --git a/src/Make_mvc.mak b/src/Make_mvc.mak
--- a/src/Make_mvc.mak
+++ b/src/Make_mvc.mak
@@ -37,7 +37,7 @@
 #	  is yes)
 #	Global IME support: GIME=yes (requires GUI=yes)
 #
-#       Terminal support: TERMINAL=yes (default is no)
+#       Terminal support: TERMINAL=yes (default is yes)
 #
 #	Lua interface:
 #	  LUA=[Path to Lua directory]
@@ -361,6 +361,14 @@ CSCOPE_OBJ   = $(OBJDIR)/if_cscope.obj
 CSCOPE_DEFS  = -DFEAT_CSCOPE
 !endif
 
+!ifndef TERMINAL
+!if "$(FEATURES)"=="HUGE"
+TERMINAL = yes
+!else
+TERMINAL = no
+!endif
+!endif
+
 !if "$(TERMINAL)" == "yes"
 TERM_OBJ = \
 	$(OBJDIR)/terminal.obj \
