# HG changeset patch
# Parent  9396eb445cafb390741b30f048ef5958aae8def5

diff --git a/src/terminal.c b/src/terminal.c
--- a/src/terminal.c
+++ b/src/terminal.c
@@ -2307,10 +2307,12 @@ create_vterm(term_T *term, int rows, int
 
     /* For unix do not use a blinking cursor.  In an xterm this causes the
      * cursor to blink if it's blinking in the xterm.
-     * We do want a blinking cursor by default on Windows, since that's what
-     * the default is for a console. */
+     * For Windows we respect the system wide setting. */
 #ifdef WIN3264
-    value.boolean = 1;
+    if (GetCaretBlinkTime() == INFINITE)
+	value.boolean = 0;
+    else
+	value.boolean = 1;
 #else
     value.boolean = 0;
 #endif
