# HG changeset patch
# Parent  77e7664f05c55e318cdf10de63859cefaf8da711

diff --git a/src/evalfunc.c b/src/evalfunc.c
--- a/src/evalfunc.c
+++ b/src/evalfunc.c
@@ -6002,7 +6002,7 @@ f_has(typval_T *argvars, typval_T *rettv
 #ifdef FEAT_TERMGUICOLORS
 	"termguicolors",
 #endif
-#ifdef FEAT_TERMINAL
+#if defined(FEAT_TERMINAL) && !defined(WIN3264)
 	"terminal",
 #endif
 #ifdef TERMINFO
@@ -6210,6 +6210,10 @@ f_has(typval_T *argvars, typval_T *rettv
 	else if (STRICMP(name, "netbeans_enabled") == 0)
 	    n = netbeans_active();
 #endif
+#if defined(FEAT_TERMINAL) && defined(WIN3264)
+	else if (STRICMP(name, "terminal") == 0)
+	    n = terminal_enabled();
+#endif
     }
 
     rettv->vval.v_number = n;
diff --git a/src/proto/terminal.pro b/src/proto/terminal.pro
--- a/src/proto/terminal.pro
+++ b/src/proto/terminal.pro
@@ -34,4 +34,5 @@ void f_term_scrape(typval_T *argvars, ty
 void f_term_sendkeys(typval_T *argvars, typval_T *rettv);
 void f_term_start(typval_T *argvars, typval_T *rettv);
 void f_term_wait(typval_T *argvars, typval_T *rettv);
+int terminal_enabled(void);
 /* vim: set ft=c : */
diff --git a/src/terminal.c b/src/terminal.c
--- a/src/terminal.c
+++ b/src/terminal.c
@@ -2709,6 +2709,16 @@ f_term_wait(typval_T *argvars, typval_T 
     }
 }
 
+# if defined(WIN3264) || defined(PROTO)
+static int dyn_winpty_init(int verbose);
+
+    int
+terminal_enabled(void)
+{
+    return dyn_winpty_init(FALSE);
+}
+# endif
+
 # ifdef WIN3264
 
 /**************************************
@@ -2738,8 +2748,8 @@ HANDLE (*winpty_agent_process)(void*);
 
 static HINSTANCE hWinPtyDLL = NULL;
 
-    int
-dyn_winpty_init(void)
+    static int
+dyn_winpty_init(int verbose)
 {
     int i;
     static struct
@@ -2777,7 +2787,8 @@ dyn_winpty_init(void)
 	hWinPtyDLL = vimLoadLib(WINPTY_DLL);
     if (!hWinPtyDLL)
     {
-	EMSG2(_(e_loadlib), *p_winptydll != NUL ? p_winptydll : WINPTY_DLL);
+	if (verbose)
+	    EMSG2(_(e_loadlib), *p_winptydll != NUL ? p_winptydll : WINPTY_DLL);
 	return 0;
     }
     for (i = 0; winpty_entry[i].name != NULL
@@ -2786,7 +2797,8 @@ dyn_winpty_init(void)
 	if ((*winpty_entry[i].ptr = (FARPROC)GetProcAddress(hWinPtyDLL,
 					      winpty_entry[i].name)) == NULL)
 	{
-	    EMSG2(_(e_loadfunc), winpty_entry[i].name);
+	    if (verbose)
+		EMSG2(_(e_loadfunc), winpty_entry[i].name);
 	    return 0;
 	}
     }
@@ -2813,7 +2825,7 @@ term_and_job_init(term_T *term, int rows
     garray_T	    ga;
     char_u	    *cmd;
 
-    if (!dyn_winpty_init())
+    if (!dyn_winpty_init(TRUE))
 	return FAIL;
 
     if (argvar->v_type == VAR_STRING)
