# HG changeset patch
# Parent  85a27bb6e221f7f95731584c25f9c4b07cb19930

diff --git a/src/charset.c b/src/charset.c
--- a/src/charset.c
+++ b/src/charset.c
@@ -884,6 +884,7 @@ win_linetabsize_cts(chartabsize_T *cts, 
 	cts->cts_ptr = saved_ptr;
 	cts->cts_vcol = vcol = saved_vcol;
     }
+    Trace("win_linetabsize_cts 1: vcol=%d cts_ptr-cts_line=%d", cts->cts_vcol, cts->cts_ptr - cts->cts_line);
     for ( ; *cts->cts_ptr != NUL && (len == MAXCOL || cts->cts_ptr < cts->cts_line + len);
 						      MB_PTR_ADV(cts->cts_ptr))
     {
@@ -920,6 +921,7 @@ win_linetabsize_cts(chartabsize_T *cts, 
 	cts->cts_vcol = vcol > MAXCOL ? MAXCOL : (int)vcol;
     }
 #endif
+    Trace("win_linetabsize_cts 2: vcol=%d", cts->cts_vcol);
 }
 
 /*
@@ -1792,6 +1794,7 @@ getvcol(
 	    *cursor = vcol + head;	    // cursor at start
 	}
     }
+    Trace("getvcol: start=%d cursor=%d end=%d", start ? *start : -1, cursor ? *cursor : -1, end ? *end : -1);
 }
 
 /*
