# HG changeset patch
# Parent  be075d8a8aaec7d4b6c12b11ec030e278b7dd0ef

diff --git a/src/charset.c b/src/charset.c
--- a/src/charset.c
+++ b/src/charset.c
@@ -849,13 +849,41 @@ linetabsize_no_outer(win_T *wp, linenr_T
 #endif
 }
 
+static int s_vcol_cache_valid = FALSE;
+
+    void
+invalidate_vcol_cache(void)
+{
+    s_vcol_cache_valid = FALSE;
+}
+
     void
 win_linetabsize_cts(chartabsize_T *cts, colnr_T len)
 {
+    static char_u   *old_line = NULL;
+    static char_u   *saved_ptr = NULL;
+    static colnr_T  saved_vcol = 0;
+    colnr_T slen = len;
+    colnr_T col_save;
     vimlong_T vcol = cts->cts_vcol;
+
 #ifdef FEAT_PROP_POPUP
     cts->cts_with_trailing = len == MAXCOL;
 #endif
+    if (slen == MAXCOL)
+	slen = STRLEN(cts->cts_line);
+    col_save = slen * 4 / 5;
+    if (slen - col_save > 4096)
+	col_save = slen - 4096;
+    if (s_vcol_cache_valid
+	    && (old_line != NULL)
+	    && (old_line == cts->cts_line)
+	    && (saved_ptr >= cts->cts_line)
+	    && (saved_ptr < cts->cts_line + slen))
+    {
+	cts->cts_ptr = saved_ptr;
+	cts->cts_vcol = vcol = saved_vcol;
+    }
     for ( ; *cts->cts_ptr != NUL && (len == MAXCOL || cts->cts_ptr < cts->cts_line + len);
 						      MB_PTR_ADV(cts->cts_ptr))
     {
@@ -867,6 +895,17 @@ win_linetabsize_cts(chartabsize_T *cts, 
 	}
 	else
 	    cts->cts_vcol = (int)vcol;
+
+	if ((cts->cts_ptr > cts->cts_line)
+		&& (cts->cts_ptr - cts->cts_line <= col_save)
+		&& (cts->cts_ptr + (*mb_ptr2len)(cts->cts_ptr) - cts->cts_line
+			>= col_save))
+	{
+	    saved_ptr = cts->cts_ptr;
+	    saved_vcol = vcol;
+	    old_line = cts->cts_line;
+	    s_vcol_cache_valid = TRUE;
+	}
     }
 #ifdef FEAT_PROP_POPUP
     // check for a virtual text at the end of a line or on an empty line
@@ -1569,6 +1608,10 @@ getvcol(
 #ifdef FEAT_PROP_POPUP
     int		on_NUL = FALSE;
 #endif
+    colnr_T	col_save;
+    static char_u   *old_line = NULL;
+    static char_u   *saved_ptr = NULL;
+    static colnr_T  saved_vcol = 0;
 
     vcol = 0;
     line = ptr = ml_get_buf(wp->w_buffer, pos->lnum, FALSE);
@@ -1576,6 +1619,9 @@ getvcol(
     init_chartabsize_arg(&cts, wp, pos->lnum, 0, line, line);
     cts.cts_max_head_vcol = -1;
 
+    col_save = pos->col * 4 / 5;
+    if (pos->col - col_save > 4096)
+	col_save = pos->col - 4096;
     /*
      * This function is used very often, do some speed optimizations.
      * When 'list', 'linebreak', 'showbreak' and 'breakindent' are not set
@@ -1591,6 +1637,15 @@ getvcol(
 #endif
        )
     {
+	if (s_vcol_cache_valid
+		&& (old_line != NULL)
+		&& (old_line == line)
+		&& (saved_ptr >= line)
+		&& (saved_ptr < line + pos->col))
+	{
+	    ptr = saved_ptr;
+	    vcol = saved_vcol;
+	}
 	for (;;)
 	{
 	    head = 0;
@@ -1637,12 +1692,30 @@ getvcol(
 	    if (next_ptr - line > pos->col) // character at pos->col
 		break;
 
+	    if ((ptr > line)
+		    && (ptr - line <= col_save)
+		    && (next_ptr - line >= col_save))
+	    {
+		saved_ptr = next_ptr;
+		saved_vcol = vcol + incr;
+		old_line = line;
+		s_vcol_cache_valid = TRUE;
+	    }
 	    vcol += incr;
 	    ptr = next_ptr;
 	}
     }
     else
     {
+	if (s_vcol_cache_valid
+		&& (old_line != NULL)
+		&& (old_line == line)
+		&& (saved_ptr >= line)
+		&& (saved_ptr < line + pos->col))
+	{
+	    cts.cts_ptr = saved_ptr;
+	    cts.cts_vcol = saved_vcol;
+	}
 	for (;;)
 	{
 	    // A tab gets expanded, depending on the current column.
@@ -1670,6 +1743,15 @@ getvcol(
 	    if (next_ptr - line > pos->col) // character at pos->col
 		break;
 
+	    if ((cts.cts_ptr > line)
+		    && (cts.cts_ptr - line <= col_save)
+		    && (next_ptr - line >= col_save))
+	    {
+		saved_ptr = next_ptr;
+		saved_vcol = cts.cts_vcol + incr;
+		old_line = line;
+		s_vcol_cache_valid = TRUE;
+	    }
 	    cts.cts_vcol += incr;
 	    cts.cts_ptr = next_ptr;
 	}
diff --git a/src/option.c b/src/option.c
--- a/src/option.c
+++ b/src/option.c
@@ -2751,6 +2751,7 @@ did_set_option(
 	*p = *p | P_INSECURE;
     else if (new_value)
 	*p = *p & ~P_INSECURE;
+    invalidate_vcol_cache();
 }
 
 /*
diff --git a/src/proto/charset.pro b/src/proto/charset.pro
--- a/src/proto/charset.pro
+++ b/src/proto/charset.pro
@@ -21,6 +21,7 @@ int linetabsize_col(int startcol, char_u
 int win_linetabsize(win_T *wp, linenr_T lnum, char_u *line, colnr_T len);
 int linetabsize(win_T *wp, linenr_T lnum);
 int linetabsize_no_outer(win_T *wp, linenr_T lnum);
+void invalidate_vcol_cache(void);
 void win_linetabsize_cts(chartabsize_T *cts, colnr_T len);
 int vim_isIDc(int c);
 int vim_isNormalIDc(int c);
