# HG changeset patch
# Parent  d96353ee9c1a241def5c1d638795544c7582eccf

diff --git a/nsis/gvim.nsi b/nsis/gvim.nsi
--- a/nsis/gvim.nsi
+++ b/nsis/gvim.nsi
@@ -113,7 +113,8 @@ Function .onInit
   # run the install program to check for already installed versions
   SetOutPath $TEMP
   File /oname=install.exe ${VIMSRC}\installw32.exe
-  ExecWait "$TEMP\install.exe -uninstall-check"
+  nsExec::Exec "$TEMP\install.exe -uninstall-check"
+  Pop $3
   Delete $TEMP\install.exe
 
   # We may have been put to the background when uninstall did something.
@@ -489,7 +490,8 @@ SectionEnd
 ##########################################################
 Section -call_install_exe
 	SetOutPath $0
-	ExecWait "$0\install.exe $1"
+	nsExec::Exec "$0\install.exe $1"
+	Pop $3
 SectionEnd
 
 ##########################################################
@@ -569,11 +571,13 @@ Section Uninstall
 
 	# If VisVim was installed, unregister the DLL.
 	${If} ${FileExists} "$0\VisVim.dll"
-	  ExecWait "regsvr32.exe /u /s $0\VisVim.dll"
+	  nsExec::Exec "regsvr32.exe /u /s $0\VisVim.dll"
+	  Pop $3
 	${EndIf}
 
 	# delete the context menu entry and batch files
-	ExecWait "$0\uninstal.exe -nsis"
+	nsExec::Exec "$0\uninstal.exe -nsis"
+	Pop $3
 
 	# We may have been put to the background when uninstall did something.
 	BringToFront
