# HG changeset patch
# Parent  84cde1c2c2c3f86a3c53e832e6f18ffc3fdaf7ff

diff --git a/runtime/doc/autocmd.txt b/runtime/doc/autocmd.txt
--- a/runtime/doc/autocmd.txt
+++ b/runtime/doc/autocmd.txt
@@ -412,6 +412,9 @@ Name			triggered by ~
 |TextChangedT|		after a change was made to the text in Terminal mode
 |TextYankPost|		after text has been yanked or deleted
 
+|SearchHitBottom|	when searching hit the bottom of the file
+|SearchHitTop|		when searching hit the top of the file
+
 |SafeState|		nothing pending, going to wait for the user to type a
 			character
 |SafeStateAgain|	repeated SafeState
@@ -1105,6 +1108,11 @@ SafeStateAgain			Like SafeState but afte
 				triggered often, don't do something that takes
 				time.
 
+							*SearchHitBottom*
+SearchHitBottom			When searching hit the bottom of the file.
+							*SearchHitTop*
+SearchHitTop			When searching hit the top of the file.
+
 							*SessionLoadPost*
 SessionLoadPost			After loading the session file created using
 				the |:mksession| command.
diff --git a/src/autocmd.c b/src/autocmd.c
--- a/src/autocmd.c
+++ b/src/autocmd.c
@@ -159,6 +159,8 @@ static struct event_name
     {"RemoteReply",	EVENT_REMOTEREPLY},
     {"SafeState",	EVENT_SAFESTATE},
     {"SafeStateAgain",	EVENT_SAFESTATEAGAIN},
+    {"SearchHitBottom",	EVENT_SEARCHHITBOTTOM},
+    {"SearchHitTop",	EVENT_SEARCHHITTOP},
     {"SessionLoadPost",	EVENT_SESSIONLOADPOST},
     {"ShellCmdPost",	EVENT_SHELLCMDPOST},
     {"ShellFilterPost",	EVENT_SHELLFILTERPOST},
diff --git a/src/search.c b/src/search.c
--- a/src/search.c
+++ b/src/search.c
@@ -1087,9 +1087,17 @@ searchit(
 	     * written.
 	     */
 	    if (dir == BACKWARD)    // start second loop at the other end
+	    {
 		lnum = buf->b_ml.ml_line_count;
+		apply_autocmds(EVENT_SEARCHHITTOP, NULL, NULL, FALSE, curbuf);
+	    }
 	    else
+	    {
 		lnum = 1;
+		apply_autocmds(EVENT_SEARCHHITBOTTOM, NULL, NULL, FALSE,
+								      curbuf);
+	    }
+
 	    if (!shortmess(SHM_SEARCH) && (options & SEARCH_MSG))
 		give_warning((char_u *)_(dir == BACKWARD
 					  ? top_bot_msg : bot_top_msg), TRUE);
diff --git a/src/vim.h b/src/vim.h
--- a/src/vim.h
+++ b/src/vim.h
@@ -1377,6 +1377,8 @@ enum auto_event
     EVENT_REMOTEREPLY,		// upon string reception from a remote vim
     EVENT_SAFESTATE,		// going to wait for a character
     EVENT_SAFESTATEAGAIN,	// still waiting for a character
+    EVENT_SEARCHHITBOTTOM,	// hit bottom when searching
+    EVENT_SEARCHHITTOP,		// hit top when searching
     EVENT_SESSIONLOADPOST,	// after loading a session file
     EVENT_SHELLCMDPOST,		// after ":!cmd"
     EVENT_SHELLFILTERPOST,	// after ":1,2!cmd", ":w !cmd", ":r !cmd".
