# HG changeset patch
# Parent  1ae4e0a6f7d59573b2776d097f3be3f6168993c8

diff --git a/nsis/gvim.nsi b/nsis/gvim.nsi
--- a/nsis/gvim.nsi
+++ b/nsis/gvim.nsi
@@ -36,6 +36,9 @@ Unicode true
 # Comment the next line if you do not want to add Native Language Support
 !define HAVE_NLS
 
+# Comment the next line if you do not want to include VisVim extension:
+!define HAVE_VIS_VIM
+
 # Comment the following line to create a multilanguage installer:
 !define HAVE_MULTI_LANG
 
@@ -586,6 +589,7 @@ SectionGroup $(str_group_plugin) id_grou
 SectionGroupEnd
 
 ##########################################################
+!ifdef HAVE_VIS_VIM
 Section "$(str_section_vis_vim)" id_section_visvim
 	SectionIn 3
 
@@ -593,6 +597,7 @@ Section "$(str_section_vis_vim)" id_sect
 	!insertmacro UpgradeDLL "${VIMSRC}\VisVim\VisVim.dll" "$0\VisVim.dll" "$0"
 	File ${VIMSRC}\VisVim\README_VisVim.txt
 SectionEnd
+!endif
 
 ##########################################################
 !ifdef HAVE_NLS
@@ -672,14 +677,18 @@ Section -post
 	  SectionGetSize ${id_section_editwith} $4
 	  IntOp $3 $3 + $4
 	${EndIf}
+!ifdef HAVE_VIS_VIM
 	${If} ${SectionIsSelected} ${id_section_visvim}
 	  SectionGetSize ${id_section_visvim} $4
 	  IntOp $3 $3 + $4
 	${EndIf}
+!endif
+!ifdef HAVE_NLS
 	${If} ${SectionIsSelected} ${id_section_nls}
 	  SectionGetSize ${id_section_nls} $4
 	  IntOp $3 $3 + $4
 	${EndIf}
+!endif
 
 	# Register EstimatedSize.
 	# Other information will be set by the install.exe.
@@ -831,11 +840,13 @@ Section "un.$(str_unsection_register)" i
 	# created.  Thus the "vim61" directory is included in it.
 	StrCpy $0 "$INSTDIR"
 
+!ifdef HAVE_VIS_VIM
 	# If VisVim was installed, unregister the DLL.
 	${If} ${FileExists} "$0\VisVim.dll"
 	  nsExec::Exec "regsvr32.exe /u /s $0\VisVim.dll"
 	  Pop $3
 	${EndIf}
+!endif
 
 	# delete the context menu entry and batch files
 	nsExec::Exec "$0\uninstal.exe -nsis"
@@ -906,7 +917,9 @@ Section "un.$(str_unsection_exe)" id_uns
 	RMDir /r $0\syntax
 	RMDir /r $0\tools
 	RMDir /r $0\tutor
+!ifdef HAVE_VIS_VIM
 	RMDir /r $0\VisVim
+!endif
 	RMDir /r $0\lang
 	RMDir /r $0\keymap
 	Delete $0\*.exe
