# HG changeset patch
# Parent  9459d517bf6548e4e994c8b99bf442bdb3134c44

diff --git a/nsis/gvim.nsi b/nsis/gvim.nsi
--- a/nsis/gvim.nsi
+++ b/nsis/gvim.nsi
@@ -68,7 +68,6 @@ SetCompressorDictSize 64
 ManifestDPIAware true
 SetDatablockOptimize on
 RequestExecutionLevel highest
-XPStyle on
 
 !ifdef HAVE_UPX
   !packhdr temp.dat "upx --best --compress-icons=1 temp.dat"
@@ -101,8 +100,8 @@ XPStyle on
 !define MUI_FINISHPAGE_RUN_TEXT            $(str_show_readme)
 !define MUI_FINISHPAGE_RUN_PARAMETERS      "-R $\"$0\README.txt$\""
 
-# This adds '\vim' to the user choice automagically.  The actual value is
-# obtained below with ReadINIStr.
+# This adds '\Vim' to the user choice automagically.  The actual value is
+# obtained below with CheckOldVim.
 InstallDir "$PROGRAMFILES\Vim"
 
 # Types of installs we can perform:
@@ -274,7 +273,7 @@ Section "$(str_section_old_ver)" id_sect
 	nsExec::Exec "$TEMP\install.exe -uninstall-check"
 	Pop $3
 	Delete $TEMP\install.exe
-	Delete $TEMP\vimini.ini
+	Delete $TEMP\vimini.ini   # install.exe creates this, but we don't need it.
 
 	# We may have been put to the background when uninstall did something.
 	BringToFront
@@ -320,7 +319,7 @@ Function .onInstSuccess
 FunctionEnd
 
 Function .onInstFailed
-  MessageBox MB_OK|MB_ICONEXCLAMATION "Installation failed. Better luck next time."
+  MessageBox MB_OK|MB_ICONEXCLAMATION "$(str_msg_install_fail)"
 FunctionEnd
 
 ##########################################################
diff --git a/nsis/lang/japanese.nsi b/nsis/lang/japanese.nsi
--- a/nsis/lang/japanese.nsi
+++ b/nsis/lang/japanese.nsi
@@ -22,7 +22,7 @@ LangString ^UninstallCaption ${LANG_JAPA
 # These should be better to be fixed by the NSIS upstream.
 LangString ^SpaceAvailable   ${LANG_JAPANESE} \
         "利用可能なディスク容量："
-LangString ^SpaceRequired ${LANG_JAPANESE} \
+LangString ^SpaceRequired    ${LANG_JAPANESE} \
         "必要なディスク容量："
 
 ##############################################################################
