# HG changeset patch
# Parent  0778674761f2fdf7492a19610526c4ab9d57ac3f

diff --git a/src/testdir/test_gui.vim b/src/testdir/test_gui.vim
--- a/src/testdir/test_gui.vim
+++ b/src/testdir/test_gui.vim
@@ -1711,6 +1711,7 @@ func Test_gui_lowlevel_keyevent()
   " Test for <Ctrl-A> to <Ctrl-Z> keys
   for kc in range(65, 90)
     call SendKeys([0x11, kc])
+    sleep 1m
     try
       let ch = getcharstr()
     catch /^Vim:Interrupt$/
