diff --git a/src/testdir/Make_dos.mak b/src/testdir/Make_dos.mak
--- a/src/testdir/Make_dos.mak
+++ b/src/testdir/Make_dos.mak
@@ -94,7 +94,7 @@ win32:	nolog $(SCRIPTS16) $(SCRIPTS) $(S
 fixff:
 	-$(VIMPROG) -u dos.vim --noplugin "+argdo set ff=dos|upd" +q *.in *.ok
 	-$(VIMPROG) -u dos.vim --noplugin "+argdo set ff=unix|upd" +q \
-		dotest.in test60.ok test71.ok test74.ok
+		dotest.in test60.ok test71.ok test74.ok test100.ok
 
 report:
 	@ECHO ""
diff --git a/src/testdir/test100.in b/src/testdir/test100.in
--- a/src/testdir/test100.in
+++ b/src/testdir/test100.in
@@ -17,6 +17,7 @@ STARTTEST
 :0put ='ONE: expecting global undolevels: 5, local undolevels: -123456 (default)'
 :call FillBuffer()
 :call feedkeys(":earlier 10\n", 't')
+:
 :call UndoLevel()
 :set ff=unix
 :%w! test.out
@@ -25,6 +26,7 @@ STARTTEST
 :setlocal ul=2
 :call FillBuffer()
 :call feedkeys(":earlier 10\n", 't')
+:
 :call UndoLevel()
 :setlocal ul=10
 :call UndoLevel()
