# HG changeset patch
# Parent  5f67cd572e9dac7b25b8d536b22dee86f4216f15

diff --git a/src/testdir/test_terminal.vim b/src/testdir/test_terminal.vim
--- a/src/testdir/test_terminal.vim
+++ b/src/testdir/test_terminal.vim
@@ -2436,6 +2436,7 @@ func Test_hidden_terminal()
 endfunc
 
 func Test_term_nasty_callback()
+  CheckExecutable sh
   func OpenTerms()
     set hidden
     let g:buf0 = term_start('sh', #{hidden: 1})
@@ -2453,7 +2454,7 @@ func Test_term_nasty_callback()
   call OpenTerms()
 
   call term_sendkeys(g:buf0, "exit\<CR>")
-  sleep 50m
+  sleep 100m
   exe g:buf0 .. 'bwipe'
   set hidden&
 endfunc
