# HG changeset patch
# Parent 2a3a688fa2d01d0e51f726087234b77dc380d3b5
diff -r 2a3a688fa2d0 -r 742c007fdf28 src/os_mswin.c
--- a/src/os_mswin.c	Sun Sep 01 13:47:36 2013 +0900
+++ b/src/os_mswin.c	Sun Sep 01 14:16:41 2013 +0900
@@ -498,6 +498,86 @@
     }
 }
 
+    static int
+stat_symlink_aware(const char *name, struct stat *stp)
+{
+#if defined(_MSC_VER) && _MSC_VER < 1700
+    WIN32_FIND_DATA	findData;
+    HANDLE		hFind, h;
+    DWORD		attr;
+    BOOL		is_symlink = FALSE;
+
+    /* Work around for VC10 or earlier. stat() can't handle a symlink to a
+     * directory. */
+    hFind = FindFirstFile(name, &findData);
+    if (hFind != INVALID_HANDLE_VALUE)
+    {
+	attr = findData.dwFileAttributes;
+	if ((attr & FILE_ATTRIBUTE_REPARSE_POINT)
+		&& (attr & FILE_ATTRIBUTE_DIRECTORY)
+		&& (findData.dwReserved0 == IO_REPARSE_TAG_SYMLINK))
+	    is_symlink = TRUE;
+	FindClose(hFind);
+    }
+    if (is_symlink)
+    {
+	h = CreateFile(name, FILE_READ_ATTRIBUTES,
+		FILE_SHARE_READ | FILE_SHARE_WRITE, NULL,
+		OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+	if (h != INVALID_HANDLE_VALUE)
+	{
+	    int	    fd, n;
+
+	    fd = _open_osfhandle((intptr_t)h, _O_RDONLY);
+	    n = _fstat(fd, (struct _stat*)stp);
+	    _close(fd);
+	    return n;
+	}
+    }
+#endif
+    return stat(name, stp);
+}
+
+    static int
+wstat_symlink_aware(const WCHAR *name, struct _stat *stp)
+{
+# if defined(_MSC_VER) && _MSC_VER < 1700
+    /* Work around for VC10 or earlier. _wstat() can't handle a symlink to a
+     * directory. */
+    int			n;
+    BOOL		is_symlink = FALSE;
+    HANDLE		hFind, h;
+    DWORD		attr;
+    WIN32_FIND_DATAW	findDataW;
+
+    hFind = FindFirstFileW(name, &findDataW);
+    if (hFind != INVALID_HANDLE_VALUE)
+    {
+	attr = findDataW.dwFileAttributes;
+	if ((attr & FILE_ATTRIBUTE_REPARSE_POINT)
+		&& (attr & FILE_ATTRIBUTE_DIRECTORY)
+		&& (findDataW.dwReserved0 == IO_REPARSE_TAG_SYMLINK))
+	    is_symlink = TRUE;
+	FindClose(hFind);
+    }
+    if (is_symlink)
+    {
+	h = CreateFileW(name, FILE_READ_ATTRIBUTES,
+		FILE_SHARE_READ | FILE_SHARE_WRITE, NULL,
+		OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+	if (h != INVALID_HANDLE_VALUE)
+	{
+	    int	    fd;
+
+	    fd = _open_osfhandle((intptr_t)h, _O_RDONLY);
+	    n = _fstat(fd, (struct _stat*)stp);
+	    _close(fd);
+	    return n;
+	}
+    }
+# endif
+    return _wstat(name, (struct _stat *)stp);
+}
 
 /*
  * stat() can't handle a trailing '/' or '\', remove it first.
@@ -514,12 +594,6 @@
     char	buf[_MAX_PATH + 1];
 #endif
     char	*p;
-#if defined(_MSC_VER) && _MSC_VER < 1700
-    WIN32_FIND_DATA	findData;
-    HANDLE		hFind, h;
-    DWORD		attr;
-    BOOL		is_symlink = FALSE;
-#endif
 
     vim_strncpy((char_u *)buf, (char_u *)name, sizeof(buf) - 1);
     p = buf + strlen(buf);
@@ -540,39 +614,7 @@
 
 	if (wp != NULL)
 	{
-# if defined(_MSC_VER) && _MSC_VER < 1700
-	    /* Work around for VC10 or earlier. _wstat() can't handle a
-	     * symlink to a directory. */
-	    WIN32_FIND_DATAW	findDataW;
-
-	    hFind = FindFirstFileW(wp, &findDataW);
-	    if (hFind != INVALID_HANDLE_VALUE)
-	    {
-		attr = findDataW.dwFileAttributes;
-		if ((attr & FILE_ATTRIBUTE_REPARSE_POINT)
-			&& (attr & FILE_ATTRIBUTE_DIRECTORY)
-			&& (findDataW.dwReserved0 == IO_REPARSE_TAG_SYMLINK))
-		    is_symlink = TRUE;
-		FindClose(hFind);
-	    }
-	    if (is_symlink)
-	    {
-		h = CreateFileW(wp, FILE_READ_ATTRIBUTES,
-			FILE_SHARE_READ | FILE_SHARE_WRITE, NULL,
-			OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
-		if (h != INVALID_HANDLE_VALUE)
-		{
-		    int	    fd;
-
-		    fd = _open_osfhandle((intptr_t)h, _O_RDONLY);
-		    n = _fstat(fd, (struct _stat*)stp);
-		    _close(fd);
-		    vim_free(wp);
-		    return n;
-		}
-	    }
-# endif
-	    n = _wstat(wp, (struct _stat *)stp);
+	    n = wstat_symlink_aware(wp, (struct _stat *)stp);
 	    vim_free(wp);
 	    if (n >= 0)
 		return n;
@@ -582,36 +624,7 @@
 	}
     }
 #endif
-#if defined(_MSC_VER) && _MSC_VER < 1700
-    /* Work around for VC10 or earlier. _stat() can't handle a
-     * symlink to a directory. */
-    hFind = FindFirstFile(buf, &findData);
-    if (hFind != INVALID_HANDLE_VALUE)
-    {
-	attr = findData.dwFileAttributes;
-	if ((attr & FILE_ATTRIBUTE_REPARSE_POINT)
-		&& (attr & FILE_ATTRIBUTE_DIRECTORY)
-		&& (findData.dwReserved0 == IO_REPARSE_TAG_SYMLINK))
-	    is_symlink = TRUE;
-	FindClose(hFind);
-    }
-    if (is_symlink)
-    {
-	h = CreateFile(buf, FILE_READ_ATTRIBUTES,
-		FILE_SHARE_READ | FILE_SHARE_WRITE, NULL,
-		OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
-	if (h != INVALID_HANDLE_VALUE)
-	{
-	    int	    fd, n;
-
-	    fd = _open_osfhandle((intptr_t)h, _O_RDONLY);
-	    n = _fstat(fd, (struct _stat*)stp);
-	    _close(fd);
-	    return n;
-	}
-    }
-#endif
-    return stat(buf, stp);
+    return stat_symlink_aware(buf, stp);
 }
 
 #if defined(FEAT_GUI_MSWIN) || defined(PROTO)
